---
title: "Orchestrating Large-Scale Single-Cell Analysis with Bioconductor"
author: Davide Risso^[University of Padova],
    Ludwig Geistlinger^[Harvard Medical School],
    Ilaria Billato^[University of Padova],
    Stefania Pirrotta^[University of Padova],
    Dario Righelli^[University of Padova],
    Marcel Ramos^[CUNY School of Public Health]
output: rmarkdown::html_vignette
date: "Last modified: 12 April, 2023."
vignette: >
  %\VignetteIndexEntry{An Example Workshop}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---


```{r, include = FALSE}
knitr::opts_chunk$set(
  collapse = TRUE,
  comment = "#>"
)
```

# Overview

## Description

In the last few years, the profiling of a large number of genome-wide features in individual cells has become routine. Consequently, a plethora of tools for the analysis of single-cell data has been developed, making it hard to understand the critical steps in the analysis workflow and the best methods for each objective of one's study.

This tutorial aims to provide a solid foundation in using Bioconductor tools for single-cell RNA-seq analysis by walking through various steps of typical workflows using example datasets. 

This tutorial uses as a "text-book" the online book [Orchestrating Single-Cell Analysis with Bioconductor (OSCA)](https://bioconductor.org/books/release/OSCA/), started in 2018 by Aaron Lun and continuously updated by many contributors from the Bioconductor community. Like the book, this tutorial strives to be of interest to the experimental biologists wanting to analyze their data and to the bioinformaticians approaching single-cell data.

## Learning objectives

Attendees will learn how to analyze multi-condition single-cell RNA-seq from raw data to statistical analyses and result interpretation. Students will learn where the critical steps and methods choices are and will be able to leverage large-data resources to analyze datasets comprising millions of cells.

In particular, students will learn:

- How to access publicly available data, such as those from the Human Cell Atlas.
- How to perform data exploration, normalization, and dimensionality reduction.
- How to identify cell types/states and marker genes.
- How to correct for batch effects and integrate multiple samples.
- How to perform differential expression and differential abundance analysis between conditions.
- How to work with large out-of-memory datasets.

## Pre-requisites

No previous experience with the analysis of single-cell data is required. 

We assume that the students have a working knowledge of single-cell RNA-seq technology (e.g., what is RNA sequencing, what kind of data are generated from the experiments, ...).

Basic knowledge of R is also expected, but no previous knowledge of Bioconductor and its packages is required as all the basics will be covered as part of the tutorial.

## Participation

Each session will include a short presentation with slides, followed by hands-on tutorial-style lectures, in which the speaker will show how to perform the tasks directly in RStudio. There will be ample room for individual exercises in each session. The last two sessions will be dedicated to hands-on case studies, in which students will be divided into small groups and will have the opportunity to work on a real dataset at their own pace.

## _R_ / _Bioconductor_ packages used

- SingleCellExperiment
- ...

## Time outline


| Activity                     | Time |
|------------------------------|------|
| Introduction and Setup.                                         | 9:00-9:30  |
| Introduction to Bioconductor and the SingleCellExperiment class | 9:30-10:00  |
| Contributing to Bioconductor                                    | 10:00-10:45   |
| Exploratory Data Analysis and Quality Control (EDA/QC)          | 10:45-11:00  |
| Clustering and cell type annotation                             | 11:00-12:00  |
| Multi-sample analyses                                           | 12:00-13:00  |
| Lunch break                                                     | 13:00-14:00  |
| Accessing the Human Cell Atlas (HCA) Data from R/Bioconductor   | 14:00-15:00  |
| Working with large data                                         | 15:00-16:00  |
| Coffee break                                                    | 16:00-16:15  |
| Case study: from data import to DE and DA                       | 16:15-17:00  |
| Case study: discussion                                          | 17:00-18:00  |





