---
title: Cell type annotation
vignette: >
  % \VignetteIndexEntry{Cell type annotation}
  % \VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
output:
  html_document:
    mathjax: null   
---

```{r chunk-opts, include=FALSE}
rm(list = ls())
gc()
knitr::opts_chunk$set(echo = TRUE, cache = TRUE)
library(BiocStyle)
```

# Setup

```{r setup, message = FALSE}
library(AUCell)
library(MouseGastrulationData)
library(SingleR)
library(bluster)
library(scater)
library(scran)
```

# Data retrieval

```{r data, message = FALSE}
sce <- WTChimeraData(samples = 5, type = "processed")
sce
```

To speed up the computations, we subsample the dataset to 1,000 cells.

```{r}
set.seed(123)
ind <- sample(ncol(sce), 1000)
sce <- sce[,ind]
```

# Preprocessing

```{r preproc, warning = FALSE}
sce <- logNormCounts(sce)
sce <- runPCA(sce)
```

# Clustering

Clustering is an unsupervised learning procedure that is used to empirically 
define groups of cells with similar expression profiles. 
Its primary purpose is to summarize complex scRNA-seq data into a digestible 
format for human interpretation. 
This allows us to describe population heterogeneity in terms of discrete labels 
that are easily understood, rather than attempting to comprehend the 
high-dimensional manifold on which the cells truly reside. After annotation 
based on marker genes, the clusters can be treated as proxies for more abstract 
biological concepts such as cell types or states.

Popularized by its use in
[Seurat](https://cran.r-project.org/web/packages/Seurat/index.html), 
graph-based clustering is a flexible and
scalable technique for clustering large scRNA-seq datasets. We first build a 
graph where each node is a cell that is connected to its nearest neighbors in
the high-dimensional space. Edges are weighted based on the similarity between 
the cells involved, with higher weight given to cells that are more closely 
related. We then apply algorithms to identify "communities" of cells that are 
more connected to cells in the same community than they are to cells of
different communities. Each community represents a cluster that we can use for 
downstream interpretation.

Here, we use the `clusterCells()` function from the 
[scran](https://bioconductor.org/packages/scran) package to perform graph-based
clustering using the 
[Louvain algorithm](https://doi.org/10.1088/1742-5468/2008/10/P10008)
for community detection. All calculations are performed using the top PCs to 
take advantage of data compression and denoising. This function returns a vector
containing cluster assignments for each cell in our `SingleCellExperiment` object.

```{r cluster}
colLabels(sce) <- clusterCells(sce, use.dimred = "PCA",
                               BLUSPARAM = NNGraphParam(cluster.fun = "louvain"))
table(colLabels(sce))
```

We assign the cluster assignments back into our `SingleCellExperiment` object as
a `factor` in the column metadata. This allows us to conveniently visualize the
distribution of clusters in eg. a *t*-SNE or a UMAP.

```{r cluster-viz}
sce <- runUMAP(sce, dimred = "PCA")
plotReducedDim(sce, "UMAP", color_by = "label")
```

# Marker gene detection

To interpret clustering results as obtained in the previous section, we identify
the genes that drive separation between clusters. These marker genes allow us to 
assign biological meaning to each cluster based on their functional annotation. 
In the simplest case, we have *a priori* knowledge of the marker genes associated 
with particular cell types, allowing us to treat the clustering as a proxy for 
cell type identity.
 
The most straightforward approach to marker gene detection involves testing for 
differential expression between clusters. If a gene is strongly DE between 
clusters, it is likely to have driven the separation of cells in the clustering 
algorithm.

Here, we perform a Wilcoxon rank sum test against a log2 fold change threshold
of 1, focusing on up-regulated (positive) markers in one cluster when compared
to another cluster. 

```{r marker-detect}
rownames(sce) <- rowData(sce)$SYMBOL
markers <- findMarkers(sce, test.type = "wilcox", direction = "up", lfc = 1)
markers
```

The resulting object contains a sorted marker gene list for each cluster,
in which the top genes are those that contribute the most to the separation of
that cluster from mall other clusters. 

Here, we inspect the ranked marker gene list for the first cluster.

```{r marker-clust1}
markers[[1]]
```

The `Top` field provides the the minimum rank across all pairwise comparisons.
The `p.value` field provides the combined *p*-value across all comparisons, and 
the `FDR` field the BH-adjusted *p*-value for each gene.
The `summary.AUC` provides area under the curve (here the concordance probability)
from the comparison with the lowest *p*-value, the `AUC.n` fields provide the
AUC for each pairwise comparison. The AUC is the probability that a randomly
selected cell in cluster *A* has a greater expression of gene *X* than a randomly
selected cell in *B*.

We can then inspect the top marker genes for the first cluster using the 
`plotExpression` function from the 
[scater](https://bioconductor.org/packages/scater) package.

```{r plot-markers, fig.width = 10, fig.height = 10}
top.markers <- head(rownames(markers[[1]]))
plotExpression(sce, features = top.markers, x = "label", color_by = "label")
```

# Cell type annotation

The most challenging task in scRNA-seq data analysis is arguably the
interpretation of the results.
Obtaining clusters of cells is fairly straightforward, but it is more difficult
to determine what biological state is represented by each of those clusters. 
Doing so requires us to bridge the gap between the current dataset and prior 
biological knowledge, and the latter is not always available in a consistent 
and quantitative manner.
Indeed, even the concept of a "cell type" is
[not clearly defined](https://doi.org/10.1016/j.cels.2017.03.006), with most
practitioners possessing a "I'll know it when I see it" intuition that is not
amenable to computational analysis.
As such, interpretation of scRNA-seq data is often manual and a common
bottleneck in the analysis workflow.

To expedite this step, we can use various computational approaches that exploit
prior information to assign meaning to an uncharacterized scRNA-seq dataset.
The most obvious sources of prior information are the curated gene sets associated
with particular biological processes, e.g., from the Gene Ontology (GO) or the
Kyoto Encyclopedia of Genes and Genomes (KEGG) collections.
Alternatively, we can directly compare our expression profiles to published
reference datasets where each sample or cell has already been annotated with its
putative biological state by domain experts.
Here, we will demonstrate both approaches on the wild-type chimera dataset.

## Assigning cell labels from reference data

A conceptually straightforward annotation approach is to compare the single-cell
expression profiles with previously annotated reference datasets.
Labels can then be assigned to each cell in our uncharacterized test dataset
based on the most similar reference sample(s), for some definition of "similar".
This is a standard classification challenge that can be tackled by standard
machine learning techniques such as random forests and support vector machines.
Any published and labelled RNA-seq dataset (bulk or single-cell) can be used as
a reference, though its reliability depends greatly on the expertise of the
original authors who assigned the labels in the first place. 

In this section, we will demonstrate the use of the `r Biocpkg("SingleR")`
method for cell type annotation
[Aran et al., 2019](https://www.nature.com/articles/s41590-018-0276-y).
This method assigns labels to cells based on the reference samples with the
highest Spearman rank correlations, using only the marker genes between pairs of
labels to focus on the relevant differences between cell types.
It also performs a fine-tuning step for each cell where the correlations are
recomputed with just the marker genes for the top-scoring labels.
This aims to resolve any ambiguity between those labels by removing noise from
irrelevant markers for other labels.
Further details can be found in the
[_SingleR_ book](https://bioconductor.org/books/release/SingleRBook) from which
most of the examples here are derived.

```{r ref-data, message = FALSE}
ref <- EmbryoAtlasData(samples = 29)
ref
table(ref$stage)
```

To speed up the computations, we subsample the dataset to 1,000 cells.

```{r}
set.seed(123)
ind <- sample(ncol(ref), 1000)
ref <- ref[,ind]
```

```{r ref-celltypes}
tab <- sort(table(ref$celltype), decreasing = TRUE)
tab
```

```{r ref-preproc}
ref <- logNormCounts(ref)
```

Some cleaning - remove cells of the reference dataset for which the cell type
annotation is missing.

```{r na-celltype}
nna <- !is.na(ref$celltype)
ref <- ref[,nna]
```

Also remove cell types of very low abundance (here less than 10 cells) to remove
noise prior to subsequent annotation tasks.

```{r low-abu-ct}
abu.ct <- names(tab)[tab >= 10]
ind <- ref$celltype %in% abu.ct
ref <- ref[,ind] 
```

Restrict to genes shared between query and reference dataset.

```{r shared-genes}
rownames(ref) <- rowData(ref)$SYMBOL
isect <- intersect(rownames(sce), rownames(ref))
sce <- sce[isect,]
ref <- ref[isect,]
```

Convert sparse assay matrices to regular dense matrices for input to SingleR. 

```{r sparse-to-dense-mat}
sce.mat <- as.matrix(assay(sce, "logcounts"))
ref.mat <- as.matrix(assay(ref, "logcounts"))
```

```{r singler}
res <- SingleR(test = sce.mat, ref = ref.mat, labels = ref$celltype)
res
```

We inspect the results using a heatmap of the per-cell and label scores.
Ideally, each cell should exhibit a high score in one label relative to all of
the others, indicating that the assignment to that label was unambiguous. 
This is largely the case for mesenchyme and endothelial cells, whereas we see
expectedly more ambiguity between the two erythroid cell populations.

```{r score-heat, fig.width = 10, fig.height = 10}
plotScoreHeatmap(res)
```

We also compare the cell type assignments with the clustering results to determine
the identity of each cluster. Here, several cell type classes are nested within
the same cluster, indicating that these clusters are composed of several 
transcriptomically similar cell populations (such as cluster 4 and 6).
On the other hand, there are also instances where we have several clusters for
the same cell type, indicating that the clustering represents finer subdivisions
within these cell types. 

```{r, fig.width = 10, fig.height = 10}
library(pheatmap)
tab <- table(anno = res$pruned.labels, cluster = colLabels(sce))
pheatmap(log2(tab + 10), color = colorRampPalette(c("white", "blue"))(101))
```

As it so happens, we are in the fortunate position where our test dataset also
contains independently defined labels. We see strong consistency between the two
sets of labels, indicating that our automatic annotation is comparable to that
generated manually by domain experts.

```{r anno-vs-preanno, fig.width = 10, fig.height = 10}
tab <- table(res$pruned.labels, sce$celltype.mapped)
pheatmap(log2(tab + 10), color = colorRampPalette(c("white", "blue"))(101))
```

## Assigning cell labels from gene sets

A related strategy is to explicitly identify sets of marker genes that are highly
expressed in each individual cell.
This does not require matching of individual cells to the expression values of
the reference dataset, which is faster and more convenient when only the
identities of the markers are available.
We demonstrate this approach using cell type markers derived from the
mouse embryo atlas dataset.

```{r atlas-markers}
library(scran)
wilcox.z <- pairwiseWilcox(ref, ref$celltype, lfc = 1, direction = "up")
markers.z <- getTopMarkers(wilcox.z$statistics, wilcox.z$pairs, 
                           pairwise = FALSE, n = 50)
lengths(markers.z)
```

Our test dataset will be as before the wild-type chimera dataset.

```{r wt-sce}
sce
```

We use the `r Biocpkg("AUCell")` package to identify marker sets that are highly
expressed in each cell.
This method ranks genes by their expression values within each cell and constructs
a response curve of the number of genes from each marker set that are present
with increasing rank.
It then computes the area under the curve (AUC) for each marker set, quantifying
the enrichment of those markers among the most highly expressed genes in that cell.
This is roughly similar to performing a Wilcoxon rank sum test between genes in
and outside of the set, but involving only the top ranking genes by expression
in each cell.

```{r create-gsets, message = FALSE}
library(GSEABase)
all.sets <- lapply(names(markers.z), 
                   function(x) GeneSet(markers.z[[x]], setName = x))
all.sets <- GeneSetCollection(all.sets)
all.sets
```

```{r aucell}
library(AUCell)
rankings <- AUCell_buildRankings(as.matrix(counts(sce)),
                                 plotStats = FALSE, verbose = FALSE)
cell.aucs <- AUCell_calcAUC(all.sets, rankings)
results <- t(assay(cell.aucs))
head(results)
```

We assign cell type identity to each cell in the test dataset by taking the
marker set with the top AUC as the label for that cell.
Our new labels mostly agree with the original annotation (and, thus, also with
the reference-based annotation).
Instances where the original annotation is divided into several new label groups
typically points to large overlaps in their marker sets.
In the absence of prior annotation, a more general diagnostic check is to compare
the assigned labels to cluster identities, under the expectation that most cells
of a single cluster would have the same label (or, if multiple labels are present,
they should at least represent closely related cell states).

```{r anno-vs-cluster}
new.labels <- colnames(results)[max.col(results)]
tab <- table(new.labels, sce$celltype.mapped)
tab
```

As a diagnostic measure, we examine the distribution of AUCs across cells for
each label.
In heterogeneous populations, the distribution for each label should be bimodal
with one high-scoring peak containing cells of that cell type and a low-scoring
peak containing cells of other types.
The gap between these two peaks can be used to derive a threshold for whether a
label is "active" for a particular cell.
(In this case, we simply take the single highest-scoring label per cell as the
labels should be mutually exclusive.)
In populations where a particular cell type is expected, lack of clear bimodality
for the corresponding label may indicate that its gene set is not sufficiently
informative.

```{r auc-dist, results="hide", fig.width = 10, fig.height = 10}
par(mfrow = c(3,3))
AUCell_exploreThresholds(cell.aucs[1:9], plotHist = TRUE, assign = TRUE) 
```

Shown is the distribution of AUCs in the wild-type chimera dataset for each label in
the embryo atlas dataset. The blue curve represents the density estimate, the red curve
represents a fitted two-component mixture of normals, the pink curve represents
a fitted three-component mixture, and the grey curve represents a fitted normal
distribution. Vertical lines represent threshold estimates corresponding to each
estimate of the distribution. 

# Session Info

```{r sessionInfo}
sessionInfo()
```

# Further Reading

* OSCA book, [Chapters 5-7](https://bioconductor.org/books/release/OSCA.basic/clustering.html)
* Assigning cell types with SingleR ([the book](https://bioconductor.org/books/release/SingleRBook/)).
* The [AUCell](https://bioconductor.org/packages/AUCell) package vignette.

# Exercises

Exercise: The [Leiden algorithm](https://www.nature.com/articles/s41598-019-41695-z)
is similar to the Louvain algorithm, but it is faster and has been shown to result
in better connected communities. Modify the above call to `clusterCells` to
carry out the community detection with the Leiden algorithm instead. Visualize
the results in a UMAP plot.

Hint: The `NNGraphParam` constructor has an argument `cluster.args`. This allows
to specify arguments passed on to the `cluster_leiden` function from the 
[igraph](https://cran.r-project.org/web/packages/igraph/index.html) package.
Use the `cluster.args` argument to parameterize the clustering to use modularity
as the objective function and a resolution parameter of 0.5.
