<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>Accessing data from the Human Cell Atlas (HCA) • ISMB.OSCA</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link href="https://cdnjs.cloudflare.com/ajax/libs/bootswatch/3.4.0/flatly/bootstrap.min.css" rel="stylesheet" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="Accessing data from the Human Cell Atlas (HCA)">
<meta property="og:description" content="ISMB.OSCA">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">
    

    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">ISMB.OSCA</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">0.99.0</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Vignettes
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../articles/intro_bioc_sce.html">Introduction to Bioconductor</a>
    </li>
    <li>
      <a href="../articles/eda_qc.html">Exploratory data analysis</a>
    </li>
    <li>
      <a href="../articles/cell_type_annotation.html">Cell type annotation</a>
    </li>
    <li>
      <a href="../articles/multi_sample.html">Multi-sample analyses</a>
    </li>
    <li>
      <a href="../articles/large_data.html">Working with large data</a>
    </li>
    <li>
      <a href="../articles/hca.html">Accessing data from the Human Cell Atlas</a>
    </li>
  </ul>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://github.com/Bioconductor/ISMB.OSCA" class="external-link">
    <span class="fa fa-github"></span>
     
  </a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1 data-toc-skip>Accessing data from the Human Cell Atlas
(HCA)</h1>
            
      
      <small class="dont-index">Source: <a href="https://github.com/Bioconductor/ISMB.OSCA/blob/HEAD/vignettes/hca.Rmd" class="external-link"><code>vignettes/hca.Rmd</code></a></small>
      <div class="hidden name"><code>hca.Rmd</code></div>

    </div>

    
    
<div class="section level2">
<h2 id="hca-project">HCA Project<a class="anchor" aria-label="anchor" href="#hca-project"></a>
</h2>
<p>The Human Cell Atlas (HCA) is a large project that aims to learn from
and map every cell type in the human body. The project extracts spatial
and molecular characteristics in order to understand cellular function
and networks. It is an international collaborative that charts healthy
cells in the human body at all ages. There are about 37.2 trillion cells
in the human body. To read more about the project, head over to their
website at <a href="https://www.humancellatlas.org" class="external-link uri">https://www.humancellatlas.org</a>.</p>
</div>
<div class="section level2">
<h2 id="cellxgene">CELLxGENE<a class="anchor" aria-label="anchor" href="#cellxgene"></a>
</h2>
<p>CELLxGENE is a database and a suite of tools that help scientists to
find, download, explore, analyze, annotate, and publish single cell
data. It includes several analytic and visualization tools to help you
to discover single cell data patterns. To see the list of tools, browse
to <a href="https://cellxgene.cziscience.com/" class="external-link uri">https://cellxgene.cziscience.com/</a>.</p>
</div>
<div class="section level2">
<h2 id="the-curatedatlasqueryr-project">The CuratedAtlasQueryR Project<a class="anchor" aria-label="anchor" href="#the-curatedatlasqueryr-project"></a>
</h2>
<p>To systematically characterize the immune system across tissues,
demographics and multiple studies, single cell transcriptomics data was
harmonized from the CELLxGENE database. Data from 28,975,366 cells that
cover 156 tissues (excluding cell cultures), 12,981 samples, and 324
studies were collected. The metadata was standardized, including sample
identifiers, tissue labels (based on anatomy) and age. Also, the
gene-transcript abundance of all samples was harmonized by putting
values on the positive natural scale (i.e. non-logarithmic).</p>
<p>To model the immune system across studies, we adopted a consistent
immune cell-type ontology appropriate for lymphoid and non-lymphoid
tissues. We applied a consensus cell labeling strategy between the
Seurat blueprint and Monaco <span class="citation">(2019)</span> to
minimize biases in immune cell classification from study-specific
standards.</p>
<p><code>CuratedAtlasQueryR</code> supports data access and programmatic
exploration of the harmonized atlas. Cells of interest can be selected
based on ontology, tissue of origin, demographics, and disease. For
example, the user can select CD4 T helper cells across healthy and
diseased lymphoid tissue. The data for the selected cells can be
downloaded locally into popular single-cell data containers. Pseudo bulk
counts are also available to facilitate large-scale, summary analyses of
transcriptional profiles. This platform offers a standardized workflow
for accessing atlas-level datasets programmatically and
reproducibly.</p>
<p><img src="figures/HCA_sccomp_SUPPLEMENTARY_technical_cartoon_curatedAtlasQuery.png" width="1110"></p>
</div>
<div class="section level2">
<h2 id="data-sources-in-r-bioconductor">Data Sources in R / Bioconductor<a class="anchor" aria-label="anchor" href="#data-sources-in-r-bioconductor"></a>
</h2>
<p>There are a few options to access single cell data with R /
Bioconductor.</p>
<table class="table">
<colgroup>
<col width="29%">
<col width="41%">
<col width="29%">
</colgroup>
<thead><tr class="header">
<th>Package</th>
<th>Target</th>
<th>Description</th>
</tr></thead>
<tbody>
<tr class="odd">
<td><a href="https://bioconductor.org/packages/hca" class="external-link">hca</a></td>
<td><a href="https://www.humancellatlas.org/data-portal/" class="external-link">HCA Data
Portal API</a></td>
<td>Project, Sample, and File level HCA data</td>
</tr>
<tr class="even">
<td><a href="https://bioconductor.org/packages/cellxgenedp" class="external-link">cellxgenedp</a></td>
<td><a href="https://cellxgene.cziscience.com/" class="external-link">CellxGene</a></td>
<td>Human and mouse SC data including HCA</td>
</tr>
<tr class="odd">
<td><a href="https://stemangiola.github.io/CuratedAtlasQueryR/" class="external-link">CuratedAtlasQueryR</a></td>
<td><a href="https://cellxgene.cziscience.com/" class="external-link">CellxGene</a></td>
<td>fine-grained query capable CELLxGENE data including HCA</td>
</tr>
</tbody>
</table>
</div>
<div class="section level2">
<h2 id="installation">Installation<a class="anchor" aria-label="anchor" href="#installation"></a>
</h2>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw">if</span> <span class="op">(</span><span class="op">!</span><span class="fu"><a href="https://rdrr.io/r/base/ns-load.html" class="external-link">requireNamespace</a></span><span class="op">(</span><span class="st">"BiocManager"</span>, quietly <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span><span class="op">)</span></span>
<span>    <span class="fu"><a href="https://rdrr.io/r/utils/install.packages.html" class="external-link">install.packages</a></span><span class="op">(</span><span class="st">"BiocManager"</span><span class="op">)</span></span>
<span></span>
<span><span class="fu">BiocManager</span><span class="fu">::</span><span class="fu"><a href="https://bioconductor.github.io/BiocManager/reference/install.html" class="external-link">install</a></span><span class="op">(</span><span class="st">"stemangiola/CuratedAtlasQueryR"</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="package-load">Package load<a class="anchor" aria-label="anchor" href="#package-load"></a>
</h2>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/stemangiola/CuratedAtlasQueryR" class="external-link">CuratedAtlasQueryR</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://dplyr.tidyverse.org" class="external-link">dplyr</a></span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="hca-metadata">HCA Metadata<a class="anchor" aria-label="anchor" href="#hca-metadata"></a>
</h2>
<p>The metadata allows the user to get a lay of the land of what is
available via the package.</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">metadata</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/CuratedAtlasQueryR/man/get_metadata.html" class="external-link">get_metadata</a></span><span class="op">(</span>remote_url <span class="op">=</span> <span class="fu">CuratedAtlasQueryR</span><span class="fu">::</span><span class="va"><a href="https://rdrr.io/pkg/CuratedAtlasQueryR/man/SAMPLE_DATABASE_URL.html" class="external-link">SAMPLE_DATABASE_URL</a></span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="summarizing-the-metadata">Summarizing the metadata<a class="anchor" aria-label="anchor" href="#summarizing-the-metadata"></a>
</h2>
<p>For each distinct tissue and dataset combination, count the number of
datasets by tissue type.</p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">metadata</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/distinct.html" class="external-link">distinct</a></span><span class="op">(</span><span class="va">tissue</span>, <span class="va">dataset_id</span><span class="op">)</span> <span class="op">|&gt;</span> </span>
<span>  <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/count.html" class="external-link">count</a></span><span class="op">(</span><span class="va">tissue</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## <span style="color: #949494;"># Source:   SQL [?? x 2]</span></span></span>
<span><span class="co">## <span style="color: #949494;"># Database: DuckDB 0.8.0 [unknown@Linux 5.15.0-1039-azure:R 4.3.0/:memory:]</span></span></span>
<span><span class="co">##    tissue                          n</span></span>
<span><span class="co">##    <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>                       <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span></span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 1</span> transition zone of prostate     2</span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 2</span> peripheral zone of prostate     2</span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 3</span> heart left ventricle            7</span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 4</span> lung                            4</span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 5</span> blood                          17</span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 6</span> bone marrow                     4</span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 7</span> kidney                          8</span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 8</span> renal medulla                   6</span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 9</span> cortex of kidney                7</span></span>
<span><span class="co">## <span style="color: #BCBCBC;">10</span> renal pelvis                    4</span></span>
<span><span class="co">## <span style="color: #949494;"># ℹ more rows</span></span></span></code></pre>
</div>
<div class="section level2">
<h2 id="columns-available-in-the-metadata">Columns available in the metadata<a class="anchor" aria-label="anchor" href="#columns-available-in-the-metadata"></a>
</h2>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">metadata</span><span class="op">)</span>, <span class="fl">10</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">##  [1] "cell_"                             "sample_"                          </span></span>
<span><span class="co">##  [3] "cell_type"                         "cell_type_harmonised"             </span></span>
<span><span class="co">##  [5] "confidence_class"                  "cell_annotation_azimuth_l2"       </span></span>
<span><span class="co">##  [7] "cell_annotation_blueprint_singler" "cell_annotation_monaco_singler"   </span></span>
<span><span class="co">##  [9] "sample_id_db"                      "_sample_name"</span></span></code></pre>
<div class="section level3">
<h3 id="download-single-cell-rna-sequencing-counts">Download single-cell RNA sequencing counts<a class="anchor" aria-label="anchor" href="#download-single-cell-rna-sequencing-counts"></a>
</h3>
<div class="section level4">
<h4 id="query-raw-counts">Query raw counts<a class="anchor" aria-label="anchor" href="#query-raw-counts"></a>
</h4>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">single_cell_counts</span> <span class="op">&lt;-</span> </span>
<span>    <span class="va">metadata</span> <span class="op">|&gt;</span></span>
<span>    <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/filter.html" class="external-link">filter</a></span><span class="op">(</span></span>
<span>        <span class="va">ethnicity</span> <span class="op">==</span> <span class="st">"African"</span> <span class="op">&amp;</span></span>
<span>        <span class="fu">stringr</span><span class="fu">::</span><span class="fu"><a href="https://stringr.tidyverse.org/reference/str_like.html" class="external-link">str_like</a></span><span class="op">(</span><span class="va">assay</span>, <span class="st">"%10x%"</span><span class="op">)</span> <span class="op">&amp;</span></span>
<span>        <span class="va">tissue</span> <span class="op">==</span> <span class="st">"lung parenchyma"</span> <span class="op">&amp;</span></span>
<span>        <span class="fu">stringr</span><span class="fu">::</span><span class="fu"><a href="https://stringr.tidyverse.org/reference/str_like.html" class="external-link">str_like</a></span><span class="op">(</span><span class="va">cell_type</span>, <span class="st">"%CD4%"</span><span class="op">)</span></span>
<span>    <span class="op">)</span> <span class="op">|&gt;</span></span>
<span>    <span class="fu"><a href="https://rdrr.io/pkg/CuratedAtlasQueryR/man/get_single_cell_experiment.html" class="external-link">get_single_cell_experiment</a></span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## <span style="color: #00BBBB;">ℹ</span> Realising metadata.</span></span></code></pre>
<pre><code><span><span class="co">## <span style="color: #00BBBB;">ℹ</span> Synchronising files</span></span></code></pre>
<pre><code><span><span class="co">## <span style="color: #00BBBB;">ℹ</span> Downloading 0 files, totalling 0 GB</span></span></code></pre>
<pre><code><span><span class="co">## <span style="color: #00BBBB;">ℹ</span> Reading files.</span></span></code></pre>
<pre><code><span><span class="co">## <span style="color: #00BBBB;">ℹ</span> Compiling Single Cell Experiment.</span></span></code></pre>
<div class="sourceCode" id="cb14"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">single_cell_counts</span></span></code></pre></div>
<pre><code><span><span class="co">## class: SingleCellExperiment </span></span>
<span><span class="co">## dim: 36229 1571 </span></span>
<span><span class="co">## metadata(0):</span></span>
<span><span class="co">## assays(1): counts</span></span>
<span><span class="co">## rownames(36229): A1BG A1BG-AS1 ... ZZEF1 ZZZ3</span></span>
<span><span class="co">## rowData names(0):</span></span>
<span><span class="co">## colnames(1571): ACACCAAAGCCACCTG_SC18_1 TCAGCTCCAGACAAGC_SC18_1 ...</span></span>
<span><span class="co">##   CAGCATAAGCTAACAA_F02607_1 AAGGAGCGTATAATGG_F02607_1</span></span>
<span><span class="co">## colData names(56): sample_ cell_type ... updated_at_y original_cell_id</span></span>
<span><span class="co">## reducedDimNames(0):</span></span>
<span><span class="co">## mainExpName: NULL</span></span>
<span><span class="co">## altExpNames(0):</span></span></code></pre>
</div>
<div class="section level4">
<h4 id="query-counts-scaled-per-million">Query counts scaled per million<a class="anchor" aria-label="anchor" href="#query-counts-scaled-per-million"></a>
</h4>
<p>This is helpful if just few genes are of interest, as they can be
compared across samples.</p>
<div class="sourceCode" id="cb16"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">metadata</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/filter.html" class="external-link">filter</a></span><span class="op">(</span></span>
<span>      <span class="va">ethnicity</span> <span class="op">==</span> <span class="st">"African"</span> <span class="op">&amp;</span></span>
<span>      <span class="fu">stringr</span><span class="fu">::</span><span class="fu"><a href="https://stringr.tidyverse.org/reference/str_like.html" class="external-link">str_like</a></span><span class="op">(</span><span class="va">assay</span>, <span class="st">"%10x%"</span><span class="op">)</span> <span class="op">&amp;</span></span>
<span>      <span class="va">tissue</span> <span class="op">==</span> <span class="st">"lung parenchyma"</span> <span class="op">&amp;</span></span>
<span>      <span class="fu">stringr</span><span class="fu">::</span><span class="fu"><a href="https://stringr.tidyverse.org/reference/str_like.html" class="external-link">str_like</a></span><span class="op">(</span><span class="va">cell_type</span>, <span class="st">"%CD4%"</span><span class="op">)</span></span>
<span>  <span class="op">)</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu"><a href="https://rdrr.io/pkg/CuratedAtlasQueryR/man/get_single_cell_experiment.html" class="external-link">get_single_cell_experiment</a></span><span class="op">(</span>assays <span class="op">=</span> <span class="st">"cpm"</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## <span style="color: #00BBBB;">ℹ</span> Realising metadata.</span></span></code></pre>
<pre><code><span><span class="co">## <span style="color: #00BBBB;">ℹ</span> Synchronising files</span></span></code></pre>
<pre><code><span><span class="co">## <span style="color: #00BBBB;">ℹ</span> Downloading 0 files, totalling 0 GB</span></span></code></pre>
<pre><code><span><span class="co">## <span style="color: #00BBBB;">ℹ</span> Reading files.</span></span></code></pre>
<pre><code><span><span class="co">## <span style="color: #00BBBB;">ℹ</span> Compiling Single Cell Experiment.</span></span></code></pre>
<pre><code><span><span class="co">## class: SingleCellExperiment </span></span>
<span><span class="co">## dim: 36229 1571 </span></span>
<span><span class="co">## metadata(0):</span></span>
<span><span class="co">## assays(1): cpm</span></span>
<span><span class="co">## rownames(36229): A1BG A1BG-AS1 ... ZZEF1 ZZZ3</span></span>
<span><span class="co">## rowData names(0):</span></span>
<span><span class="co">## colnames(1571): ACACCAAAGCCACCTG_SC18_1 TCAGCTCCAGACAAGC_SC18_1 ...</span></span>
<span><span class="co">##   CAGCATAAGCTAACAA_F02607_1 AAGGAGCGTATAATGG_F02607_1</span></span>
<span><span class="co">## colData names(56): sample_ cell_type ... updated_at_y original_cell_id</span></span>
<span><span class="co">## reducedDimNames(0):</span></span>
<span><span class="co">## mainExpName: NULL</span></span>
<span><span class="co">## altExpNames(0):</span></span></code></pre>
</div>
<div class="section level4">
<h4 id="extract-only-a-subset-of-genes">Extract only a subset of genes<a class="anchor" aria-label="anchor" href="#extract-only-a-subset-of-genes"></a>
</h4>
<div class="sourceCode" id="cb23"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">single_cell_counts</span> <span class="op">&lt;-</span></span>
<span>    <span class="va">metadata</span> <span class="op">|&gt;</span></span>
<span>    <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/filter.html" class="external-link">filter</a></span><span class="op">(</span></span>
<span>        <span class="va">ethnicity</span> <span class="op">==</span> <span class="st">"African"</span> <span class="op">&amp;</span></span>
<span>        <span class="fu">stringr</span><span class="fu">::</span><span class="fu"><a href="https://stringr.tidyverse.org/reference/str_like.html" class="external-link">str_like</a></span><span class="op">(</span><span class="va">assay</span>, <span class="st">"%10x%"</span><span class="op">)</span> <span class="op">&amp;</span></span>
<span>        <span class="va">tissue</span> <span class="op">==</span> <span class="st">"lung parenchyma"</span> <span class="op">&amp;</span></span>
<span>        <span class="fu">stringr</span><span class="fu">::</span><span class="fu"><a href="https://stringr.tidyverse.org/reference/str_like.html" class="external-link">str_like</a></span><span class="op">(</span><span class="va">cell_type</span>, <span class="st">"%CD4%"</span><span class="op">)</span></span>
<span>    <span class="op">)</span> <span class="op">|&gt;</span></span>
<span>    <span class="fu"><a href="https://rdrr.io/pkg/CuratedAtlasQueryR/man/get_single_cell_experiment.html" class="external-link">get_single_cell_experiment</a></span><span class="op">(</span>assays <span class="op">=</span> <span class="st">"cpm"</span>, features <span class="op">=</span> <span class="st">"PUM1"</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## <span style="color: #00BBBB;">ℹ</span> Realising metadata.</span></span></code></pre>
<pre><code><span><span class="co">## <span style="color: #00BBBB;">ℹ</span> Synchronising files</span></span></code></pre>
<pre><code><span><span class="co">## <span style="color: #00BBBB;">ℹ</span> Downloading 0 files, totalling 0 GB</span></span></code></pre>
<pre><code><span><span class="co">## <span style="color: #00BBBB;">ℹ</span> Reading files.</span></span></code></pre>
<pre><code><span><span class="co">## <span style="color: #00BBBB;">ℹ</span> Compiling Single Cell Experiment.</span></span></code></pre>
<div class="sourceCode" id="cb29"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">single_cell_counts</span></span></code></pre></div>
<pre><code><span><span class="co">## class: SingleCellExperiment </span></span>
<span><span class="co">## dim: 1 1571 </span></span>
<span><span class="co">## metadata(0):</span></span>
<span><span class="co">## assays(1): cpm</span></span>
<span><span class="co">## rownames(1): PUM1</span></span>
<span><span class="co">## rowData names(0):</span></span>
<span><span class="co">## colnames(1571): ACACCAAAGCCACCTG_SC18_1 TCAGCTCCAGACAAGC_SC18_1 ...</span></span>
<span><span class="co">##   CAGCATAAGCTAACAA_F02607_1 AAGGAGCGTATAATGG_F02607_1</span></span>
<span><span class="co">## colData names(56): sample_ cell_type ... updated_at_y original_cell_id</span></span>
<span><span class="co">## reducedDimNames(0):</span></span>
<span><span class="co">## mainExpName: NULL</span></span>
<span><span class="co">## altExpNames(0):</span></span></code></pre>
</div>
<div class="section level4">
<h4 id="extracting-counts-as-a-seurat-object">Extracting counts as a Seurat object<a class="anchor" aria-label="anchor" href="#extracting-counts-as-a-seurat-object"></a>
</h4>
<p>If needed, the H5 <code>SingleCellExperiment</code> can be converted
into a Seurat object. Note that it may take a long time and use a lot of
memory depending on how many cells you are requesting.</p>
<div class="sourceCode" id="cb31"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">single_cell_counts</span> <span class="op">&lt;-</span></span>
<span>    <span class="va">metadata</span> <span class="op">|&gt;</span></span>
<span>    <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/filter.html" class="external-link">filter</a></span><span class="op">(</span></span>
<span>        <span class="va">ethnicity</span> <span class="op">==</span> <span class="st">"African"</span> <span class="op">&amp;</span></span>
<span>        <span class="fu">stringr</span><span class="fu">::</span><span class="fu"><a href="https://stringr.tidyverse.org/reference/str_like.html" class="external-link">str_like</a></span><span class="op">(</span><span class="va">assay</span>, <span class="st">"%10x%"</span><span class="op">)</span> <span class="op">&amp;</span></span>
<span>        <span class="va">tissue</span> <span class="op">==</span> <span class="st">"lung parenchyma"</span> <span class="op">&amp;</span></span>
<span>        <span class="fu">stringr</span><span class="fu">::</span><span class="fu"><a href="https://stringr.tidyverse.org/reference/str_like.html" class="external-link">str_like</a></span><span class="op">(</span><span class="va">cell_type</span>, <span class="st">"%CD4%"</span><span class="op">)</span></span>
<span>    <span class="op">)</span> <span class="op">|&gt;</span></span>
<span>    <span class="fu"><a href="https://rdrr.io/pkg/CuratedAtlasQueryR/man/get_seurat.html" class="external-link">get_seurat</a></span><span class="op">(</span><span class="op">)</span></span>
<span></span>
<span><span class="va">single_cell_counts</span></span></code></pre></div>
</div>
</div>
<div class="section level3">
<h3 id="save-your-singlecellexperiment">Save your <code>SingleCellExperiment</code><a class="anchor" aria-label="anchor" href="#save-your-singlecellexperiment"></a>
</h3>
<div class="section level4">
<h4 id="saving-as-hdf5">Saving as HDF5<a class="anchor" aria-label="anchor" href="#saving-as-hdf5"></a>
</h4>
<p>The recommended way of saving these <code>SingleCellExperiment</code>
objects, if necessary, is to use
<code>saveHDF5SummarizedExperiment</code> from the
<code>HDF5Array</code> package.</p>
<div class="sourceCode" id="cb32"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">single_cell_counts</span> <span class="op">|&gt;</span> <span class="fu">saveHDF5SummarizedExperiment</span><span class="op">(</span><span class="st">"single_cell_counts"</span><span class="op">)</span></span></code></pre></div>
</div>
</div>
</div>
<div class="section level2">
<h2 id="session-info">Session Info<a class="anchor" aria-label="anchor" href="#session-info"></a>
</h2>
<div class="sourceCode" id="cb33"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/sessionInfo.html" class="external-link">sessionInfo</a></span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## R version 4.3.0 (2023-04-21)</span></span>
<span><span class="co">## Platform: x86_64-pc-linux-gnu (64-bit)</span></span>
<span><span class="co">## Running under: Ubuntu 22.04.2 LTS</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## Matrix products: default</span></span>
<span><span class="co">## BLAS:   /usr/lib/x86_64-linux-gnu/openblas-pthread/libblas.so.3 </span></span>
<span><span class="co">## LAPACK: /usr/lib/x86_64-linux-gnu/openblas-pthread/libopenblasp-r0.3.20.so;  LAPACK version 3.10.0</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## locale:</span></span>
<span><span class="co">##  [1] LC_CTYPE=en_US.UTF-8       LC_NUMERIC=C              </span></span>
<span><span class="co">##  [3] LC_TIME=en_US.UTF-8        LC_COLLATE=en_US.UTF-8    </span></span>
<span><span class="co">##  [5] LC_MONETARY=en_US.UTF-8    LC_MESSAGES=en_US.UTF-8   </span></span>
<span><span class="co">##  [7] LC_PAPER=en_US.UTF-8       LC_NAME=C                 </span></span>
<span><span class="co">##  [9] LC_ADDRESS=C               LC_TELEPHONE=C            </span></span>
<span><span class="co">## [11] LC_MEASUREMENT=en_US.UTF-8 LC_IDENTIFICATION=C       </span></span>
<span><span class="co">## </span></span>
<span><span class="co">## time zone: Etc/UTC</span></span>
<span><span class="co">## tzcode source: system (glibc)</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## attached base packages:</span></span>
<span><span class="co">## [1] stats     graphics  grDevices utils     datasets  methods   base     </span></span>
<span><span class="co">## </span></span>
<span><span class="co">## other attached packages:</span></span>
<span><span class="co">## [1] dplyr_1.1.2               CuratedAtlasQueryR_0.99.2</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## loaded via a namespace (and not attached):</span></span>
<span><span class="co">##   [1] RColorBrewer_1.1-3          jsonlite_1.8.5             </span></span>
<span><span class="co">##   [3] magrittr_2.0.3              spatstat.utils_3.0-3       </span></span>
<span><span class="co">##   [5] rmarkdown_2.22              zlibbioc_1.46.0            </span></span>
<span><span class="co">##   [7] fs_1.6.2                    ragg_1.2.5                 </span></span>
<span><span class="co">##   [9] vctrs_0.6.2                 ROCR_1.0-11                </span></span>
<span><span class="co">##  [11] memoise_2.0.1               spatstat.explore_3.2-1     </span></span>
<span><span class="co">##  [13] RCurl_1.98-1.12             htmltools_0.5.5            </span></span>
<span><span class="co">##  [15] S4Arrays_1.0.4              Rhdf5lib_1.22.0            </span></span>
<span><span class="co">##  [17] rhdf5_2.44.0                sass_0.4.6                 </span></span>
<span><span class="co">##  [19] sctransform_0.3.5           parallelly_1.36.0          </span></span>
<span><span class="co">##  [21] KernSmooth_2.23-21          bslib_0.5.0                </span></span>
<span><span class="co">##  [23] htmlwidgets_1.6.2           desc_1.4.2                 </span></span>
<span><span class="co">##  [25] ica_1.0-3                   plyr_1.8.8                 </span></span>
<span><span class="co">##  [27] plotly_4.10.2               zoo_1.8-12                 </span></span>
<span><span class="co">##  [29] cachem_1.0.8                igraph_1.4.3               </span></span>
<span><span class="co">##  [31] mime_0.12                   lifecycle_1.0.3            </span></span>
<span><span class="co">##  [33] pkgconfig_2.0.3             Matrix_1.5-4.1             </span></span>
<span><span class="co">##  [35] R6_2.5.1                    fastmap_1.1.1              </span></span>
<span><span class="co">##  [37] GenomeInfoDbData_1.2.10     MatrixGenerics_1.12.2      </span></span>
<span><span class="co">##  [39] fitdistrplus_1.1-11         future_1.32.0              </span></span>
<span><span class="co">##  [41] shiny_1.7.4                 digest_0.6.31              </span></span>
<span><span class="co">##  [43] colorspace_2.1-0            patchwork_1.1.2            </span></span>
<span><span class="co">##  [45] S4Vectors_0.38.1            rprojroot_2.0.3            </span></span>
<span><span class="co">##  [47] Seurat_4.3.0                tensor_1.5                 </span></span>
<span><span class="co">##  [49] irlba_2.3.5.1               GenomicRanges_1.52.0       </span></span>
<span><span class="co">##  [51] textshaping_0.3.6           progressr_0.13.0           </span></span>
<span><span class="co">##  [53] fansi_1.0.4                 spatstat.sparse_3.0-1      </span></span>
<span><span class="co">##  [55] polyclip_1.10-4             httr_1.4.6                 </span></span>
<span><span class="co">##  [57] abind_1.4-5                 compiler_4.3.0             </span></span>
<span><span class="co">##  [59] withr_2.5.0                 DBI_1.1.3                  </span></span>
<span><span class="co">##  [61] highr_0.10                  HDF5Array_1.28.1           </span></span>
<span><span class="co">##  [63] duckdb_0.8.0                MASS_7.3-60                </span></span>
<span><span class="co">##  [65] DelayedArray_0.26.3         tools_4.3.0                </span></span>
<span><span class="co">##  [67] lmtest_0.9-40               httpuv_1.6.11              </span></span>
<span><span class="co">##  [69] future.apply_1.11.0         goftest_1.2-3              </span></span>
<span><span class="co">##  [71] glue_1.6.2                  nlme_3.1-162               </span></span>
<span><span class="co">##  [73] rhdf5filters_1.12.1         promises_1.2.0.1           </span></span>
<span><span class="co">##  [75] grid_4.3.0                  Rtsne_0.16                 </span></span>
<span><span class="co">##  [77] cluster_2.1.4               reshape2_1.4.4             </span></span>
<span><span class="co">##  [79] generics_0.1.3              gtable_0.3.3               </span></span>
<span><span class="co">##  [81] spatstat.data_3.0-1         tidyr_1.3.0                </span></span>
<span><span class="co">##  [83] data.table_1.14.8           XVector_0.40.0             </span></span>
<span><span class="co">##  [85] sp_1.6-1                    utf8_1.2.3                 </span></span>
<span><span class="co">##  [87] BiocGenerics_0.46.0         spatstat.geom_3.2-1        </span></span>
<span><span class="co">##  [89] RcppAnnoy_0.0.20            ggrepel_0.9.3              </span></span>
<span><span class="co">##  [91] RANN_2.6.1                  pillar_1.9.0               </span></span>
<span><span class="co">##  [93] stringr_1.5.0               later_1.3.1                </span></span>
<span><span class="co">##  [95] splines_4.3.0               lattice_0.21-8             </span></span>
<span><span class="co">##  [97] deldir_1.0-9                survival_3.5-5             </span></span>
<span><span class="co">##  [99] tidyselect_1.2.0            SingleCellExperiment_1.22.0</span></span>
<span><span class="co">## [101] miniUI_0.1.1.1              pbapply_1.7-0              </span></span>
<span><span class="co">## [103] knitr_1.43                  gridExtra_2.3              </span></span>
<span><span class="co">## [105] IRanges_2.34.0              SummarizedExperiment_1.30.2</span></span>
<span><span class="co">## [107] scattermore_1.2             stats4_4.3.0               </span></span>
<span><span class="co">## [109] xfun_0.39                   Biobase_2.60.0             </span></span>
<span><span class="co">## [111] matrixStats_1.0.0           stringi_1.7.12             </span></span>
<span><span class="co">## [113] lazyeval_0.2.2              yaml_2.3.7                 </span></span>
<span><span class="co">## [115] evaluate_0.21               codetools_0.2-19           </span></span>
<span><span class="co">## [117] tibble_3.2.1                cli_3.6.1                  </span></span>
<span><span class="co">## [119] uwot_0.1.14                 xtable_1.8-4               </span></span>
<span><span class="co">## [121] reticulate_1.30             systemfonts_1.0.4          </span></span>
<span><span class="co">## [123] munsell_0.5.0               jquerylib_0.1.4            </span></span>
<span><span class="co">## [125] GenomeInfoDb_1.36.0         Rcpp_1.0.10                </span></span>
<span><span class="co">## [127] spatstat.random_3.1-5       globals_0.16.2             </span></span>
<span><span class="co">## [129] dbplyr_2.3.2                png_0.1-8                  </span></span>
<span><span class="co">## [131] parallel_4.3.0              ellipsis_0.3.2             </span></span>
<span><span class="co">## [133] blob_1.2.4                  assertthat_0.2.1           </span></span>
<span><span class="co">## [135] pkgdown_2.0.7               ggplot2_3.4.2              </span></span>
<span><span class="co">## [137] bitops_1.0-7                listenv_0.9.0              </span></span>
<span><span class="co">## [139] viridisLite_0.4.2           scales_1.2.1               </span></span>
<span><span class="co">## [141] ggridges_0.5.4              SeuratObject_4.1.3         </span></span>
<span><span class="co">## [143] leiden_0.4.3                purrr_1.0.1                </span></span>
<span><span class="co">## [145] crayon_1.5.2                rlang_1.1.1                </span></span>
<span><span class="co">## [147] cowplot_1.1.1</span></span></code></pre>
</div>
<div class="section level2">
<h2 class="unnumbered" id="references">References<a class="anchor" aria-label="anchor" href="#references"></a>
</h2>
<div id="refs" class="references csl-bib-body hanging-indent">
<div id="ref-Monaco2019" class="csl-entry">
Monaco, Gianni, Bernett Lee, Weili Xu, Seri Mustafah, You Yi Hwang,
Christophe Carré, Nicolas Burdin, et al. 2019.
<span>“<span>RNA-Seq</span> Signatures Normalized by <span class="nocase">mRNA</span> Abundance Allow Absolute Deconvolution of
Human Immune Cell Types.”</span> <em>Cell Rep.</em> 26 (6):
1627–1640.e7.
</div>
</div>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">

        <nav id="toc" data-toggle="toc"><h2 data-toc-skip>Contents</h2>
    </nav>
</div>

</div>



      <footer><div class="copyright">
  <p></p>
<p>Developed by Davide Risso, Ludwig Geistlinger.</p>
</div>

<div class="pkgdown">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.7.</p>
</div>

      </footer>
</div>

  


  

  </body>
</html>
