<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>Working with large data • ISMB.OSCA</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link href="https://cdnjs.cloudflare.com/ajax/libs/bootswatch/3.4.0/flatly/bootstrap.min.css" rel="stylesheet" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="Working with large data">
<meta property="og:description" content="ISMB.OSCA">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">
    

    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">ISMB.OSCA</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">0.99.11</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Vignettes
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../articles/intro_bioc_sce.html">Introduction to Bioconductor</a>
    </li>
    <li>
      <a href="../articles/eda_qc.html">Exploratory data analysis</a>
    </li>
    <li>
      <a href="../articles/cell_type_annotation.html">Cell type annotation</a>
    </li>
    <li>
      <a href="../articles/multi_sample.html">Multi-sample analyses</a>
    </li>
    <li>
      <a href="../articles/large_data.html">Working with large data</a>
    </li>
    <li>
      <a href="../articles/hca.html">Accessing data from the Human Cell Atlas</a>
    </li>
  </ul>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://github.com/Bioconductor/ISMB.OSCA" class="external-link">
    <span class="fa fa-github"></span>
     
  </a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1 data-toc-skip>Working with large data</h1>
            
      
      <small class="dont-index">Source: <a href="https://github.com/Bioconductor/ISMB.OSCA/blob/HEAD/vignettes/large_data.Rmd" class="external-link"><code>vignettes/large_data.Rmd</code></a></small>
      <div class="hidden name"><code>large_data.Rmd</code></div>

    </div>

    
    
<div class="section level2">
<h2 id="motivation">Motivation<a class="anchor" aria-label="anchor" href="#motivation"></a>
</h2>
<p>Advances in scRNA-seq technologies have increased the number of cells
that can be assayed in routine experiments. Public databases such as <a href="https://www.ncbi.nlm.nih.gov/geo/" class="external-link">GEO</a> are continually
expanding with more scRNA-seq studies, while large-scale projects such
as the <a href="https://www.humancellatlas.org/" class="external-link">Human Cell Atlas</a>
are expected to generate data for billions of cells. For effective data
analysis, the computational methods need to scale with the increasing
size of scRNA-seq data sets. This section discusses how we can use
various aspects of the Bioconductor ecosystem to tune our analysis
pipelines for greater speed and efficiency.</p>
</div>
<div class="section level2">
<h2 id="out-of-memory-representations">Out of memory representations<a class="anchor" aria-label="anchor" href="#out-of-memory-representations"></a>
</h2>
<p>The count matrix is the central structure around which our analyses
are based. In most of the previous chapters, this has been held fully in
memory as a dense <code>matrix</code> or as a sparse
<code>dgCMatrix</code>. Howevever, in-memory representations may not be
feasible for very large data sets, especially on machines with limited
memory. For example, the 1.3 million brain cell data set from 10X
Genomics (<a href="https://doi.org/10.1038/ncomms14049" class="external-link">Zheng et al.,
2017</a>) would require over 100 GB of RAM to hold as a
<code>matrix</code> and around 30 GB as a <code>dgCMatrix</code>. This
makes it challenging to explore the data on anything less than a HPC
system.</p>
<p>The obvious solution is to use a file-backed matrix representation
where the data are held on disk and subsets are retrieved into memory as
requested. While a number of implementations of file-backed matrices are
available (e.g., <a href="https://cran.r-project.org/web/packages/bigmemory/index.html" class="external-link">bigmemory</a>,
<a href="https://bioconductor.org/packages/matter" class="external-link">matter</a>), we will
be using the implementation from the <a href="https://bioconductor.org/packages/HDF5Array" class="external-link">HDF5Array</a>
package. This uses the popular HDF5 format as the underlying data store,
which provides a measure of standardization and portability across
systems. We demonstrate with a subset of 20,000 cells from the 1.3
million brain cell data set, as provided by the <a href="https://bioconductor.org/packages/TENxBrainData" class="external-link">TENxBrainData</a>
package.</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">TENxBrainData</span><span class="op">)</span></span>
<span><span class="va">sce.brain</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/TENxBrainData/man/TENxBrainData.html" class="external-link">TENxBrainData20k</a></span><span class="op">(</span><span class="op">)</span> </span>
<span><span class="va">sce.brain</span></span></code></pre></div>
<pre><code><span><span class="co">## class: SingleCellExperiment </span></span>
<span><span class="co">## dim: 27998 20000 </span></span>
<span><span class="co">## metadata(0):</span></span>
<span><span class="co">## assays(1): counts</span></span>
<span><span class="co">## rownames: NULL</span></span>
<span><span class="co">## rowData names(2): Ensembl Symbol</span></span>
<span><span class="co">## colnames: NULL</span></span>
<span><span class="co">## colData names(4): Barcode Sequence Library Mouse</span></span>
<span><span class="co">## reducedDimNames(0):</span></span>
<span><span class="co">## mainExpName: NULL</span></span>
<span><span class="co">## altExpNames(0):</span></span></code></pre>
<p>Examination of the <code>SingleCellExperiment</code> object indicates
that the count matrix is a <code>HDF5Matrix</code>. From a comparison of
the memory usage, it is clear that this matrix object is simply a stub
that points to the much larger HDF5 file that actually contains the
data. This avoids the need for large RAM availability during
analyses.</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/dge.html" class="external-link">counts</a></span><span class="op">(</span><span class="va">sce.brain</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## &lt;27998 x 20000&gt; HDF5Matrix object of type "integer":</span></span>
<span><span class="co">##              [,1]     [,2]     [,3]     [,4] ... [,19997] [,19998] [,19999]</span></span>
<span><span class="co">##     [1,]        0        0        0        0   .        0        0        0</span></span>
<span><span class="co">##     [2,]        0        0        0        0   .        0        0        0</span></span>
<span><span class="co">##     [3,]        0        0        0        0   .        0        0        0</span></span>
<span><span class="co">##     [4,]        0        0        0        0   .        0        0        0</span></span>
<span><span class="co">##     [5,]        0        0        0        0   .        0        0        0</span></span>
<span><span class="co">##      ...        .        .        .        .   .        .        .        .</span></span>
<span><span class="co">## [27994,]        0        0        0        0   .        0        0        0</span></span>
<span><span class="co">## [27995,]        0        0        0        1   .        0        2        0</span></span>
<span><span class="co">## [27996,]        0        0        0        0   .        0        1        0</span></span>
<span><span class="co">## [27997,]        0        0        0        0   .        0        0        0</span></span>
<span><span class="co">## [27998,]        0        0        0        0   .        0        0        0</span></span>
<span><span class="co">##          [,20000]</span></span>
<span><span class="co">##     [1,]        0</span></span>
<span><span class="co">##     [2,]        0</span></span>
<span><span class="co">##     [3,]        0</span></span>
<span><span class="co">##     [4,]        0</span></span>
<span><span class="co">##     [5,]        0</span></span>
<span><span class="co">##      ...        .</span></span>
<span><span class="co">## [27994,]        0</span></span>
<span><span class="co">## [27995,]        0</span></span>
<span><span class="co">## [27996,]        0</span></span>
<span><span class="co">## [27997,]        0</span></span>
<span><span class="co">## [27998,]        0</span></span></code></pre>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/object.size.html" class="external-link">object.size</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/dge.html" class="external-link">counts</a></span><span class="op">(</span><span class="va">sce.brain</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## 2496 bytes</span></span></code></pre>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/file.info.html" class="external-link">file.info</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/path.html" class="external-link">path</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/dge.html" class="external-link">counts</a></span><span class="op">(</span><span class="va">sce.brain</span><span class="op">)</span><span class="op">)</span><span class="op">)</span><span class="op">$</span><span class="va">size</span></span></code></pre></div>
<pre><code><span><span class="co">## [1] 76264332</span></span></code></pre>
<p>Manipulation of the count matrix will generally result in the
creation of a <code>DelayedArray</code> object from the <a href="https://bioconductor.org/packages/DelayedArray" class="external-link">DelayedArray</a>
package. This remembers the operations to be applied to the counts and
stores them in the object, to be executed when the modified matrix
values are realized for use in calculations. The use of delayed
operations avoids the need to write the modified values to a new file at
every operation, which would unnecessarily require time-consuming disk
I/O.</p>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">tmp</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/dge.html" class="external-link">counts</a></span><span class="op">(</span><span class="va">sce.brain</span><span class="op">)</span></span>
<span><span class="va">tmp</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/Log.html" class="external-link">log2</a></span><span class="op">(</span><span class="va">tmp</span> <span class="op">+</span> <span class="fl">1</span><span class="op">)</span></span>
<span><span class="va">tmp</span></span></code></pre></div>
<pre><code><span><span class="co">## &lt;27998 x 20000&gt; DelayedMatrix object of type "double":</span></span>
<span><span class="co">##              [,1]     [,2]     [,3] ... [,19999] [,20000]</span></span>
<span><span class="co">##     [1,]        0        0        0   .        0        0</span></span>
<span><span class="co">##     [2,]        0        0        0   .        0        0</span></span>
<span><span class="co">##     [3,]        0        0        0   .        0        0</span></span>
<span><span class="co">##     [4,]        0        0        0   .        0        0</span></span>
<span><span class="co">##     [5,]        0        0        0   .        0        0</span></span>
<span><span class="co">##      ...        .        .        .   .        .        .</span></span>
<span><span class="co">## [27994,]        0        0        0   .        0        0</span></span>
<span><span class="co">## [27995,]        0        0        0   .        0        0</span></span>
<span><span class="co">## [27996,]        0        0        0   .        0        0</span></span>
<span><span class="co">## [27997,]        0        0        0   .        0        0</span></span>
<span><span class="co">## [27998,]        0        0        0   .        0        0</span></span></code></pre>
<p>Many functions described in the previous workflows are capable of
accepting <code>HDF5Matrix</code> objects. This is powered by the
availability of common methods for all matrix representations (e.g.,
subsetting, combining, methods from <a href="https://bioconductor.org/packages/DelayedMatrixStats" class="external-link">DelayedMatrixStats</a>
as well as representation-agnostic C++ code using <a href="https://bioconductor.org/packages/beachmat" class="external-link">beachmat</a>. For
example, we compute QC metrics below with the same
<code><a href="https://rdrr.io/pkg/scater/man/defunct.html" class="external-link">calculateQCMetrics()</a></code> function that we used in the other
workflows.</p>
<div class="sourceCode" id="cb11"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="http://bioconductor.org/packages/scater/" class="external-link">scater</a></span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## Loading required package: scuttle</span></span></code></pre>
<pre><code><span><span class="co">## Loading required package: ggplot2</span></span></code></pre>
<div class="sourceCode" id="cb14"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">is.mito</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/grep.html" class="external-link">grepl</a></span><span class="op">(</span><span class="st">"^mt-"</span>, <span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">rowData</a></span><span class="op">(</span><span class="va">sce.brain</span><span class="op">)</span><span class="op">$</span><span class="va">Symbol</span><span class="op">)</span></span>
<span><span class="va">qcstats</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/scuttle/man/perCellQCMetrics.html" class="external-link">perCellQCMetrics</a></span><span class="op">(</span><span class="va">sce.brain</span>, subsets <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span>Mt <span class="op">=</span> <span class="va">is.mito</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">qcstats</span></span></code></pre></div>
<pre><code><span><span class="co">## DataFrame with 20000 rows and 6 columns</span></span>
<span><span class="co">##             sum  detected subsets_Mt_sum subsets_Mt_detected subsets_Mt_percent</span></span>
<span><span class="co">##       &lt;numeric&gt; &lt;numeric&gt;      &lt;numeric&gt;           &lt;numeric&gt;          &lt;numeric&gt;</span></span>
<span><span class="co">## 1          3060      1546            123                  10            4.01961</span></span>
<span><span class="co">## 2          3500      1694            118                  11            3.37143</span></span>
<span><span class="co">## 3          3092      1613             58                   9            1.87581</span></span>
<span><span class="co">## 4          4420      2050            131                  10            2.96380</span></span>
<span><span class="co">## 5          3771      1813            100                   8            2.65182</span></span>
<span><span class="co">## ...         ...       ...            ...                 ...                ...</span></span>
<span><span class="co">## 19996      4431      2050            127                   9           2.866170</span></span>
<span><span class="co">## 19997      6988      2704             60                   9           0.858615</span></span>
<span><span class="co">## 19998      8749      2988            305                  11           3.486113</span></span>
<span><span class="co">## 19999      3842      1711            129                   8           3.357626</span></span>
<span><span class="co">## 20000      1775       945             26                   6           1.464789</span></span>
<span><span class="co">##           total</span></span>
<span><span class="co">##       &lt;numeric&gt;</span></span>
<span><span class="co">## 1          3060</span></span>
<span><span class="co">## 2          3500</span></span>
<span><span class="co">## 3          3092</span></span>
<span><span class="co">## 4          4420</span></span>
<span><span class="co">## 5          3771</span></span>
<span><span class="co">## ...         ...</span></span>
<span><span class="co">## 19996      4431</span></span>
<span><span class="co">## 19997      6988</span></span>
<span><span class="co">## 19998      8749</span></span>
<span><span class="co">## 19999      3842</span></span>
<span><span class="co">## 20000      1775</span></span></code></pre>
<p>Needless to say, data access from file-backed representations is
slower than that from in-memory representations. The time spent
retrieving data from disk is an unavoidable cost of reducing memory
usage. Whether this is tolerable depends on the application. One example
usage pattern involves performing the heavy computing quickly with
in-memory representations on HPC systems with plentiful memory, and then
distributing file-backed counterparts to individual users for
exploration and visualization on their personal machines.</p>
</div>
<div class="section level2">
<h2 id="parallelization">Parallelization<a class="anchor" aria-label="anchor" href="#parallelization"></a>
</h2>
<p>Parallelization of calculations across genes or cells is an obvious
strategy for speeding up scRNA-seq analysis workflows.</p>
<p>The <em><a href="https://bioconductor.org/packages/3.17/BiocParallel" class="external-link">BiocParallel</a></em>
package provides a common interface for parallel computing throughout
the Bioconductor ecosystem, manifesting as a <code>BPPARAM</code>
argument in compatible functions. We can also use
<code>BiocParallel</code> with more expressive functions directly
through the package’s interface.</p>
<div class="section level4">
<h4 id="basic-use">Basic use<a class="anchor" aria-label="anchor" href="#basic-use"></a>
</h4>
<div class="sourceCode" id="cb16"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/Bioconductor/BiocParallel" class="external-link">BiocParallel</a></span><span class="op">)</span></span></code></pre></div>
<p><code>BiocParallel</code> makes it quite easy to iterate over a
vector and distribute the computation across workers using the
<code>bplapply</code> function. Basic knowledge of <code>lapply</code>
is required.</p>
<p>In this example, we find the square root of a vector of numbers in
parallel by indicating the <code>BPPARAM</code> argument in
<code>bplapply</code>.</p>
<div class="sourceCode" id="cb17"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">param</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/BiocParallel/man/MulticoreParam-class.html" class="external-link">MulticoreParam</a></span><span class="op">(</span>workers <span class="op">=</span> <span class="fl">1</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/BiocParallel/man/bplapply.html" class="external-link">bplapply</a></span><span class="op">(</span></span>
<span>    X <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">4</span>, <span class="fl">9</span>, <span class="fl">16</span>, <span class="fl">25</span><span class="op">)</span>,</span>
<span>    FUN <span class="op">=</span> <span class="kw">function</span><span class="op">(</span><span class="va">x</span><span class="op">)</span> <span class="op">{</span> <span class="fu"><a href="https://rdrr.io/r/base/MathFun.html" class="external-link">sqrt</a></span><span class="op">(</span><span class="va">x</span><span class="op">)</span> <span class="op">}</span>,</span>
<span>    BPPARAM <span class="op">=</span> <span class="va">param</span></span>
<span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## [[1]]</span></span>
<span><span class="co">## [1] 2</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## [[2]]</span></span>
<span><span class="co">## [1] 3</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## [[3]]</span></span>
<span><span class="co">## [1] 4</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## [[4]]</span></span>
<span><span class="co">## [1] 5</span></span></code></pre>
<p><strong>Note</strong>. The number of workers is set to 1 due to
continuous testing resource limitations.</p>
<p>There exists a diverse set of parallelization backends depending on
available hardware and operating systems.</p>
<p>For example, we might use forking across two cores to parallelize the
variance calculations on a Unix system:</p>
<div class="sourceCode" id="cb19"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/MarioniLab/MouseGastrulationData" class="external-link">MouseGastrulationData</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">scran</span><span class="op">)</span></span>
<span><span class="va">sce</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/MouseGastrulationData/man/WTChimeraData.html" class="external-link">WTChimeraData</a></span><span class="op">(</span>samples <span class="op">=</span> <span class="fl">5</span>, type <span class="op">=</span> <span class="st">"processed"</span><span class="op">)</span></span>
<span><span class="va">sce</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/scuttle/man/logNormCounts.html" class="external-link">logNormCounts</a></span><span class="op">(</span><span class="va">sce</span><span class="op">)</span></span>
<span><span class="va">dec.mc</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/scran/man/modelGeneVar.html" class="external-link">modelGeneVar</a></span><span class="op">(</span><span class="va">sce</span>, BPPARAM <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/BiocParallel/man/MulticoreParam-class.html" class="external-link">MulticoreParam</a></span><span class="op">(</span><span class="fl">2</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">dec.mc</span></span></code></pre></div>
<pre><code><span><span class="co">## DataFrame with 29453 rows and 6 columns</span></span>
<span><span class="co">##                           mean       total        tech          bio     p.value</span></span>
<span><span class="co">##                      &lt;numeric&gt;   &lt;numeric&gt;   &lt;numeric&gt;    &lt;numeric&gt;   &lt;numeric&gt;</span></span>
<span><span class="co">## ENSMUSG00000051951 0.002800256 0.003504940 0.002856726  6.48214e-04 1.20918e-01</span></span>
<span><span class="co">## ENSMUSG00000089699 0.000000000 0.000000000 0.000000000  0.00000e+00         NaN</span></span>
<span><span class="co">## ENSMUSG00000102343 0.000000000 0.000000000 0.000000000  0.00000e+00         NaN</span></span>
<span><span class="co">## ENSMUSG00000025900 0.000794995 0.000863633 0.000811027  5.26059e-05 3.68975e-01</span></span>
<span><span class="co">## ENSMUSG00000025902 0.170777718 0.388633677 0.170893240  2.17740e-01 2.48097e-11</span></span>
<span><span class="co">## ...                        ...         ...         ...          ...         ...</span></span>
<span><span class="co">## ENSMUSG00000095041  0.35571083  0.34572194   0.3364126  0.009309342    0.443249</span></span>
<span><span class="co">## ENSMUSG00000063897  0.49007956  0.41924282   0.4407846 -0.021541821    0.599512</span></span>
<span><span class="co">## ENSMUSG00000096730  0.00000000  0.00000000   0.0000000  0.000000000         NaN</span></span>
<span><span class="co">## ENSMUSG00000095742  0.00177158  0.00211619   0.0018073  0.000308881    0.189009</span></span>
<span><span class="co">## tomato-td           0.57257331  0.47487832   0.4971974 -0.022319130    0.591555</span></span>
<span><span class="co">##                            FDR</span></span>
<span><span class="co">##                      &lt;numeric&gt;</span></span>
<span><span class="co">## ENSMUSG00000051951 6.76330e-01</span></span>
<span><span class="co">## ENSMUSG00000089699         NaN</span></span>
<span><span class="co">## ENSMUSG00000102343         NaN</span></span>
<span><span class="co">## ENSMUSG00000025900 7.56203e-01</span></span>
<span><span class="co">## ENSMUSG00000025902 1.35619e-09</span></span>
<span><span class="co">## ...                        ...</span></span>
<span><span class="co">## ENSMUSG00000095041    0.756203</span></span>
<span><span class="co">## ENSMUSG00000063897    0.756203</span></span>
<span><span class="co">## ENSMUSG00000096730         NaN</span></span>
<span><span class="co">## ENSMUSG00000095742    0.756203</span></span>
<span><span class="co">## tomato-td             0.756203</span></span></code></pre>
<p>Another approach would be to distribute jobs across a network of
computers, which yields the same result:</p>
<div class="sourceCode" id="cb21"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">dec.snow</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/scran/man/modelGeneVar.html" class="external-link">modelGeneVar</a></span><span class="op">(</span><span class="va">sce</span>, BPPARAM <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/BiocParallel/man/SnowParam-class.html" class="external-link">SnowParam</a></span><span class="op">(</span><span class="fl">2</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<p>For high-performance computing (HPC) systems with a cluster of
compute nodes, we can distribute jobs via the job scheduler using the
<code>BatchtoolsParam</code> class. The example below assumes a SLURM
cluster, though the settings can be easily configured for a particular
system (see <a href="https://bioconductor.org/packages/3.17/BiocParallel/vignettes/BiocParallel_BatchtoolsParam.pdf" class="external-link">here</a>
for details).</p>
<div class="sourceCode" id="cb22"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># 2 hours, 8 GB, 1 CPU per task, for 10 tasks.</span></span>
<span><span class="va">rs</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span>walltime <span class="op">=</span> <span class="fl">7200</span>, memory <span class="op">=</span> <span class="fl">8000</span>, ncpus <span class="op">=</span> <span class="fl">1</span><span class="op">)</span></span>
<span><span class="va">bpp</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/BiocParallel/man/BatchtoolsParam-class.html" class="external-link">BatchtoolsParam</a></span><span class="op">(</span><span class="fl">10</span>, cluster <span class="op">=</span> <span class="st">"slurm"</span>, resources <span class="op">=</span> <span class="va">rs</span><span class="op">)</span></span></code></pre></div>
<p>Parallelization is best suited for CPU-intensive calculations where
the division of labor results in a concomitant reduction in compute
time. It is not suited for tasks that are bounded by other compute
resources, e.g., memory or file I/O (though the latter is less of an
issue on HPC systems with parallel read/write). In particular, R itself
is inherently single-core, so many of the parallelization backends
involve (i) setting up one or more separate R sessions, (ii) loading the
relevant packages and (iii) transmitting the data to that session.
Depending on the nature and size of the task, this overhead may outweigh
any benefit from parallel computing.</p>
</div>
</div>
<div class="section level2">
<h2 id="fast-approximations">Fast approximations<a class="anchor" aria-label="anchor" href="#fast-approximations"></a>
</h2>
<div class="section level3">
<h3 id="nearest-neighbor-searching">Nearest neighbor searching<a class="anchor" aria-label="anchor" href="#nearest-neighbor-searching"></a>
</h3>
<p>Identification of neighbouring cells in PC or expression space is a
common procedure that is used in many functions, e.g.,
<code><a href="https://rdrr.io/pkg/scran/man/buildSNNGraph.html" class="external-link">buildSNNGraph()</a></code>, <code><a href="https://rdrr.io/pkg/scran/man/defunct.html" class="external-link">doubletCells()</a></code>. The default
is to favour accuracy over speed by using an exact nearest neighbour
(NN) search, implemented with the <span class="math inline">\(k\)</span>-means for <span class="math inline">\(k\)</span>-nearest neighbours algorithm. However,
for large data sets, it may be preferable to use a faster approximate
approach.</p>
<p>The <em><a href="https://bioconductor.org/packages/3.17/BiocNeighbors" class="external-link">BiocNeighbors</a></em>
framework makes it easy to switch between search options by simply
changing the <code>BNPARAM</code> argument in compatible functions. To
demonstrate, we will use the wild-type chimera data for which we had
applied graph-based clustering using the Louvain algorithm for community
detection:</p>
<div class="sourceCode" id="cb23"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">bluster</span><span class="op">)</span></span>
<span><span class="va">sce</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/BiocSingular/man/runPCA.html" class="external-link">runPCA</a></span><span class="op">(</span><span class="va">sce</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/SingleCellExperiment/man/colLabels.html" class="external-link">colLabels</a></span><span class="op">(</span><span class="va">sce</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/scran/man/clusterCells.html" class="external-link">clusterCells</a></span><span class="op">(</span><span class="va">sce</span>, use.dimred <span class="op">=</span> <span class="st">"PCA"</span>,</span>
<span>                               BLUSPARAM <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/bluster/man/NNGraphParam-class.html" class="external-link">NNGraphParam</a></span><span class="op">(</span>cluster.fun <span class="op">=</span> <span class="st">"louvain"</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<p>The above clusters on a nearest neighbor graph generated with an
exact neighbour search. We repeat this below using an approximate
search, implemented using the <a href="https://github.com/spotify/Annoy" class="external-link">Annoy</a> algorithm. This
involves constructing a <code>AnnoyParam</code> object to specify the
search algorithm and then passing it to the parameterization of the
<code><a href="https://rdrr.io/pkg/bluster/man/NNGraphParam-class.html" class="external-link">NNGraphParam()</a></code> function. The results from the exact and
approximate searches are consistent with most clusters from the former
re-appearing in the latter. This suggests that the inaccuracy from the
approximation can be largely ignored.</p>
<div class="sourceCode" id="cb24"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">scran</span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">BiocNeighbors</span><span class="op">)</span></span>
<span><span class="va">clusters</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/scran/man/clusterCells.html" class="external-link">clusterCells</a></span><span class="op">(</span><span class="va">sce</span>, use.dimred <span class="op">=</span> <span class="st">"PCA"</span>,</span>
<span>                         BLUSPARAM <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/bluster/man/NNGraphParam-class.html" class="external-link">NNGraphParam</a></span><span class="op">(</span>cluster.fun <span class="op">=</span> <span class="st">"louvain"</span>,</span>
<span>                                                  BNPARAM <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/BiocNeighbors/man/AnnoyParam.html" class="external-link">AnnoyParam</a></span><span class="op">(</span><span class="op">)</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/table.html" class="external-link">table</a></span><span class="op">(</span>exact <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/SingleCellExperiment/man/colLabels.html" class="external-link">colLabels</a></span><span class="op">(</span><span class="va">sce</span><span class="op">)</span>, approx <span class="op">=</span> <span class="va">clusters</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">##      approx</span></span>
<span><span class="co">## exact   1   2   3   4   5   6   7   8   9  10  11  12  13  14  15</span></span>
<span><span class="co">##    1   91   0   0   0   0   0   0   0   0   0   0   0   0   0   0</span></span>
<span><span class="co">##    2    0 143   0   0   0   0   0   0   0   0   0   0   0   0   1</span></span>
<span><span class="co">##    3    0   0  77   0   0   0   0   0   0   0   0   0   0   0   0</span></span>
<span><span class="co">##    4    0   0   0 341   0   0   0   0   0   0   0   0   0   0  56</span></span>
<span><span class="co">##    5    4   0   0   0 389   0   0   0   0   0   0   0   0   0   0</span></span>
<span><span class="co">##    6    0   0   0   0   0 205   5   0   0   0   0   0   0   0   0</span></span>
<span><span class="co">##    7    0   0   0   0   0   0 246   0   0   0   0   0   0   0   0</span></span>
<span><span class="co">##    8    0   0   0   0   1   0   0  95   0   0   0   0   0   0   0</span></span>
<span><span class="co">##    9    0   0   0   0   0   0   0   0 108   0   0   0   0   0   0</span></span>
<span><span class="co">##    10   0   0   0   0   0   0   0   0   0 113   0   0   0   0   0</span></span>
<span><span class="co">##    11   0   0   0   0   0   0   0   0   0   0 153   0   0   0   0</span></span>
<span><span class="co">##    12   0   0   0   0   5   0   0   0   0   1   0 211   0   0   0</span></span>
<span><span class="co">##    13   0   0   0   0   0   0   1   0   0   0   0   0 144   0   1</span></span>
<span><span class="co">##    14   0   0   0   0   0   0   0   0   0   0   0   0   0  20   0</span></span></code></pre>
<p>Note that Annoy writes the NN index to disk prior to performing the
search. Thus, it may not actually be faster than the default exact
algorithm for small datasets, depending on whether the overhead of disk
write is offset by the computational complexity of the search. It is
also not difficult to find situations where the approximation
deteriorates, especially at high dimensions, though this may not have an
appreciable impact on the biological conclusions.</p>
<div class="sourceCode" id="cb26"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/Random.html" class="external-link">set.seed</a></span><span class="op">(</span><span class="fl">1000</span><span class="op">)</span></span>
<span><span class="va">y1</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/matrix.html" class="external-link">matrix</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/stats/Normal.html" class="external-link">rnorm</a></span><span class="op">(</span><span class="fl">50000</span><span class="op">)</span>, nrow <span class="op">=</span> <span class="fl">1000</span><span class="op">)</span></span>
<span><span class="va">y2</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/matrix.html" class="external-link">matrix</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/stats/Normal.html" class="external-link">rnorm</a></span><span class="op">(</span><span class="fl">50000</span><span class="op">)</span>, nrow <span class="op">=</span> <span class="fl">1000</span><span class="op">)</span></span>
<span><span class="va">Y</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/cbind.html" class="external-link">rbind</a></span><span class="op">(</span><span class="va">y1</span>, <span class="va">y2</span><span class="op">)</span></span>
<span><span class="va">exact</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/BiocNeighbors/man/findKNN-methods.html" class="external-link">findKNN</a></span><span class="op">(</span><span class="va">Y</span>, k <span class="op">=</span> <span class="fl">20</span><span class="op">)</span></span>
<span><span class="va">approx</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/BiocNeighbors/man/findKNN-methods.html" class="external-link">findKNN</a></span><span class="op">(</span><span class="va">Y</span>, k <span class="op">=</span> <span class="fl">20</span>, BNPARAM <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/BiocNeighbors/man/AnnoyParam.html" class="external-link">AnnoyParam</a></span><span class="op">(</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/mean.html" class="external-link">mean</a></span><span class="op">(</span><span class="va">exact</span><span class="op">$</span><span class="va">index</span> <span class="op">!=</span> <span class="va">approx</span><span class="op">$</span><span class="va">index</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## [1] 0.561925</span></span></code></pre>
</div>
<div class="section level3">
<h3 id="singular-value-decomposition">Singular value decomposition<a class="anchor" aria-label="anchor" href="#singular-value-decomposition"></a>
</h3>
<p>The singular value decomposition (SVD) underlies the PCA used
throughout our analyses, e.g., in <code><a href="https://rdrr.io/pkg/scran/man/denoisePCA.html" class="external-link">denoisePCA()</a></code>,
<code>fastMNN()</code>, <code><a href="https://rdrr.io/pkg/scran/man/defunct.html" class="external-link">doubletCells()</a></code>. (Briefly, the right
singular vectors are the eigenvectors of the gene-gene covariance
matrix, where each eigenvector represents the axis of maximum remaining
variation in the PCA.) The default <code><a href="https://rdrr.io/r/base/svd.html" class="external-link">base::svd()</a></code> function
performs an exact SVD that is not performant for large datasets.
Instead, we use fast approximate methods from the <em><a href="https://CRAN.R-project.org/package=irlba" class="external-link">irlba</a></em> and
<em><a href="https://CRAN.R-project.org/package=rsvd" class="external-link">rsvd</a></em>
packages, conveniently wrapped into the <em><a href="https://bioconductor.org/packages/3.17/BiocSingular" class="external-link">BiocSingular</a></em>
package for ease of use and package development. Specifically, we can
change the SVD algorithm used in any of these functions by simply
specifying an alternative value for the <code>BSPARAM</code>
argument.</p>
<div class="sourceCode" id="cb28"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="http://bioconductor.org/packages/scater/" class="external-link">scater</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/LTLA/BiocSingular" class="external-link">BiocSingular</a></span><span class="op">)</span></span>
<span></span>
<span><span class="co"># As the name suggests, it is random, so we need to set the seed.</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/Random.html" class="external-link">set.seed</a></span><span class="op">(</span><span class="fl">101000</span><span class="op">)</span></span>
<span><span class="va">r.out</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/BiocSingular/man/runPCA.html" class="external-link">runPCA</a></span><span class="op">(</span><span class="va">sce</span>, ncomponents <span class="op">=</span> <span class="fl">20</span>, BSPARAM <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/BiocSingular/man/BiocSingularParam.html" class="external-link">RandomParam</a></span><span class="op">(</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/str.html" class="external-link">str</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/SingleCellExperiment/man/reducedDims.html" class="external-link">reducedDim</a></span><span class="op">(</span><span class="va">r.out</span>, <span class="st">"PCA"</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">##  num [1:2411, 1:20] 14.79 5.79 13.07 -32.19 -26.45 ...</span></span>
<span><span class="co">##  - attr(*, "dimnames")=List of 2</span></span>
<span><span class="co">##   ..$ : chr [1:2411] "cell_9769" "cell_9770" "cell_9771" "cell_9772" ...</span></span>
<span><span class="co">##   ..$ : chr [1:20] "PC1" "PC2" "PC3" "PC4" ...</span></span>
<span><span class="co">##  - attr(*, "varExplained")= num [1:20] 192.6 87 29.4 23.1 21.6 ...</span></span>
<span><span class="co">##  - attr(*, "percentVar")= num [1:20] 25.84 11.67 3.94 3.1 2.89 ...</span></span>
<span><span class="co">##  - attr(*, "rotation")= num [1:500, 1:20] -0.174 -0.173 -0.157 0.105 -0.132 ...</span></span>
<span><span class="co">##   ..- attr(*, "dimnames")=List of 2</span></span>
<span><span class="co">##   .. ..$ : chr [1:500] "ENSMUSG00000055609" "ENSMUSG00000052217" "ENSMUSG00000069919" "ENSMUSG00000048583" ...</span></span>
<span><span class="co">##   .. ..$ : chr [1:20] "PC1" "PC2" "PC3" "PC4" ...</span></span></code></pre>
<div class="sourceCode" id="cb30"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/Random.html" class="external-link">set.seed</a></span><span class="op">(</span><span class="fl">101001</span><span class="op">)</span></span>
<span><span class="va">i.out</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/BiocSingular/man/runPCA.html" class="external-link">runPCA</a></span><span class="op">(</span><span class="va">sce</span>, ncomponents <span class="op">=</span> <span class="fl">20</span>, BSPARAM <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/BiocSingular/man/BiocSingularParam.html" class="external-link">IrlbaParam</a></span><span class="op">(</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/str.html" class="external-link">str</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/SingleCellExperiment/man/reducedDims.html" class="external-link">reducedDim</a></span><span class="op">(</span><span class="va">i.out</span>, <span class="st">"PCA"</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">##  num [1:2411, 1:20] -14.79 -5.79 -13.07 32.19 26.45 ...</span></span>
<span><span class="co">##  - attr(*, "dimnames")=List of 2</span></span>
<span><span class="co">##   ..$ : chr [1:2411] "cell_9769" "cell_9770" "cell_9771" "cell_9772" ...</span></span>
<span><span class="co">##   ..$ : chr [1:20] "PC1" "PC2" "PC3" "PC4" ...</span></span>
<span><span class="co">##  - attr(*, "varExplained")= num [1:20] 192.6 87 29.4 23.1 21.6 ...</span></span>
<span><span class="co">##  - attr(*, "percentVar")= num [1:20] 25.84 11.67 3.94 3.1 2.89 ...</span></span>
<span><span class="co">##  - attr(*, "rotation")= num [1:500, 1:20] 0.174 0.173 0.157 -0.105 0.132 ...</span></span>
<span><span class="co">##   ..- attr(*, "dimnames")=List of 2</span></span>
<span><span class="co">##   .. ..$ : chr [1:500] "ENSMUSG00000055609" "ENSMUSG00000052217" "ENSMUSG00000069919" "ENSMUSG00000048583" ...</span></span>
<span><span class="co">##   .. ..$ : chr [1:20] "PC1" "PC2" "PC3" "PC4" ...</span></span></code></pre>
<p>Both IRLBA and randomized SVD (RSVD) are much faster than the exact
SVD with negligible loss of accuracy. This motivates their default use
in many <em><a href="https://bioconductor.org/packages/3.17/scran" class="external-link">scran</a></em> and
<em><a href="https://bioconductor.org/packages/3.17/scater" class="external-link">scater</a></em>
functions, at the cost of requiring users to set the seed to guarantee
reproducibility. IRLBA can occasionally fail to converge and require
more iterations (passed via <code>maxit=</code> in
<code><a href="https://rdrr.io/pkg/BiocSingular/man/BiocSingularParam.html" class="external-link">IrlbaParam()</a></code>), while RSVD involves an explicit trade-off
between accuracy and speed based on its oversampling parameter
(<code>p=</code>) and number of power iterations (<code>q=</code>). We
tend to prefer IRLBA as its default behavior is more accurate, though
RSVD is much faster for file-backed matrices.</p>
</div>
</div>
<div class="section level2">
<h2 id="interoperability-with-popular-single-cell-analysis-ecosytems">Interoperability with popular single-cell analysis ecosytems<a class="anchor" aria-label="anchor" href="#interoperability-with-popular-single-cell-analysis-ecosytems"></a>
</h2>
<div class="section level3">
<h3 id="seurat">Seurat<a class="anchor" aria-label="anchor" href="#seurat"></a>
</h3>
<p><a href="https://satijalab.org/seurat" class="external-link">Seurat</a> is an R package
designed for QC, analysis, and exploration of single-cell RNA-seq data.
Seurat can be used to identify and interpret sources of heterogeneity
from single-cell transcriptomic measurements, and to integrate diverse
types of single-cell data. Seurat is developed and maintained by the <a href="https://satijalab.org/seurat/authors.html" class="external-link">Satija lab</a> and is
released under the <a href="https://opensource.org/license/mit/" class="external-link">MIT
license</a>.</p>
<div class="sourceCode" id="cb32"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://satijalab.org/seurat" class="external-link">Seurat</a></span><span class="op">)</span></span></code></pre></div>
<p>Although the basic processing of single-cell data with Bioconductor
packages (described in the <a href="https://bioconductor.org/books/release/OSCA/" class="external-link">OSCA book</a>) and
with Seurat is very similar and will produce overall roughly identical
results, there is also complementary functionality with regard to cell
type annotation, dataset integration, and downstream analysis. To make
the most of both ecosystems it is therefore beneficial to be able to
easily switch between a <code>SeuratObject</code> and a
<code>SingleCellExperiment</code>. See also the Seurat <a href="https://satijalab.org/seurat/articles/conversion_vignette.html" class="external-link">conversion
vignette</a> for conversion to/from other popular single cell formats
such as the AnnData format used by <a href="https://scanpy.readthedocs.io/en/stable/" class="external-link">scanpy</a>.</p>
<p>Here, we demonstrate converting the Seurat object produced in
Seurat’s <a href="https://satijalab.org/seurat/articles/pbmc3k_tutorial.html" class="external-link">PBMC
tutorial</a> to a <code>SingleCellExperiment</code> for further analysis
with functionality from OSCA/Bioconductor. We therefore need to first
install the <a href="https://github.com/satijalab/seurat-data" class="external-link">SeuratData</a> package,
which is available from GitHub only.</p>
<div class="sourceCode" id="cb33"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">BiocManager</span><span class="fu">::</span><span class="fu"><a href="https://bioconductor.github.io/BiocManager/reference/install.html" class="external-link">install</a></span><span class="op">(</span><span class="st">"satijalab/seurat-data"</span><span class="op">)</span></span></code></pre></div>
<p>We then proceed by loading all required packages and installing the
PBMC dataset:</p>
<div class="sourceCode" id="cb34"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="http://www.satijalab.org/seurat" class="external-link">SeuratData</a></span><span class="op">)</span></span>
<span><span class="fu"><a href="https://satijalab.org/seurat/reference/InstallData.html" class="external-link">InstallData</a></span><span class="op">(</span><span class="st">"pbmc3k"</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## Installing package into '/tmp/RtmpUXXVga/temp_libpath1fa48da3c7d'</span></span>
<span><span class="co">## (as 'lib' is unspecified)</span></span></code></pre>
<p>We then load the dataset as an <code>SeuratObject</code> and convert
it to a <code>SingleCellExperiment</code>.</p>
<div class="sourceCode" id="cb36"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Use PBMC3K from SeuratData</span></span>
<span><span class="va">pbmc</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://satijalab.org/seurat/reference/LoadData.html" class="external-link">LoadData</a></span><span class="op">(</span>ds <span class="op">=</span> <span class="st">"pbmc3k"</span>, type <span class="op">=</span> <span class="st">"pbmc3k.final"</span><span class="op">)</span></span>
<span><span class="va">pbmc</span></span></code></pre></div>
<pre><code><span><span class="co">## An object of class Seurat </span></span>
<span><span class="co">## 13714 features across 2638 samples within 1 assay </span></span>
<span><span class="co">## Active assay: RNA (13714 features, 2000 variable features)</span></span>
<span><span class="co">##  2 dimensional reductions calculated: pca, umap</span></span></code></pre>
<div class="sourceCode" id="cb38"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">pbmc.sce</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://satijalab.org/seurat/reference/as.SingleCellExperiment.html" class="external-link">as.SingleCellExperiment</a></span><span class="op">(</span><span class="va">pbmc</span><span class="op">)</span></span>
<span><span class="va">pbmc.sce</span></span></code></pre></div>
<pre><code><span><span class="co">## class: SingleCellExperiment </span></span>
<span><span class="co">## dim: 13714 2638 </span></span>
<span><span class="co">## metadata(0):</span></span>
<span><span class="co">## assays(3): counts logcounts scaledata</span></span>
<span><span class="co">## rownames(13714): AL627309.1 AP006222.2 ... PNRC2.1 SRSF10.1</span></span>
<span><span class="co">## rowData names(0):</span></span>
<span><span class="co">## colnames(2638): AAACATACAACCAC AAACATTGAGCTAC ... TTTGCATGAGAGGC</span></span>
<span><span class="co">##   TTTGCATGCCTCAC</span></span>
<span><span class="co">## colData names(8): orig.ident nCount_RNA ... seurat_clusters ident</span></span>
<span><span class="co">## reducedDimNames(2): PCA UMAP</span></span>
<span><span class="co">## mainExpName: RNA</span></span>
<span><span class="co">## altExpNames(0):</span></span></code></pre>
<p>Seurat also allows conversion from <code>SingleCellExperiment</code>
objects to Seurat objects; we demonstrate this here on the wild-type
chimera mouse gastrulation dataset.</p>
<div class="sourceCode" id="cb40"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">sce</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/MouseGastrulationData/man/WTChimeraData.html" class="external-link">WTChimeraData</a></span><span class="op">(</span>samples <span class="op">=</span> <span class="fl">5</span>, type <span class="op">=</span> <span class="st">"processed"</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">assay</a></span><span class="op">(</span><span class="va">sce</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/matrix.html" class="external-link">as.matrix</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">assay</a></span><span class="op">(</span><span class="va">sce</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">sce</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/scuttle/man/logNormCounts.html" class="external-link">logNormCounts</a></span><span class="op">(</span><span class="va">sce</span><span class="op">)</span></span>
<span><span class="va">sce</span></span></code></pre></div>
<pre><code><span><span class="co">## class: SingleCellExperiment </span></span>
<span><span class="co">## dim: 29453 2411 </span></span>
<span><span class="co">## metadata(0):</span></span>
<span><span class="co">## assays(2): counts logcounts</span></span>
<span><span class="co">## rownames(29453): ENSMUSG00000051951 ENSMUSG00000089699 ...</span></span>
<span><span class="co">##   ENSMUSG00000095742 tomato-td</span></span>
<span><span class="co">## rowData names(2): ENSEMBL SYMBOL</span></span>
<span><span class="co">## colnames(2411): cell_9769 cell_9770 ... cell_12178 cell_12179</span></span>
<span><span class="co">## colData names(11): cell barcode ... doub.density sizeFactor</span></span>
<span><span class="co">## reducedDimNames(2): pca.corrected.E7.5 pca.corrected.E8.5</span></span>
<span><span class="co">## mainExpName: NULL</span></span>
<span><span class="co">## altExpNames(0):</span></span></code></pre>
<p>After some processing of the dataset, the actual conversion is
carried out with the <code>as.Seurat</code> function.</p>
<div class="sourceCode" id="cb42"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">sobj</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://mojaveazure.github.io/seurat-object/reference/as.Seurat.html" class="external-link">as.Seurat</a></span><span class="op">(</span><span class="va">sce</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://mojaveazure.github.io/seurat-object/reference/Idents.html" class="external-link">Idents</a></span><span class="op">(</span><span class="va">sobj</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="st">"celltype.mapped"</span></span>
<span><span class="va">sobj</span></span></code></pre></div>
<pre><code><span><span class="co">## An object of class Seurat </span></span>
<span><span class="co">## 29453 features across 2411 samples within 1 assay </span></span>
<span><span class="co">## Active assay: originalexp (29453 features, 0 variable features)</span></span>
<span><span class="co">##  2 dimensional reductions calculated: pca.corrected.E7.5, pca.corrected.E8.5</span></span></code></pre>
</div>
<div class="section level3">
<h3 id="scanpy">Scanpy<a class="anchor" aria-label="anchor" href="#scanpy"></a>
</h3>
<p><a href="https://scanpy.readthedocs.io" class="external-link">Scanpy</a> is a scalable
toolkit for analyzing single-cell gene expression data built jointly
with <a href="https://anndata.readthedocs.io/" class="external-link">anndata</a>. It includes
preprocessing, visualization, clustering, trajectory inference and
differential expression testing. The Python-based implementation
efficiently deals with datasets of more than one million cells. Scanpy
is developed and maintained by the <a href="">Theis lab</a> and is
released under a <a href="https://github.com/scverse/scanpy/blob/master/LICENSE" class="external-link">BSD-3-Clause
license</a>. Scanpy is part of the <a href="https://scverse.org/" class="external-link">scverse</a>, a Python-based ecosystem for
single-cell omics data analysis.</p>
<p>At the core of scanpy’s single-cell functionality is the
<code>anndata</code> data structure, scanpy’s integrated single-cell
data container, which is conceptually very similar to Bioconductor’s
<code>SingleCellExperiment</code> class.</p>
<p>Bioconductor’s <em><a href="https://bioconductor.org/packages/3.17/zellkonverter" class="external-link">zellkonverter</a></em>
package provides a lightweight interface between the Bioconductor
<code>SingleCellExperiment</code> data structure and the Python
<code>AnnData</code>-based single-cell analysis environment. The idea is
to enable users and developers to easily move data between these
frameworks to construct a multi-language analysis pipeline across
R/Bioconductor and Python.</p>
<div class="sourceCode" id="cb44"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/theislab/zellkonverter" class="external-link">zellkonverter</a></span><span class="op">)</span></span></code></pre></div>
<p>The <code><a href="https://rdrr.io/pkg/zellkonverter/man/readH5AD.html" class="external-link">readH5AD()</a></code> function can be used to read a
<code>SingleCellExperiment</code> from an H5AD file. Here, we use an
example H5AD file contained in the <em><a href="https://bioconductor.org/packages/3.17/zellkonverter" class="external-link">zellkonverter</a></em>
package.</p>
<div class="sourceCode" id="cb45"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">example_h5ad</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">"extdata"</span>, <span class="st">"krumsiek11.h5ad"</span>,</span>
<span>                            package <span class="op">=</span> <span class="st">"zellkonverter"</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/zellkonverter/man/readH5AD.html" class="external-link">readH5AD</a></span><span class="op">(</span><span class="va">example_h5ad</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## Warning: The names of these selected <span style="color: #00BB00;">uns$highlights</span> items have been modified to match R</span></span>
<span><span class="co">## conventions: <span style="color: #00BB00;">'0' -&gt; 'X0'</span>, <span style="color: #00BB00;">'159' -&gt; 'X159'</span>, <span style="color: #00BB00;">'319' -&gt; 'X319'</span>, <span style="color: #00BB00;">'459' -&gt; 'X459'</span>,</span></span>
<span><span class="co">## and <span style="color: #00BB00;">'619' -&gt; 'X619'</span></span></span></code></pre>
<pre><code><span><span class="co">## class: SingleCellExperiment </span></span>
<span><span class="co">## dim: 11 640 </span></span>
<span><span class="co">## metadata(2): highlights iroot</span></span>
<span><span class="co">## assays(1): X</span></span>
<span><span class="co">## rownames(11): Gata2 Gata1 ... EgrNab Gfi1</span></span>
<span><span class="co">## rowData names(0):</span></span>
<span><span class="co">## colnames(640): 0 1 ... 158-3 159-3</span></span>
<span><span class="co">## colData names(1): cell_type</span></span>
<span><span class="co">## reducedDimNames(0):</span></span>
<span><span class="co">## mainExpName: NULL</span></span>
<span><span class="co">## altExpNames(0):</span></span></code></pre>
<p>We can also write a <code>SingleCellExperiment</code> to an H5AD file
with the <code><a href="https://rdrr.io/pkg/zellkonverter/man/writeH5AD.html" class="external-link">writeH5AD()</a></code> function. This is demonstrated below
on the wild-type chimera mouse gastrulation dataset.</p>
<div class="sourceCode" id="cb48"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">out.file</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/tempfile.html" class="external-link">tempfile</a></span><span class="op">(</span>pattern <span class="op">=</span> <span class="st">".h5ad"</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/zellkonverter/man/writeH5AD.html" class="external-link">writeH5AD</a></span><span class="op">(</span><span class="va">sce</span>, file <span class="op">=</span> <span class="va">out.file</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## <span style="color: #00BBBB;">ℹ</span> Using the <span style="color: #00BB00;">'counts'</span> assay as the <span style="color: #00BB00;">X matrix</span></span></span></code></pre>
<p>The resulting H5AD file can then be read into Python using scanpy’s
<a href="https://scanpy.readthedocs.io/en/stable/generated/scanpy.read_h5ad.html" class="external-link">read_h5ad</a>
function and then directly used in compatible Python-based analysis
frameworks.</p>
</div>
</div>
<div class="section level2">
<h2 id="session-info">Session Info<a class="anchor" aria-label="anchor" href="#session-info"></a>
</h2>
<div class="sourceCode" id="cb50"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/sessionInfo.html" class="external-link">sessionInfo</a></span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## R version 4.3.0 (2023-04-21)</span></span>
<span><span class="co">## Platform: x86_64-pc-linux-gnu (64-bit)</span></span>
<span><span class="co">## Running under: Ubuntu 22.04.2 LTS</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## Matrix products: default</span></span>
<span><span class="co">## BLAS:   /usr/lib/x86_64-linux-gnu/openblas-pthread/libblas.so.3 </span></span>
<span><span class="co">## LAPACK: /usr/lib/x86_64-linux-gnu/openblas-pthread/libopenblasp-r0.3.20.so;  LAPACK version 3.10.0</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## locale:</span></span>
<span><span class="co">##  [1] LC_CTYPE=en_US.UTF-8       LC_NUMERIC=C              </span></span>
<span><span class="co">##  [3] LC_TIME=en_US.UTF-8        LC_COLLATE=en_US.UTF-8    </span></span>
<span><span class="co">##  [5] LC_MONETARY=en_US.UTF-8    LC_MESSAGES=en_US.UTF-8   </span></span>
<span><span class="co">##  [7] LC_PAPER=en_US.UTF-8       LC_NAME=C                 </span></span>
<span><span class="co">##  [9] LC_ADDRESS=C               LC_TELEPHONE=C            </span></span>
<span><span class="co">## [11] LC_MEASUREMENT=en_US.UTF-8 LC_IDENTIFICATION=C       </span></span>
<span><span class="co">## </span></span>
<span><span class="co">## time zone: Etc/UTC</span></span>
<span><span class="co">## tzcode source: system (glibc)</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## attached base packages:</span></span>
<span><span class="co">## [1] stats4    stats     graphics  grDevices utils     datasets  methods  </span></span>
<span><span class="co">## [8] base     </span></span>
<span><span class="co">## </span></span>
<span><span class="co">## other attached packages:</span></span>
<span><span class="co">##  [1] zellkonverter_1.10.1         pbmc3k.SeuratData_3.1.4     </span></span>
<span><span class="co">##  [3] SeuratData_0.2.2             SeuratObject_4.1.3          </span></span>
<span><span class="co">##  [5] Seurat_4.3.0.1               BiocSingular_1.16.0         </span></span>
<span><span class="co">##  [7] BiocNeighbors_1.18.0         bluster_1.10.0              </span></span>
<span><span class="co">##  [9] scran_1.28.1                 MouseGastrulationData_1.14.0</span></span>
<span><span class="co">## [11] SpatialExperiment_1.10.0     BiocParallel_1.34.2         </span></span>
<span><span class="co">## [13] scater_1.28.0                ggplot2_3.4.2               </span></span>
<span><span class="co">## [15] scuttle_1.10.1               TENxBrainData_1.20.0        </span></span>
<span><span class="co">## [17] HDF5Array_1.28.1             rhdf5_2.44.0                </span></span>
<span><span class="co">## [19] DelayedArray_0.26.6          S4Arrays_1.0.4              </span></span>
<span><span class="co">## [21] Matrix_1.5-4.1               SingleCellExperiment_1.22.0 </span></span>
<span><span class="co">## [23] SummarizedExperiment_1.30.2  Biobase_2.60.0              </span></span>
<span><span class="co">## [25] GenomicRanges_1.52.0         GenomeInfoDb_1.36.1         </span></span>
<span><span class="co">## [27] IRanges_2.34.1               S4Vectors_0.38.1            </span></span>
<span><span class="co">## [29] BiocGenerics_0.46.0          MatrixGenerics_1.12.2       </span></span>
<span><span class="co">## [31] matrixStats_1.0.0            BiocStyle_2.28.0            </span></span>
<span><span class="co">## </span></span>
<span><span class="co">## loaded via a namespace (and not attached):</span></span>
<span><span class="co">##   [1] fs_1.6.2                      spatstat.sparse_3.0-2        </span></span>
<span><span class="co">##   [3] bitops_1.0-7                  httr_1.4.6                   </span></span>
<span><span class="co">##   [5] RColorBrewer_1.1-3            tools_4.3.0                  </span></span>
<span><span class="co">##   [7] sctransform_0.3.5             utf8_1.2.3                   </span></span>
<span><span class="co">##   [9] R6_2.5.1                      lazyeval_0.2.2               </span></span>
<span><span class="co">##  [11] uwot_0.1.16                   rhdf5filters_1.12.1          </span></span>
<span><span class="co">##  [13] withr_2.5.0                   sp_2.0-0                     </span></span>
<span><span class="co">##  [15] gridExtra_2.3                 progressr_0.13.0             </span></span>
<span><span class="co">##  [17] cli_3.6.1                     textshaping_0.3.6            </span></span>
<span><span class="co">##  [19] spatstat.explore_3.2-1        sass_0.4.6                   </span></span>
<span><span class="co">##  [21] spatstat.data_3.0-1           ggridges_0.5.4               </span></span>
<span><span class="co">##  [23] pbapply_1.7-2                 pkgdown_2.0.7                </span></span>
<span><span class="co">##  [25] systemfonts_1.0.4             R.utils_2.12.2               </span></span>
<span><span class="co">##  [27] parallelly_1.36.0             limma_3.56.2                 </span></span>
<span><span class="co">##  [29] RSQLite_2.3.1                 generics_0.1.3               </span></span>
<span><span class="co">##  [31] ica_1.0-3                     spatstat.random_3.1-5        </span></span>
<span><span class="co">##  [33] dplyr_1.1.2                   ggbeeswarm_0.7.2             </span></span>
<span><span class="co">##  [35] fansi_1.0.4                   abind_1.4-5                  </span></span>
<span><span class="co">##  [37] R.methodsS3_1.8.2             lifecycle_1.0.3              </span></span>
<span><span class="co">##  [39] yaml_2.3.7                    edgeR_3.42.4                 </span></span>
<span><span class="co">##  [41] BiocFileCache_2.8.0           Rtsne_0.16                   </span></span>
<span><span class="co">##  [43] grid_4.3.0                    blob_1.2.4                   </span></span>
<span><span class="co">##  [45] promises_1.2.0.1              dqrng_0.3.0                  </span></span>
<span><span class="co">##  [47] ExperimentHub_2.8.0           crayon_1.5.2                 </span></span>
<span><span class="co">##  [49] dir.expiry_1.8.0              miniUI_0.1.1.1               </span></span>
<span><span class="co">##  [51] lattice_0.21-8                beachmat_2.16.0              </span></span>
<span><span class="co">##  [53] cowplot_1.1.1                 KEGGREST_1.40.0              </span></span>
<span><span class="co">##  [55] magick_2.7.4                  pillar_1.9.0                 </span></span>
<span><span class="co">##  [57] knitr_1.43                    metapod_1.8.0                </span></span>
<span><span class="co">##  [59] rjson_0.2.21                  future.apply_1.11.0          </span></span>
<span><span class="co">##  [61] codetools_0.2-19              leiden_0.4.3                 </span></span>
<span><span class="co">##  [63] glue_1.6.2                    data.table_1.14.8            </span></span>
<span><span class="co">##  [65] vctrs_0.6.3                   png_0.1-8                    </span></span>
<span><span class="co">##  [67] gtable_0.3.3                  cachem_1.0.8                 </span></span>
<span><span class="co">##  [69] xfun_0.39                     mime_0.12                    </span></span>
<span><span class="co">##  [71] DropletUtils_1.20.0           survival_3.5-5               </span></span>
<span><span class="co">##  [73] statmod_1.5.0                 interactiveDisplayBase_1.38.0</span></span>
<span><span class="co">##  [75] ellipsis_0.3.2                fitdistrplus_1.1-11          </span></span>
<span><span class="co">##  [77] ROCR_1.0-11                   nlme_3.1-162                 </span></span>
<span><span class="co">##  [79] bit64_4.0.5                   filelock_1.0.2               </span></span>
<span><span class="co">##  [81] RcppAnnoy_0.0.21              BumpyMatrix_1.8.0            </span></span>
<span><span class="co">##  [83] rprojroot_2.0.3               bslib_0.5.0                  </span></span>
<span><span class="co">##  [85] irlba_2.3.5.1                 vipor_0.4.5                  </span></span>
<span><span class="co">##  [87] KernSmooth_2.23-21            colorspace_2.1-0             </span></span>
<span><span class="co">##  [89] DBI_1.1.3                     tidyselect_1.2.0             </span></span>
<span><span class="co">##  [91] bit_4.0.5                     compiler_4.3.0               </span></span>
<span><span class="co">##  [93] curl_5.0.1                    basilisk.utils_1.12.1        </span></span>
<span><span class="co">##  [95] desc_1.4.2                    plotly_4.10.2                </span></span>
<span><span class="co">##  [97] scales_1.2.1                  lmtest_0.9-40                </span></span>
<span><span class="co">##  [99] rappdirs_0.3.3                stringr_1.5.0                </span></span>
<span><span class="co">## [101] digest_0.6.32                 goftest_1.2-3                </span></span>
<span><span class="co">## [103] spatstat.utils_3.0-3          rmarkdown_2.23               </span></span>
<span><span class="co">## [105] basilisk_1.12.1               XVector_0.40.0               </span></span>
<span><span class="co">## [107] htmltools_0.5.5               pkgconfig_2.0.3              </span></span>
<span><span class="co">## [109] sparseMatrixStats_1.12.2      dbplyr_2.3.2                 </span></span>
<span><span class="co">## [111] fastmap_1.1.1                 rlang_1.1.1                  </span></span>
<span><span class="co">## [113] htmlwidgets_1.6.2             shiny_1.7.4                  </span></span>
<span><span class="co">## [115] DelayedMatrixStats_1.22.1     jquerylib_0.1.4              </span></span>
<span><span class="co">## [117] zoo_1.8-12                    jsonlite_1.8.7               </span></span>
<span><span class="co">## [119] R.oo_1.25.0                   RCurl_1.98-1.12              </span></span>
<span><span class="co">## [121] magrittr_2.0.3                GenomeInfoDbData_1.2.10      </span></span>
<span><span class="co">## [123] patchwork_1.1.2               Rhdf5lib_1.22.0              </span></span>
<span><span class="co">## [125] munsell_0.5.0                 Rcpp_1.0.10                  </span></span>
<span><span class="co">## [127] viridis_0.6.3                 reticulate_1.30              </span></span>
<span><span class="co">## [129] stringi_1.7.12                zlibbioc_1.46.0              </span></span>
<span><span class="co">## [131] MASS_7.3-60                   AnnotationHub_3.8.0          </span></span>
<span><span class="co">## [133] plyr_1.8.8                    parallel_4.3.0               </span></span>
<span><span class="co">## [135] listenv_0.9.0                 ggrepel_0.9.3                </span></span>
<span><span class="co">## [137] deldir_1.0-9                  Biostrings_2.68.1            </span></span>
<span><span class="co">## [139] splines_4.3.0                 tensor_1.5                   </span></span>
<span><span class="co">## [141] locfit_1.5-9.8                igraph_1.5.0                 </span></span>
<span><span class="co">## [143] spatstat.geom_3.2-1           reshape2_1.4.4               </span></span>
<span><span class="co">## [145] ScaledMatrix_1.8.1            BiocVersion_3.17.1           </span></span>
<span><span class="co">## [147] evaluate_0.21                 BiocManager_1.30.21          </span></span>
<span><span class="co">## [149] httpuv_1.6.11                 RANN_2.6.1                   </span></span>
<span><span class="co">## [151] tidyr_1.3.0                   purrr_1.0.1                  </span></span>
<span><span class="co">## [153] polyclip_1.10-4               future_1.33.0                </span></span>
<span><span class="co">## [155] scattermore_1.2               rsvd_1.0.5                   </span></span>
<span><span class="co">## [157] xtable_1.8-4                  later_1.3.1                  </span></span>
<span><span class="co">## [159] viridisLite_0.4.2             ragg_1.2.5                   </span></span>
<span><span class="co">## [161] tibble_3.2.1                  memoise_2.0.1                </span></span>
<span><span class="co">## [163] beeswarm_0.4.0                AnnotationDbi_1.62.2         </span></span>
<span><span class="co">## [165] cluster_2.1.4                 globals_0.16.2               </span></span>
<span><span class="co">## [167] here_1.0.1</span></span></code></pre>
</div>
<div class="section level2">
<h2 id="further-reading">Further Reading<a class="anchor" aria-label="anchor" href="#further-reading"></a>
</h2>
<ul>
<li>OSCA book, <a href="https://bioconductor.org/books/release/OSCA.advanced/dealing-with-big-data.html" class="external-link">Chapter
14</a>: Dealing with big data</li>
<li>The <code>BiocParallel</code> <a href="https://bioconductor.org/packages/3.17/BiocParallel/vignettes/Introduction_To_BiocParallel.html" class="external-link">intro
vignette</a>.</li>
</ul>
</div>
<div class="section level2">
<h2 id="exercises">Exercises<a class="anchor" aria-label="anchor" href="#exercises"></a>
</h2>
<ol style="list-style-type: decimal">
<li>Out of memory representation: Write the counts matrix of the
wild-type chimera mouse gastrulation dataset to an HDF5 file. Create
another counts matrix that reads the data from the HDF5 file. Compare
memory usage of holding the entire matrix in memory as opposed to
holding the data out of memory.</li>
</ol>
<p>Hint: see the <code>HDF5Array</code> function for reading from HDF5
and the <code>writeHDF5Array</code> function for writing to HDF5 from
the <em><a href="https://bioconductor.org/packages/3.17/HDF5Array" class="external-link">HDF5Array</a></em>
package.</p>
<ol start="2" style="list-style-type: decimal">
<li>Parallelization: Perform a PCA analysis of the wild-type chimera
mouse gastrulation dataset using a multicore backend for parallel
computation. Compare the runtime of performing the PCA either in serial
execution mode, in multicore execution mode with 2 workers, and in
multicore execution mode with 3 workers.</li>
</ol>
<p>Hint: use the function <code>system.time</code> to obtain the runtime
of each job.</p>
<ol start="3" style="list-style-type: decimal">
<li>Conversion to Seurat: The <a href="https://bioconductor.org/packages/scRNAseq" class="external-link">scRNAseq</a> package
provides gene-level counts for a collection of public scRNA-seq
datasets, stored as <code>SingleCellExperiment</code> objects with
annotated cell- and gene-level metadata. Consult the vignette of the <a href="https://bioconductor.org/packages/scRNAseq" class="external-link">scRNAseq</a> package
to inspect all available datasets and select a dataset of your choice.
Convert the chosen dataset to a Seurat object and produce a PCA plot
with cells colored by a cell metadata column of your choice.</li>
</ol>
<p>Hint: use Seurat’s <code>DimPlot</code> function.</p>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">

        <nav id="toc" data-toggle="toc"><h2 data-toc-skip>Contents</h2>
    </nav>
</div>

</div>



      <footer><div class="copyright">
  <p></p>
<p>Developed by Davide Risso, Ludwig Geistlinger.</p>
</div>

<div class="pkgdown">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.7.</p>
</div>

      </footer>
</div>

  


  

  </body>
</html>
