<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>Multi-sample analyses • ISMB.OSCA</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link href="https://cdnjs.cloudflare.com/ajax/libs/bootswatch/3.4.0/flatly/bootstrap.min.css" rel="stylesheet" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="Multi-sample analyses">
<meta property="og:description" content="ISMB.OSCA">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">
    

    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">ISMB.OSCA</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">0.99.13</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Vignettes
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../articles/intro_bioc_sce.html">Introduction to Bioconductor</a>
    </li>
    <li>
      <a href="../articles/eda_qc.html">Exploratory data analysis</a>
    </li>
    <li>
      <a href="../articles/cell_type_annotation.html">Cell type annotation</a>
    </li>
    <li>
      <a href="../articles/multi_sample.html">Multi-sample analyses</a>
    </li>
    <li>
      <a href="../articles/large_data.html">Working with large data</a>
    </li>
    <li>
      <a href="../articles/hca.html">Accessing data from the Human Cell Atlas</a>
    </li>
  </ul>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://github.com/Bioconductor/ISMB.OSCA" class="external-link">
    <span class="fa fa-github"></span>
     
  </a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1 data-toc-skip>Multi-sample analyses</h1>
            
      
      <small class="dont-index">Source: <a href="https://github.com/Bioconductor/ISMB.OSCA/blob/HEAD/vignettes/multi_sample.Rmd" class="external-link"><code>vignettes/multi_sample.Rmd</code></a></small>
      <div class="hidden name"><code>multi_sample.Rmd</code></div>

    </div>

    
    
<div class="section level2">
<h2 id="setup-and-data-exploration">Setup and data exploration<a class="anchor" aria-label="anchor" href="#setup-and-data-exploration"></a>
</h2>
<p>As said, we will use the the wild-type data from the Tal1 chimera
experiment:</p>
<ul>
<li>Sample 5: E8.5 injected cells (tomato positive), pool 3</li>
<li>Sample 6: E8.5 host cells (tomato negative), pool 3</li>
<li>Sample 7: E8.5 injected cells (tomato positive), pool 4</li>
<li>Sample 8: E8.5 host cells (tomato negative), pool 4</li>
<li>Sample 9: E8.5 injected cells (tomato positive), pool 5</li>
<li>Sample 10: E8.5 host cells (tomato negative), pool 5</li>
</ul>
<p>Note that this is a paired design in which for each biological
replicate (pool 3, 4, and 5), we have both host and injected cells.</p>
<p>We start by loading the data and doing a quick exploratory analysis,
essentially applying the normalization and visualization techniques that
we have seen in the previous lectures to all samples.</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/MarioniLab/MouseGastrulationData" class="external-link">MouseGastrulationData</a></span><span class="op">)</span></span>
<span><span class="va">sce</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/MouseGastrulationData/man/WTChimeraData.html" class="external-link">WTChimeraData</a></span><span class="op">(</span>samples<span class="op">=</span><span class="fl">5</span><span class="op">:</span><span class="fl">10</span>, type <span class="op">=</span> <span class="st">"processed"</span><span class="op">)</span></span>
<span><span class="va">sce</span></span></code></pre></div>
<pre><code><span><span class="co">## class: SingleCellExperiment </span></span>
<span><span class="co">## dim: 29453 20935 </span></span>
<span><span class="co">## metadata(0):</span></span>
<span><span class="co">## assays(1): counts</span></span>
<span><span class="co">## rownames(29453): ENSMUSG00000051951 ENSMUSG00000089699 ...</span></span>
<span><span class="co">##   ENSMUSG00000095742 tomato-td</span></span>
<span><span class="co">## rowData names(2): ENSEMBL SYMBOL</span></span>
<span><span class="co">## colnames(20935): cell_9769 cell_9770 ... cell_30702 cell_30703</span></span>
<span><span class="co">## colData names(11): cell barcode ... doub.density sizeFactor</span></span>
<span><span class="co">## reducedDimNames(2): pca.corrected.E7.5 pca.corrected.E8.5</span></span>
<span><span class="co">## mainExpName: NULL</span></span>
<span><span class="co">## altExpNames(0):</span></span></code></pre>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">colData</a></span><span class="op">(</span><span class="va">sce</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## DataFrame with 20935 rows and 11 columns</span></span>
<span><span class="co">##                   cell          barcode    sample       stage    tomato</span></span>
<span><span class="co">##            &lt;character&gt;      &lt;character&gt; &lt;integer&gt; &lt;character&gt; &lt;logical&gt;</span></span>
<span><span class="co">## cell_9769    cell_9769 AAACCTGAGACTGTAA         5        E8.5      TRUE</span></span>
<span><span class="co">## cell_9770    cell_9770 AAACCTGAGATGCCTT         5        E8.5      TRUE</span></span>
<span><span class="co">## cell_9771    cell_9771 AAACCTGAGCAGCCTC         5        E8.5      TRUE</span></span>
<span><span class="co">## cell_9772    cell_9772 AAACCTGCATACTCTT         5        E8.5      TRUE</span></span>
<span><span class="co">## cell_9773    cell_9773 AAACGGGTCAACACCA         5        E8.5      TRUE</span></span>
<span><span class="co">## ...                ...              ...       ...         ...       ...</span></span>
<span><span class="co">## cell_30699  cell_30699 TTTGTCACAGCTCGCA        10        E8.5     FALSE</span></span>
<span><span class="co">## cell_30700  cell_30700 TTTGTCAGTCTAGTCA        10        E8.5     FALSE</span></span>
<span><span class="co">## cell_30701  cell_30701 TTTGTCATCATCGGAT        10        E8.5     FALSE</span></span>
<span><span class="co">## cell_30702  cell_30702 TTTGTCATCATTATCC        10        E8.5     FALSE</span></span>
<span><span class="co">## cell_30703  cell_30703 TTTGTCATCCCATTTA        10        E8.5     FALSE</span></span>
<span><span class="co">##                 pool stage.mapped        celltype.mapped closest.cell</span></span>
<span><span class="co">##            &lt;integer&gt;  &lt;character&gt;            &lt;character&gt;  &lt;character&gt;</span></span>
<span><span class="co">## cell_9769          3        E8.25             Mesenchyme   cell_24159</span></span>
<span><span class="co">## cell_9770          3         E8.5            Endothelium   cell_96660</span></span>
<span><span class="co">## cell_9771          3         E8.5              Allantois  cell_134982</span></span>
<span><span class="co">## cell_9772          3         E8.5             Erythroid3  cell_133892</span></span>
<span><span class="co">## cell_9773          3        E8.25             Erythroid1   cell_76296</span></span>
<span><span class="co">## ...              ...          ...                    ...          ...</span></span>
<span><span class="co">## cell_30699         5         E8.5             Erythroid3   cell_38810</span></span>
<span><span class="co">## cell_30700         5         E8.5       Surface ectoderm   cell_38588</span></span>
<span><span class="co">## cell_30701         5        E8.25 Forebrain/Midbrain/H..   cell_66082</span></span>
<span><span class="co">## cell_30702         5         E8.5             Erythroid3  cell_138114</span></span>
<span><span class="co">## cell_30703         5         E8.0                Doublet   cell_92644</span></span>
<span><span class="co">##            doub.density sizeFactor</span></span>
<span><span class="co">##               &lt;numeric&gt;  &lt;numeric&gt;</span></span>
<span><span class="co">## cell_9769    0.02985045    1.41243</span></span>
<span><span class="co">## cell_9770    0.00172753    1.22757</span></span>
<span><span class="co">## cell_9771    0.01338013    1.15439</span></span>
<span><span class="co">## cell_9772    0.00218402    1.28676</span></span>
<span><span class="co">## cell_9773    0.00211723    1.78719</span></span>
<span><span class="co">## ...                 ...        ...</span></span>
<span><span class="co">## cell_30699   0.00146287   0.389311</span></span>
<span><span class="co">## cell_30700   0.00374155   0.588784</span></span>
<span><span class="co">## cell_30701   0.05651258   0.624455</span></span>
<span><span class="co">## cell_30702   0.00108837   0.550807</span></span>
<span><span class="co">## cell_30703   0.82369305   1.184919</span></span></code></pre>
<p>To speed up computations, after removing doublets, we randomly select
50% cells per sample.</p>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="http://bioconductor.org/packages/scater/" class="external-link">scater</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://ggplot2.tidyverse.org" class="external-link">ggplot2</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">scran</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># remove doublets</span></span>
<span><span class="va">drop</span> <span class="op">&lt;-</span> <span class="va">sce</span><span class="op">$</span><span class="va">celltype.mapped</span> <span class="op"><a href="https://rdrr.io/pkg/BiocGenerics/man/match.html" class="external-link">%in%</a></span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"stripped"</span>, <span class="st">"Doublet"</span><span class="op">)</span></span>
<span><span class="va">sce</span> <span class="op">&lt;-</span> <span class="va">sce</span><span class="op">[</span>,<span class="op">!</span><span class="va">drop</span><span class="op">]</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/Random.html" class="external-link">set.seed</a></span><span class="op">(</span><span class="fl">29482</span><span class="op">)</span></span>
<span><span class="va">idx</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/unlist.html" class="external-link">unlist</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/tapply.html" class="external-link">tapply</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/row_colnames.html" class="external-link">colnames</a></span><span class="op">(</span><span class="va">sce</span><span class="op">)</span>, <span class="va">sce</span><span class="op">$</span><span class="va">sample</span>, <span class="kw">function</span><span class="op">(</span><span class="va">x</span><span class="op">)</span> <span class="op">{</span></span>
<span>    <span class="va">perc</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/Round.html" class="external-link">round</a></span><span class="op">(</span><span class="fl">0.50</span> <span class="op">*</span> <span class="fu"><a href="https://rdrr.io/r/base/length.html" class="external-link">length</a></span><span class="op">(</span><span class="va">x</span><span class="op">)</span><span class="op">)</span></span>
<span>    <span class="fu"><a href="https://rdrr.io/r/base/sample.html" class="external-link">sample</a></span><span class="op">(</span><span class="va">x</span>, <span class="va">perc</span><span class="op">)</span></span>
<span><span class="op">}</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="va">sce</span> <span class="op">&lt;-</span> <span class="va">sce</span><span class="op">[</span>,<span class="va">idx</span><span class="op">]</span></span></code></pre></div>
<p>We now normalize the data and visualize them in a tSNE plot.</p>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># normalization</span></span>
<span><span class="va">sce</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/scuttle/man/logNormCounts.html" class="external-link">logNormCounts</a></span><span class="op">(</span><span class="va">sce</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># identify highly variable genes</span></span>
<span><span class="va">dec</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/scran/man/modelGeneVar.html" class="external-link">modelGeneVar</a></span><span class="op">(</span><span class="va">sce</span>, block<span class="op">=</span><span class="va">sce</span><span class="op">$</span><span class="va">sample</span><span class="op">)</span></span>
<span><span class="va">chosen.hvgs</span> <span class="op">&lt;-</span> <span class="va">dec</span><span class="op">$</span><span class="va">bio</span> <span class="op">&gt;</span> <span class="fl">0</span></span>
<span></span>
<span><span class="co"># dimensionality reduction</span></span>
<span><span class="va">sce</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/BiocSingular/man/runPCA.html" class="external-link">runPCA</a></span><span class="op">(</span><span class="va">sce</span>, subset_row <span class="op">=</span> <span class="va">chosen.hvgs</span>, ntop <span class="op">=</span> <span class="fl">1000</span><span class="op">)</span></span>
<span><span class="va">sce</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/scater/man/runTSNE.html" class="external-link">runTSNE</a></span><span class="op">(</span><span class="va">sce</span>, dimred <span class="op">=</span> <span class="st">"PCA"</span><span class="op">)</span></span>
<span></span>
<span><span class="va">sce</span><span class="op">$</span><span class="va">sample</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/factor.html" class="external-link">as.factor</a></span><span class="op">(</span><span class="va">sce</span><span class="op">$</span><span class="va">sample</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/scater/man/plot_reddim.html" class="external-link">plotTSNE</a></span><span class="op">(</span><span class="va">sce</span>, colour_by <span class="op">=</span> <span class="st">"sample"</span><span class="op">)</span></span></code></pre></div>
<p><img src="multi_sample_files/figure-html/unnamed-chunk-2-1.png" width="700"></p>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/scater/man/plot_reddim.html" class="external-link">plotTSNE</a></span><span class="op">(</span><span class="va">sce</span>, colour_by <span class="op">=</span> <span class="st">"celltype.mapped"</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_colour_discrete.html" class="external-link">scale_color_discrete</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/theme.html" class="external-link">theme</a></span><span class="op">(</span>legend.position <span class="op">=</span> <span class="st">"bottom"</span><span class="op">)</span></span></code></pre></div>
<p><img src="multi_sample_files/figure-html/unnamed-chunk-2-2.png" width="700"></p>
<p>There are evident sample effects. Depending on the analysis that you
want to perform you may want to remove or retain the sample effect. For
instance, if the goal is to identify cell types with a clustering
method, one may want to remove the sample effects with “batch effect”
correction methods.</p>
<p>For now, let’s assume that we want to remove this effect.</p>
</div>
<div class="section level2">
<h2 id="correcting-batch-effects">Correcting batch effects<a class="anchor" aria-label="anchor" href="#correcting-batch-effects"></a>
</h2>
<p>We correct the effect of samples by aid of the
<code>correctExperiment</code> function in the <code>batchelor</code>
package and using the <code>sample</code> <code>colData</code> column as
batch.</p>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">batchelor</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/Random.html" class="external-link">set.seed</a></span><span class="op">(</span><span class="fl">10102</span><span class="op">)</span></span>
<span><span class="va">merged</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/batchelor/man/correctExperiments.html" class="external-link">correctExperiments</a></span><span class="op">(</span><span class="va">sce</span>, </span>
<span>    batch<span class="op">=</span><span class="va">sce</span><span class="op">$</span><span class="va">sample</span>, </span>
<span>    subset.row<span class="op">=</span><span class="va">chosen.hvgs</span>,</span>
<span>    PARAM<span class="op">=</span><span class="fu"><a href="https://rdrr.io/pkg/batchelor/man/BatchelorParam.html" class="external-link">FastMnnParam</a></span><span class="op">(</span></span>
<span>        merge.order<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span></span>
<span>            <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span><span class="fl">1</span>,<span class="fl">3</span>,<span class="fl">5</span><span class="op">)</span>, <span class="co"># WT (3 replicates)</span></span>
<span>            <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span><span class="fl">2</span>,<span class="fl">4</span>,<span class="fl">6</span><span class="op">)</span>  <span class="co"># td-Tomato (3 replicates)</span></span>
<span>        <span class="op">)</span></span>
<span>    <span class="op">)</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="va">merged</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/scater/man/runTSNE.html" class="external-link">runTSNE</a></span><span class="op">(</span><span class="va">merged</span>, dimred<span class="op">=</span><span class="st">"corrected"</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/scater/man/plot_reddim.html" class="external-link">plotTSNE</a></span><span class="op">(</span><span class="va">merged</span>, colour_by<span class="op">=</span><span class="st">"batch"</span><span class="op">)</span></span></code></pre></div>
<p><img src="multi_sample_files/figure-html/unnamed-chunk-3-1.png" width="700"></p>
<p>Once we removed the sample batch effect, we can proceed with the
Differential Expression Analysis.</p>
</div>
<div class="section level2">
<h2 id="differential-expression">Differential Expression<a class="anchor" aria-label="anchor" href="#differential-expression"></a>
</h2>
<p>In order to perform a Differential Expression Analysis, we need to
identify groups of cells across samples/conditions (depending on the
experimental design and the final aim of the experiment).</p>
<p>As previously seen, we have two ways of grouping cells, cell
clustering and cell labeling. In our case we will focus on this second
aspect to group cells according to the already annotated cell types to
proceed with the computation of the pseudo-bulk samples.</p>
<div class="section level3">
<h3 id="pseudo-bulk-samples">Pseudo-bulk samples<a class="anchor" aria-label="anchor" href="#pseudo-bulk-samples"></a>
</h3>
<p>To compute differences between groups of cells, a possible way is to
compute pseudo-bulk samples, where we mediate the gene signal of all the
cells for each specific cell type. In this manner, we are then able to
detect differences between the same cell type across two different
conditions.</p>
<p>To compute pseudo-bulk samples, we use the
<code>aggregateAcrossCells</code> function in the <code>scuttle</code>
package, which takes as input not only a SingleCellExperiment, but also
the id to use for the identification of the group of cells. In our case,
we use as id not just the cell type, but also the sample, because we
want be able to discern between replicates and conditions during further
steps.</p>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Using 'label' and 'sample' as our two factors; each column of the output</span></span>
<span><span class="co"># corresponds to one unique combination of these two factors.</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">scuttle</span><span class="op">)</span></span>
<span><span class="va">summed</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/scuttle/man/aggregateAcrossCells.html" class="external-link">aggregateAcrossCells</a></span><span class="op">(</span><span class="va">merged</span>, </span>
<span>    id<span class="op">=</span><span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">colData</a></span><span class="op">(</span><span class="va">merged</span><span class="op">)</span><span class="op">[</span>,<span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"celltype.mapped"</span>, <span class="st">"sample"</span><span class="op">)</span><span class="op">]</span><span class="op">)</span></span>
<span><span class="va">summed</span></span></code></pre></div>
<pre><code><span><span class="co">## class: SingleCellExperiment </span></span>
<span><span class="co">## dim: 13641 179 </span></span>
<span><span class="co">## metadata(2): merge.info pca.info</span></span>
<span><span class="co">## assays(1): counts</span></span>
<span><span class="co">## rownames(13641): ENSMUSG00000051951 ENSMUSG00000025900 ...</span></span>
<span><span class="co">##   ENSMUSG00000096730 ENSMUSG00000095742</span></span>
<span><span class="co">## rowData names(3): rotation ENSEMBL SYMBOL</span></span>
<span><span class="co">## colnames: NULL</span></span>
<span><span class="co">## colData names(15): batch cell ... sample ncells</span></span>
<span><span class="co">## reducedDimNames(5): corrected pca.corrected.E7.5 pca.corrected.E8.5 PCA</span></span>
<span><span class="co">##   TSNE</span></span>
<span><span class="co">## mainExpName: NULL</span></span>
<span><span class="co">## altExpNames(0):</span></span></code></pre>
</div>
<div class="section level3">
<h3 id="differential-expression-analysis">Differential Expression Analysis<a class="anchor" aria-label="anchor" href="#differential-expression-analysis"></a>
</h3>
<p>The main advantage of using pseudo-bulk samples is the possibility to
use well-tested methods for differential analysis like
<code>edgeR</code> and <code>DESeq2</code>, we will focus on the former
for this analysis. <code>edgeR</code> uses a Negative Binomial
Generalized Linear Model.</p>
<p>First, let’s start with a specific cell type, for instance the
“Mesenchymal stem cells”, and look into differences between this cell
type across conditions.</p>
<div class="sourceCode" id="cb11"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">label</span> <span class="op">&lt;-</span> <span class="st">"Mesenchyme"</span></span>
<span><span class="va">current</span> <span class="op">&lt;-</span> <span class="va">summed</span><span class="op">[</span>,<span class="va">label</span><span class="op">==</span><span class="va">summed</span><span class="op">$</span><span class="va">celltype.mapped</span><span class="op">]</span></span>
<span></span>
<span><span class="co"># Creating up a DGEList object for use in edgeR:</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://bioinf.wehi.edu.au/edgeR/" class="external-link">edgeR</a></span><span class="op">)</span></span>
<span><span class="va">y</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/edgeR/man/DGEList.html" class="external-link">DGEList</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/dge.html" class="external-link">counts</a></span><span class="op">(</span><span class="va">current</span><span class="op">)</span>, samples<span class="op">=</span><span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">colData</a></span><span class="op">(</span><span class="va">current</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">y</span></span></code></pre></div>
<pre><code><span><span class="co">## An object of class "DGEList"</span></span>
<span><span class="co">## $counts</span></span>
<span><span class="co">##                    Sample1 Sample2 Sample3 Sample4 Sample5 Sample6</span></span>
<span><span class="co">## ENSMUSG00000051951       2       0       0       0       1       0</span></span>
<span><span class="co">## ENSMUSG00000025900       0       0       0       0       0       0</span></span>
<span><span class="co">## ENSMUSG00000025902       4       0       2       0       3       6</span></span>
<span><span class="co">## ENSMUSG00000033845     765     130     508     213     781     305</span></span>
<span><span class="co">## ENSMUSG00000002459       2       0       1       0       0       0</span></span>
<span><span class="co">## 13636 more rows ...</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## $samples</span></span>
<span><span class="co">##         group lib.size norm.factors batch cell barcode sample stage tomato pool</span></span>
<span><span class="co">## Sample1     1  2478901            1     5 &lt;NA&gt;    &lt;NA&gt;      5  E8.5   TRUE    3</span></span>
<span><span class="co">## Sample2     1   548407            1     6 &lt;NA&gt;    &lt;NA&gt;      6  E8.5  FALSE    3</span></span>
<span><span class="co">## Sample3     1  1260187            1     7 &lt;NA&gt;    &lt;NA&gt;      7  E8.5   TRUE    4</span></span>
<span><span class="co">## Sample4     1   578699            1     8 &lt;NA&gt;    &lt;NA&gt;      8  E8.5  FALSE    4</span></span>
<span><span class="co">## Sample5     1  2092329            1     9 &lt;NA&gt;    &lt;NA&gt;      9  E8.5   TRUE    5</span></span>
<span><span class="co">## Sample6     1   904929            1    10 &lt;NA&gt;    &lt;NA&gt;     10  E8.5  FALSE    5</span></span>
<span><span class="co">##         stage.mapped celltype.mapped closest.cell doub.density sizeFactor</span></span>
<span><span class="co">## Sample1         &lt;NA&gt;      Mesenchyme         &lt;NA&gt;           NA         NA</span></span>
<span><span class="co">## Sample2         &lt;NA&gt;      Mesenchyme         &lt;NA&gt;           NA         NA</span></span>
<span><span class="co">## Sample3         &lt;NA&gt;      Mesenchyme         &lt;NA&gt;           NA         NA</span></span>
<span><span class="co">## Sample4         &lt;NA&gt;      Mesenchyme         &lt;NA&gt;           NA         NA</span></span>
<span><span class="co">## Sample5         &lt;NA&gt;      Mesenchyme         &lt;NA&gt;           NA         NA</span></span>
<span><span class="co">## Sample6         &lt;NA&gt;      Mesenchyme         &lt;NA&gt;           NA         NA</span></span>
<span><span class="co">##         celltype.mapped.1 sample.1 ncells</span></span>
<span><span class="co">## Sample1        Mesenchyme        5    151</span></span>
<span><span class="co">## Sample2        Mesenchyme        6     28</span></span>
<span><span class="co">## Sample3        Mesenchyme        7    127</span></span>
<span><span class="co">## Sample4        Mesenchyme        8     75</span></span>
<span><span class="co">## Sample5        Mesenchyme        9    239</span></span>
<span><span class="co">## Sample6        Mesenchyme       10    146</span></span></code></pre>
<p>A typical step is to discard low quality samples due to low sequenced
library size. We discard these samples because they can affect further
steps like normalization and/or DEGs analysis.</p>
<p>We can see that in our case we don’t have low quality samples and we
don’t need to filter out any of them.</p>
<div class="sourceCode" id="cb13"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">discarded</span> <span class="op">&lt;-</span> <span class="va">current</span><span class="op">$</span><span class="va">ncells</span> <span class="op">&lt;</span> <span class="fl">10</span></span>
<span><span class="va">y</span> <span class="op">&lt;-</span> <span class="va">y</span><span class="op">[</span>,<span class="op">!</span><span class="va">discarded</span><span class="op">]</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/summary.html" class="external-link">summary</a></span><span class="op">(</span><span class="va">discarded</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">##    Mode   FALSE </span></span>
<span><span class="co">## logical       6</span></span></code></pre>
<p>The same idea is typically applied to the genes, indeed we need to
discard low expressed genes to improve accuracy for the DEGs
modeling.</p>
<div class="sourceCode" id="cb15"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">keep</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/edgeR/man/filterByExpr.html" class="external-link">filterByExpr</a></span><span class="op">(</span><span class="va">y</span>, group<span class="op">=</span><span class="va">current</span><span class="op">$</span><span class="va">tomato</span><span class="op">)</span></span>
<span><span class="va">y</span> <span class="op">&lt;-</span> <span class="va">y</span><span class="op">[</span><span class="va">keep</span>,<span class="op">]</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/summary.html" class="external-link">summary</a></span><span class="op">(</span><span class="va">keep</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">##    Mode   FALSE    TRUE </span></span>
<span><span class="co">## logical    9121    4520</span></span></code></pre>
<p>We can now proceed to normalize the data There are several approaches
for normalizing bulk, and hence pseudo-bulk data. Here, we use the
Trimmed Mean of M-values method, implemented in the <code>edgeR</code>
package within the <code>calcNormFactors</code> function. Keep in mind
that because we are going to normalize the pseudo-bulk counts, we don’t
need to normalize the data in “single cell form”.</p>
<div class="sourceCode" id="cb17"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">y</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/edgeR/man/calcNormFactors.html" class="external-link">calcNormFactors</a></span><span class="op">(</span><span class="va">y</span><span class="op">)</span></span>
<span><span class="va">y</span><span class="op">$</span><span class="va">samples</span></span></code></pre></div>
<pre><code><span><span class="co">##         group lib.size norm.factors batch cell barcode sample stage tomato pool</span></span>
<span><span class="co">## Sample1     1  2478901    1.0506857     5 &lt;NA&gt;    &lt;NA&gt;      5  E8.5   TRUE    3</span></span>
<span><span class="co">## Sample2     1   548407    1.0399112     6 &lt;NA&gt;    &lt;NA&gt;      6  E8.5  FALSE    3</span></span>
<span><span class="co">## Sample3     1  1260187    0.9700083     7 &lt;NA&gt;    &lt;NA&gt;      7  E8.5   TRUE    4</span></span>
<span><span class="co">## Sample4     1   578699    0.9871129     8 &lt;NA&gt;    &lt;NA&gt;      8  E8.5  FALSE    4</span></span>
<span><span class="co">## Sample5     1  2092329    0.9695559     9 &lt;NA&gt;    &lt;NA&gt;      9  E8.5   TRUE    5</span></span>
<span><span class="co">## Sample6     1   904929    0.9858611    10 &lt;NA&gt;    &lt;NA&gt;     10  E8.5  FALSE    5</span></span>
<span><span class="co">##         stage.mapped celltype.mapped closest.cell doub.density sizeFactor</span></span>
<span><span class="co">## Sample1         &lt;NA&gt;      Mesenchyme         &lt;NA&gt;           NA         NA</span></span>
<span><span class="co">## Sample2         &lt;NA&gt;      Mesenchyme         &lt;NA&gt;           NA         NA</span></span>
<span><span class="co">## Sample3         &lt;NA&gt;      Mesenchyme         &lt;NA&gt;           NA         NA</span></span>
<span><span class="co">## Sample4         &lt;NA&gt;      Mesenchyme         &lt;NA&gt;           NA         NA</span></span>
<span><span class="co">## Sample5         &lt;NA&gt;      Mesenchyme         &lt;NA&gt;           NA         NA</span></span>
<span><span class="co">## Sample6         &lt;NA&gt;      Mesenchyme         &lt;NA&gt;           NA         NA</span></span>
<span><span class="co">##         celltype.mapped.1 sample.1 ncells</span></span>
<span><span class="co">## Sample1        Mesenchyme        5    151</span></span>
<span><span class="co">## Sample2        Mesenchyme        6     28</span></span>
<span><span class="co">## Sample3        Mesenchyme        7    127</span></span>
<span><span class="co">## Sample4        Mesenchyme        8     75</span></span>
<span><span class="co">## Sample5        Mesenchyme        9    239</span></span>
<span><span class="co">## Sample6        Mesenchyme       10    146</span></span></code></pre>
<p>To investigate the effect of our normalization, we use a
Mean-Difference (MD) plot for each sample in order to detect possible
normalization problems due to insufficient cells/reads/UMIs composing a
particular pseudo-bulk profile.</p>
<p>In our case, we verify that all these plots are centered in 0 (on
y-axis) and present a trumpet shape, as expected.</p>
<div class="sourceCode" id="cb19"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/graphics/par.html" class="external-link">par</a></span><span class="op">(</span>mfrow<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">2</span>,<span class="fl">3</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="kw">for</span> <span class="op">(</span><span class="va">i</span> <span class="kw">in</span> <span class="fu"><a href="https://rdrr.io/r/base/seq.html" class="external-link">seq_len</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/nrow.html" class="external-link">ncol</a></span><span class="op">(</span><span class="va">y</span><span class="op">)</span><span class="op">)</span><span class="op">)</span> <span class="op">{</span></span>
<span>    <span class="fu"><a href="https://rdrr.io/pkg/limma/man/plotMD.html" class="external-link">plotMD</a></span><span class="op">(</span><span class="va">y</span>, column<span class="op">=</span><span class="va">i</span><span class="op">)</span></span>
<span><span class="op">}</span></span></code></pre></div>
<p><img src="multi_sample_files/figure-html/unnamed-chunk-9-1.png" width="700"></p>
<p>Furthermore, we want to check if the samples cluster together based
on their known factors (like the tomato injection in this case).</p>
<p>To do so, we use the MDS plot, which is very close to a PCA
representation.</p>
<div class="sourceCode" id="cb20"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">limma</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/limma/man/plotMDS.html" class="external-link">plotMDS</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/edgeR/man/cpm.html" class="external-link">cpm</a></span><span class="op">(</span><span class="va">y</span>, log<span class="op">=</span><span class="cn">TRUE</span><span class="op">)</span>, </span>
<span>    col<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/ifelse.html" class="external-link">ifelse</a></span><span class="op">(</span><span class="va">y</span><span class="op">$</span><span class="va">samples</span><span class="op">$</span><span class="va">tomato</span>, <span class="st">"red"</span>, <span class="st">"blue"</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<p><img src="multi_sample_files/figure-html/unnamed-chunk-10-1.png" width="700"></p>
<p>We then construct a design matrix by including both the pool and the
tomato as factors. This design indicates which samples belong to which
pool and condition, so we can use it in the next step of the
analysis.</p>
<div class="sourceCode" id="cb21"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">design</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/stats/model.matrix.html" class="external-link">model.matrix</a></span><span class="op">(</span><span class="op">~</span><span class="fu"><a href="https://rdrr.io/r/base/factor.html" class="external-link">factor</a></span><span class="op">(</span><span class="va">pool</span><span class="op">)</span> <span class="op">+</span> <span class="fu"><a href="https://rdrr.io/r/base/factor.html" class="external-link">factor</a></span><span class="op">(</span><span class="va">tomato</span><span class="op">)</span>, <span class="va">y</span><span class="op">$</span><span class="va">samples</span><span class="op">)</span></span>
<span><span class="va">design</span></span></code></pre></div>
<pre><code><span><span class="co">##         (Intercept) factor(pool)4 factor(pool)5 factor(tomato)TRUE</span></span>
<span><span class="co">## Sample1           1             0             0                  1</span></span>
<span><span class="co">## Sample2           1             0             0                  0</span></span>
<span><span class="co">## Sample3           1             1             0                  1</span></span>
<span><span class="co">## Sample4           1             1             0                  0</span></span>
<span><span class="co">## Sample5           1             0             1                  1</span></span>
<span><span class="co">## Sample6           1             0             1                  0</span></span>
<span><span class="co">## attr(,"assign")</span></span>
<span><span class="co">## [1] 0 1 1 2</span></span>
<span><span class="co">## attr(,"contrasts")</span></span>
<span><span class="co">## attr(,"contrasts")$`factor(pool)`</span></span>
<span><span class="co">## [1] "contr.treatment"</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## attr(,"contrasts")$`factor(tomato)`</span></span>
<span><span class="co">## [1] "contr.treatment"</span></span></code></pre>
<p>Now we can estimate the Negative Binomial (NB) overdispersion
parameter, to model the mean-variance trend.</p>
<div class="sourceCode" id="cb23"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">y</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/edgeR/man/estimateDisp.html" class="external-link">estimateDisp</a></span><span class="op">(</span><span class="va">y</span>, <span class="va">design</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/summary.html" class="external-link">summary</a></span><span class="op">(</span><span class="va">y</span><span class="op">$</span><span class="va">trended.dispersion</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">##     Min.  1st Qu.   Median     Mean  3rd Qu.     Max. </span></span>
<span><span class="co">## 0.009325 0.016271 0.024233 0.021603 0.026868 0.027993</span></span></code></pre>
<p>The BCV plot allows us to investigate the relation between the
Biological Coefficient of Variation and the Average log CPM for each
gene. Additionally, the Common and Trend BCV are shown in
<code>red</code> and <code>blue</code>.</p>
<div class="sourceCode" id="cb25"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/edgeR/man/plotBCV.html" class="external-link">plotBCV</a></span><span class="op">(</span><span class="va">y</span><span class="op">)</span></span></code></pre></div>
<p><img src="multi_sample_files/figure-html/unnamed-chunk-13-1.png" width="700"></p>
<p>We then fit a Quasi-Likelihood (QL) negative binomial generalized
linear model for each gene. The <code>robust=TRUE</code> parameter
avoids distorsions from highly variable clusters. The QL method includes
an additional dispersion parameter, useful to handle the uncertainty and
variability of the per-gene variance, which is not well estimated by the
NB dispersions, so the two dispersion types complement each other in the
final analysis.</p>
<div class="sourceCode" id="cb26"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">fit</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/edgeR/man/glmQLFTest.html" class="external-link">glmQLFit</a></span><span class="op">(</span><span class="va">y</span>, <span class="va">design</span>, robust<span class="op">=</span><span class="cn">TRUE</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/summary.html" class="external-link">summary</a></span><span class="op">(</span><span class="va">fit</span><span class="op">$</span><span class="va">var.prior</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">##    Min. 1st Qu.  Median    Mean 3rd Qu.    Max. </span></span>
<span><span class="co">##  0.2977  0.6640  0.8275  0.7637  0.8798  0.9670</span></span></code></pre>
<div class="sourceCode" id="cb28"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/summary.html" class="external-link">summary</a></span><span class="op">(</span><span class="va">fit</span><span class="op">$</span><span class="va">df.prior</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">##    Min. 1st Qu.  Median    Mean 3rd Qu.    Max. </span></span>
<span><span class="co">##   0.323   8.177   8.177   8.111   8.177   8.177</span></span></code></pre>
<p>QL dispersion estimates for each gene as a function of abundance. Raw
estimates (black) are shrunk towards the trend (blue) to yield squeezed
estimates (red).</p>
<div class="sourceCode" id="cb30"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/edgeR/man/plotQLDisp.html" class="external-link">plotQLDisp</a></span><span class="op">(</span><span class="va">fit</span><span class="op">)</span></span></code></pre></div>
<p><img src="multi_sample_files/figure-html/unnamed-chunk-15-1.png" width="700"></p>
<p>We then use an empirical Bayes quasi-likelihood F-test to test for
differential expression (due to tomato injection) per each gene at a
False Discovery Rate (FDR) of 5%. The low amount of DGEs highlights that
the tomato injection effect has a low influence on the mesenchyme
cells.</p>
<div class="sourceCode" id="cb31"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">res</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/edgeR/man/glmQLFTest.html" class="external-link">glmQLFTest</a></span><span class="op">(</span><span class="va">fit</span>, coef<span class="op">=</span><span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/nrow.html" class="external-link">ncol</a></span><span class="op">(</span><span class="va">design</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/summary.html" class="external-link">summary</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/limma/man/decideTests.html" class="external-link">decideTests</a></span><span class="op">(</span><span class="va">res</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">##        factor(tomato)TRUE</span></span>
<span><span class="co">## Down                    5</span></span>
<span><span class="co">## NotSig               4510</span></span>
<span><span class="co">## Up                      5</span></span></code></pre>
<div class="sourceCode" id="cb33"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/edgeR/man/topTags.html" class="external-link">topTags</a></span><span class="op">(</span><span class="va">res</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## Coefficient:  factor(tomato)TRUE </span></span>
<span><span class="co">##                         logFC   logCPM          F       PValue          FDR</span></span>
<span><span class="co">## ENSMUSG00000010760 -4.1551264 9.973704 1112.14948 9.905998e-12 4.477511e-08</span></span>
<span><span class="co">## ENSMUSG00000096768  1.9992920 8.844258  403.85294 1.594095e-09 3.602655e-06</span></span>
<span><span class="co">## ENSMUSG00000035299  1.8001627 6.904163  123.52980 5.130084e-07 7.729327e-04</span></span>
<span><span class="co">## ENSMUSG00000101609  1.3708397 7.310009   80.58075 3.745290e-06 4.232177e-03</span></span>
<span><span class="co">## ENSMUSG00000019188 -1.0195649 7.545530   61.65538 1.248303e-05 1.128466e-02</span></span>
<span><span class="co">## ENSMUSG00000024423  0.9946833 7.391075   58.34967 1.591674e-05 1.199061e-02</span></span>
<span><span class="co">## ENSMUSG00000086503 -6.5155131 7.411257  159.33690 2.625600e-05 1.695388e-02</span></span>
<span><span class="co">## ENSMUSG00000042607 -0.9567347 7.468203   45.42154 4.690293e-05 2.650016e-02</span></span>
<span><span class="co">## ENSMUSG00000036446 -0.8305889 9.401028   42.72058 6.071290e-05 3.049137e-02</span></span>
<span><span class="co">## ENSMUSG00000027520  1.5814592 6.952923   40.94715 7.775888e-05 3.514702e-02</span></span></code></pre>
<p>All the previous steps can be easily performed with the following
function for each cell type, thanks to the <code>pseudoBulkDGE</code>
function in the <code>scran</code> package.</p>
<div class="sourceCode" id="cb35"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">scran</span><span class="op">)</span></span>
<span><span class="va">summed.filt</span> <span class="op">&lt;-</span> <span class="va">summed</span><span class="op">[</span>,<span class="va">summed</span><span class="op">$</span><span class="va">ncells</span> <span class="op">&gt;=</span> <span class="fl">10</span><span class="op">]</span></span>
<span></span>
<span><span class="va">de.results</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/scran/man/pseudoBulkDGE.html" class="external-link">pseudoBulkDGE</a></span><span class="op">(</span><span class="va">summed.filt</span>, </span>
<span>    label<span class="op">=</span><span class="va">summed.filt</span><span class="op">$</span><span class="va">celltype.mapped</span>,</span>
<span>    design<span class="op">=</span><span class="op">~</span><span class="fu"><a href="https://rdrr.io/r/base/factor.html" class="external-link">factor</a></span><span class="op">(</span><span class="va">pool</span><span class="op">)</span> <span class="op">+</span> <span class="va">tomato</span>,</span>
<span>    coef<span class="op">=</span><span class="st">"tomatoTRUE"</span>,</span>
<span>    condition<span class="op">=</span><span class="va">summed.filt</span><span class="op">$</span><span class="va">tomato</span> </span>
<span><span class="op">)</span></span></code></pre></div>
<p>The returned object is a list of <code>DataFrame</code>s each with
the results for a cell type. Each of these contains also the
intermediate results in <code>edgeR</code> format to perform any
intermediate plot or diagnostic.</p>
<div class="sourceCode" id="cb36"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">cur.results</span> <span class="op">&lt;-</span> <span class="va">de.results</span><span class="op">[[</span><span class="st">"Allantois"</span><span class="op">]</span><span class="op">]</span></span>
<span><span class="va">cur.results</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/order.html" class="external-link">order</a></span><span class="op">(</span><span class="va">cur.results</span><span class="op">$</span><span class="va">PValue</span><span class="op">)</span>,<span class="op">]</span></span></code></pre></div>
<pre><code><span><span class="co">## DataFrame with 13641 rows and 5 columns</span></span>
<span><span class="co">##                        logFC    logCPM         F      PValue         FDR</span></span>
<span><span class="co">##                    &lt;numeric&gt; &lt;numeric&gt; &lt;numeric&gt;   &lt;numeric&gt;   &lt;numeric&gt;</span></span>
<span><span class="co">## ENSMUSG00000037664 -7.995130  11.55290  3180.990 7.35933e-22 3.09165e-18</span></span>
<span><span class="co">## ENSMUSG00000010760 -2.574762  12.40592  1114.529 9.22901e-18 1.93855e-14</span></span>
<span><span class="co">## ENSMUSG00000086503 -7.015319   7.49749   703.373 5.57372e-16 7.80507e-13</span></span>
<span><span class="co">## ENSMUSG00000096768  1.828480   9.33239   304.769 8.39747e-13 8.81944e-10</span></span>
<span><span class="co">## ENSMUSG00000022464  0.969837  10.28302   118.697 2.12502e-09 1.78544e-06</span></span>
<span><span class="co">## ...                      ...       ...       ...         ...         ...</span></span>
<span><span class="co">## ENSMUSG00000095247        NA        NA        NA          NA          NA</span></span>
<span><span class="co">## ENSMUSG00000096808        NA        NA        NA          NA          NA</span></span>
<span><span class="co">## ENSMUSG00000079808        NA        NA        NA          NA          NA</span></span>
<span><span class="co">## ENSMUSG00000096730        NA        NA        NA          NA          NA</span></span>
<span><span class="co">## ENSMUSG00000095742        NA        NA        NA          NA          NA</span></span></code></pre>
</div>
</div>
<div class="section level2">
<h2 id="differential-abundance">Differential Abundance<a class="anchor" aria-label="anchor" href="#differential-abundance"></a>
</h2>
<p>With DA we test for differences between clusters across conditions,
to investigate which clusters change accordingly to the treatment (the
tomato injection in our case).</p>
<p>We first setup some code and variables for further analysis, like
quantifying the number of cells per each cell type and fit a model to
catch differences between the injected cells and the background.</p>
<p>The performed steps are very similar to the ones for DEGs analysis,
but this time we start our analysis on the computed abundances and
without normalizing the data with TMM.</p>
<div class="sourceCode" id="cb38"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://bioinf.wehi.edu.au/edgeR/" class="external-link">edgeR</a></span><span class="op">)</span></span>
<span><span class="va">abundances</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/table.html" class="external-link">table</a></span><span class="op">(</span><span class="va">merged</span><span class="op">$</span><span class="va">celltype.mapped</span>, <span class="va">merged</span><span class="op">$</span><span class="va">sample</span><span class="op">)</span> </span>
<span><span class="va">abundances</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/class.html" class="external-link">unclass</a></span><span class="op">(</span><span class="va">abundances</span><span class="op">)</span> </span>
<span><span class="co"># Attaching some column metadata.</span></span>
<span><span class="va">extra.info</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">colData</a></span><span class="op">(</span><span class="va">merged</span><span class="op">)</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/match.html" class="external-link">match</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/row_colnames.html" class="external-link">colnames</a></span><span class="op">(</span><span class="va">abundances</span><span class="op">)</span>, <span class="va">merged</span><span class="op">$</span><span class="va">sample</span><span class="op">)</span>,<span class="op">]</span></span>
<span><span class="va">y.ab</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/edgeR/man/DGEList.html" class="external-link">DGEList</a></span><span class="op">(</span><span class="va">abundances</span>, samples<span class="op">=</span><span class="va">extra.info</span><span class="op">)</span></span>
<span></span>
<span><span class="va">design</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/stats/model.matrix.html" class="external-link">model.matrix</a></span><span class="op">(</span><span class="op">~</span><span class="fu"><a href="https://rdrr.io/r/base/factor.html" class="external-link">factor</a></span><span class="op">(</span><span class="va">pool</span><span class="op">)</span> <span class="op">+</span> <span class="fu"><a href="https://rdrr.io/r/base/factor.html" class="external-link">factor</a></span><span class="op">(</span><span class="va">tomato</span><span class="op">)</span>, <span class="va">y.ab</span><span class="op">$</span><span class="va">samples</span><span class="op">)</span></span>
<span><span class="va">y.ab</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/edgeR/man/estimateDisp.html" class="external-link">estimateDisp</a></span><span class="op">(</span><span class="va">y.ab</span>, <span class="va">design</span>, trend<span class="op">=</span><span class="st">"none"</span><span class="op">)</span></span>
<span><span class="va">fit.ab</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/edgeR/man/glmQLFTest.html" class="external-link">glmQLFit</a></span><span class="op">(</span><span class="va">y.ab</span>, <span class="va">design</span>, robust<span class="op">=</span><span class="cn">TRUE</span>, abundance.trend<span class="op">=</span><span class="cn">FALSE</span><span class="op">)</span></span></code></pre></div>
<div class="section level3">
<h3 id="background-on-compositional-effect">Background on compositional effect<a class="anchor" aria-label="anchor" href="#background-on-compositional-effect"></a>
</h3>
<p>As mentioned before, in DA we don’t normalize our data with
<code>calcNormFactors</code> function, because this approach considers
that most of the input features do not vary between conditions. This
cannot be applied to this analysis because we have a small number of
cell populations that all can change due to the treatment. Leaving us to
normalize only for library depth, which in pseudo-bulk data means by the
total number of cells in each sample (cell type).</p>
<p>On the other hand, this can lead our data to be susceptible to
compositional effect, that means that our conclusions can be biased by
the amount of cells present in each cell type. And this amount of cells
can be totally unbalanced between cell types.</p>
<p>For example, a specific cell type can be 40% of the total amount of
cells present in the experiment, while another just the 3%. The
differences in terms of abundance of these cell types are detected
between the different conditions, but our final interpretation could be
biased if we don’t consider this aspect.</p>
<p>We now look at different approaches for handling the compositional
effect.</p>
</div>
<div class="section level3">
<h3 id="assuming-most-labels-do-not-change">Assuming most labels do not change<a class="anchor" aria-label="anchor" href="#assuming-most-labels-do-not-change"></a>
</h3>
<p>We can use a similar approach used during the DEGs analysis, assuming
that most labels are not changing, in particular if we think about the
low number of DEGs resulted from the previous analysis.</p>
<p>To do so, we first normalize the data with
<code>calcNormFactors</code> and then we fit and estimate a QL-model for
our abundance data.</p>
<div class="sourceCode" id="cb39"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">y.ab2</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/edgeR/man/calcNormFactors.html" class="external-link">calcNormFactors</a></span><span class="op">(</span><span class="va">y.ab</span><span class="op">)</span></span>
<span><span class="va">y.ab2</span><span class="op">$</span><span class="va">samples</span><span class="op">$</span><span class="va">norm.factors</span></span></code></pre></div>
<pre><code><span><span class="co">## [1] 1.1029040 1.0228173 1.0695358 0.7686501 1.0402941 1.0365354</span></span></code></pre>
<p>We then follow the already seen edgeR analysis steps.</p>
<div class="sourceCode" id="cb41"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">y.ab2</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/edgeR/man/estimateDisp.html" class="external-link">estimateDisp</a></span><span class="op">(</span><span class="va">y.ab2</span>, <span class="va">design</span>, trend<span class="op">=</span><span class="st">"none"</span><span class="op">)</span></span>
<span><span class="va">fit.ab2</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/edgeR/man/glmQLFTest.html" class="external-link">glmQLFit</a></span><span class="op">(</span><span class="va">y.ab2</span>, <span class="va">design</span>, robust<span class="op">=</span><span class="cn">TRUE</span>, abundance.trend<span class="op">=</span><span class="cn">FALSE</span><span class="op">)</span></span>
<span><span class="va">res2</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/edgeR/man/glmQLFTest.html" class="external-link">glmQLFTest</a></span><span class="op">(</span><span class="va">fit.ab2</span>, coef<span class="op">=</span><span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/nrow.html" class="external-link">ncol</a></span><span class="op">(</span><span class="va">design</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/summary.html" class="external-link">summary</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/limma/man/decideTests.html" class="external-link">decideTests</a></span><span class="op">(</span><span class="va">res2</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">##        factor(tomato)TRUE</span></span>
<span><span class="co">## Down                    2</span></span>
<span><span class="co">## NotSig                 32</span></span>
<span><span class="co">## Up                      0</span></span></code></pre>
<div class="sourceCode" id="cb43"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/edgeR/man/topTags.html" class="external-link">topTags</a></span><span class="op">(</span><span class="va">res2</span>, n<span class="op">=</span><span class="fl">10</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## Coefficient:  factor(tomato)TRUE </span></span>
<span><span class="co">##                        logFC   logCPM         F       PValue          FDR</span></span>
<span><span class="co">## ExE ectoderm      -5.7452733 13.13490 37.367234 5.393309e-08 1.833725e-06</span></span>
<span><span class="co">## Parietal endoderm -6.9016375 12.36649 25.884039 3.062123e-06 5.205609e-05</span></span>
<span><span class="co">## Mesenchyme         0.9656118 16.32654  6.139328 1.570983e-02 1.495573e-01</span></span>
<span><span class="co">## Erythroid3        -0.9192068 17.34677  5.921070 1.759497e-02 1.495573e-01</span></span>
<span><span class="co">## Neural crest      -1.0200609 14.83912  5.276218 2.470370e-02 1.679851e-01</span></span>
<span><span class="co">## ExE endoderm      -3.9992127 10.75172  4.673383 3.415631e-02 1.935524e-01</span></span>
<span><span class="co">## Endothelium        0.8666732 14.12195  3.307087 7.338557e-02 3.564442e-01</span></span>
<span><span class="co">## Cardiomyocytes     0.6956771 14.93321  2.592279 1.120186e-01 4.760789e-01</span></span>
<span><span class="co">## Allantois          0.6001360 15.54924  2.085500 1.532954e-01 5.791158e-01</span></span>
<span><span class="co">## Erythroid2        -0.5177901 15.97357  1.614924 2.081314e-01 7.076469e-01</span></span></code></pre>
</div>
<div class="section level3">
<h3 id="testing-against-a-log-fold-change-threshold">Testing against a log-fold change threshold<a class="anchor" aria-label="anchor" href="#testing-against-a-log-fold-change-threshold"></a>
</h3>
<p>This other approach assumes that the composition bias introduces a
spurious log2-fold change of no more than a quantity for a non-DA label.
In other words, we interpret this as the maximum log-fold change in the
total number of cells given by DA in other labels. On the other hand,
when choosing , we should not consider fold-differences in the totals
due to differences in capture efficiency or the size of the original
cell population are not attributable to composition bias. We then
mitigate the effect of composition biases by testing each label for
changes in abundance beyond .</p>
<div class="sourceCode" id="cb45"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">res.lfc</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/edgeR/man/glmTreat.html" class="external-link">glmTreat</a></span><span class="op">(</span><span class="va">fit.ab</span>, coef<span class="op">=</span><span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/nrow.html" class="external-link">ncol</a></span><span class="op">(</span><span class="va">design</span><span class="op">)</span>, lfc<span class="op">=</span><span class="fl">1</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/summary.html" class="external-link">summary</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/limma/man/decideTests.html" class="external-link">decideTests</a></span><span class="op">(</span><span class="va">res.lfc</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">##        factor(tomato)TRUE</span></span>
<span><span class="co">## Down                    2</span></span>
<span><span class="co">## NotSig                 32</span></span>
<span><span class="co">## Up                      0</span></span></code></pre>
<div class="sourceCode" id="cb47"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/edgeR/man/topTags.html" class="external-link">topTags</a></span><span class="op">(</span><span class="va">res.lfc</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## Coefficient:  factor(tomato)TRUE </span></span>
<span><span class="co">##                          logFC unshrunk.logFC   logCPM       PValue</span></span>
<span><span class="co">## ExE ectoderm        -5.4835385     -5.9123463 13.06465 4.691927e-06</span></span>
<span><span class="co">## Parietal endoderm   -6.5773855    -27.4596643 12.30091 8.041229e-05</span></span>
<span><span class="co">## ExE endoderm        -3.9301490    -23.9290916 10.76159 5.050518e-02</span></span>
<span><span class="co">## Mesenchyme           1.1597987      1.1610214 16.35239 1.509665e-01</span></span>
<span><span class="co">## Endothelium          1.0396400      1.0450849 14.14422 2.268038e-01</span></span>
<span><span class="co">## Caudal neurectoderm -1.4805723     -1.6367540 11.09613 2.934582e-01</span></span>
<span><span class="co">## Cardiomyocytes       0.8713101      0.8740177 14.96579 3.377083e-01</span></span>
<span><span class="co">## Neural crest        -0.8238482     -0.8264411 14.83184 3.713878e-01</span></span>
<span><span class="co">## Allantois            0.8033893      0.8048744 15.54528 3.952001e-01</span></span>
<span><span class="co">## Def. endoderm        0.7098640      0.7228811 12.50001 4.365468e-01</span></span>
<span><span class="co">##                              FDR</span></span>
<span><span class="co">## ExE ectoderm        0.0001595255</span></span>
<span><span class="co">## Parietal endoderm   0.0013670090</span></span>
<span><span class="co">## ExE endoderm        0.5723920586</span></span>
<span><span class="co">## Mesenchyme          0.9906682053</span></span>
<span><span class="co">## Endothelium         0.9906682053</span></span>
<span><span class="co">## Caudal neurectoderm 0.9906682053</span></span>
<span><span class="co">## Cardiomyocytes      0.9906682053</span></span>
<span><span class="co">## Neural crest        0.9906682053</span></span>
<span><span class="co">## Allantois           0.9906682053</span></span>
<span><span class="co">## Def. endoderm       0.9906682053</span></span></code></pre>
<p>Addionally, the choice of can be guided by other external
experimental data, like a previous or a pilot experiment.</p>
</div>
</div>
<div class="section level2">
<h2 id="session-info">Session Info<a class="anchor" aria-label="anchor" href="#session-info"></a>
</h2>
<div class="sourceCode" id="cb49"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/sessionInfo.html" class="external-link">sessionInfo</a></span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## R version 4.3.0 (2023-04-21)</span></span>
<span><span class="co">## Platform: x86_64-pc-linux-gnu (64-bit)</span></span>
<span><span class="co">## Running under: Ubuntu 22.04.2 LTS</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## Matrix products: default</span></span>
<span><span class="co">## BLAS:   /usr/lib/x86_64-linux-gnu/openblas-pthread/libblas.so.3 </span></span>
<span><span class="co">## LAPACK: /usr/lib/x86_64-linux-gnu/openblas-pthread/libopenblasp-r0.3.20.so;  LAPACK version 3.10.0</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## locale:</span></span>
<span><span class="co">##  [1] LC_CTYPE=en_US.UTF-8       LC_NUMERIC=C              </span></span>
<span><span class="co">##  [3] LC_TIME=en_US.UTF-8        LC_COLLATE=en_US.UTF-8    </span></span>
<span><span class="co">##  [5] LC_MONETARY=en_US.UTF-8    LC_MESSAGES=en_US.UTF-8   </span></span>
<span><span class="co">##  [7] LC_PAPER=en_US.UTF-8       LC_NAME=C                 </span></span>
<span><span class="co">##  [9] LC_ADDRESS=C               LC_TELEPHONE=C            </span></span>
<span><span class="co">## [11] LC_MEASUREMENT=en_US.UTF-8 LC_IDENTIFICATION=C       </span></span>
<span><span class="co">## </span></span>
<span><span class="co">## time zone: Etc/UTC</span></span>
<span><span class="co">## tzcode source: system (glibc)</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## attached base packages:</span></span>
<span><span class="co">## [1] stats4    stats     graphics  grDevices utils     datasets  methods  </span></span>
<span><span class="co">## [8] base     </span></span>
<span><span class="co">## </span></span>
<span><span class="co">## other attached packages:</span></span>
<span><span class="co">##  [1] edgeR_3.42.4                 limma_3.56.2                </span></span>
<span><span class="co">##  [3] batchelor_1.16.0             scran_1.28.1                </span></span>
<span><span class="co">##  [5] scater_1.28.0                ggplot2_3.4.2               </span></span>
<span><span class="co">##  [7] scuttle_1.10.1               MouseGastrulationData_1.14.0</span></span>
<span><span class="co">##  [9] SpatialExperiment_1.10.0     SingleCellExperiment_1.22.0 </span></span>
<span><span class="co">## [11] SummarizedExperiment_1.30.2  Biobase_2.60.0              </span></span>
<span><span class="co">## [13] GenomicRanges_1.52.0         GenomeInfoDb_1.36.1         </span></span>
<span><span class="co">## [15] IRanges_2.34.1               S4Vectors_0.38.1            </span></span>
<span><span class="co">## [17] BiocGenerics_0.46.0          MatrixGenerics_1.12.2       </span></span>
<span><span class="co">## [19] matrixStats_1.0.0            BiocStyle_2.28.0            </span></span>
<span><span class="co">## </span></span>
<span><span class="co">## loaded via a namespace (and not attached):</span></span>
<span><span class="co">##   [1] jsonlite_1.8.7                magrittr_2.0.3               </span></span>
<span><span class="co">##   [3] ggbeeswarm_0.7.2              magick_2.7.4                 </span></span>
<span><span class="co">##   [5] farver_2.1.1                  rmarkdown_2.23               </span></span>
<span><span class="co">##   [7] fs_1.6.2                      zlibbioc_1.46.0              </span></span>
<span><span class="co">##   [9] ragg_1.2.5                    vctrs_0.6.3                  </span></span>
<span><span class="co">##  [11] memoise_2.0.1                 DelayedMatrixStats_1.22.1    </span></span>
<span><span class="co">##  [13] RCurl_1.98-1.12               htmltools_0.5.5              </span></span>
<span><span class="co">##  [15] S4Arrays_1.0.4                AnnotationHub_3.8.0          </span></span>
<span><span class="co">##  [17] curl_5.0.1                    BiocNeighbors_1.18.0         </span></span>
<span><span class="co">##  [19] Rhdf5lib_1.22.0               rhdf5_2.44.0                 </span></span>
<span><span class="co">##  [21] sass_0.4.6                    bslib_0.5.0                  </span></span>
<span><span class="co">##  [23] desc_1.4.2                    cachem_1.0.8                 </span></span>
<span><span class="co">##  [25] ResidualMatrix_1.10.0         igraph_1.5.0                 </span></span>
<span><span class="co">##  [27] mime_0.12                     lifecycle_1.0.3              </span></span>
<span><span class="co">##  [29] pkgconfig_2.0.3               rsvd_1.0.5                   </span></span>
<span><span class="co">##  [31] Matrix_1.5-4.1                R6_2.5.1                     </span></span>
<span><span class="co">##  [33] fastmap_1.1.1                 GenomeInfoDbData_1.2.10      </span></span>
<span><span class="co">##  [35] shiny_1.7.4                   digest_0.6.32                </span></span>
<span><span class="co">##  [37] colorspace_2.1-0              AnnotationDbi_1.62.2         </span></span>
<span><span class="co">##  [39] rprojroot_2.0.3               dqrng_0.3.0                  </span></span>
<span><span class="co">##  [41] irlba_2.3.5.1                 ExperimentHub_2.8.0          </span></span>
<span><span class="co">##  [43] textshaping_0.3.6             RSQLite_2.3.1                </span></span>
<span><span class="co">##  [45] beachmat_2.16.0               labeling_0.4.2               </span></span>
<span><span class="co">##  [47] filelock_1.0.2                fansi_1.0.4                  </span></span>
<span><span class="co">##  [49] httr_1.4.6                    compiler_4.3.0               </span></span>
<span><span class="co">##  [51] bit64_4.0.5                   withr_2.5.0                  </span></span>
<span><span class="co">##  [53] BiocParallel_1.34.2           viridis_0.6.3                </span></span>
<span><span class="co">##  [55] DBI_1.1.3                     highr_0.10                   </span></span>
<span><span class="co">##  [57] HDF5Array_1.28.1              R.utils_2.12.2               </span></span>
<span><span class="co">##  [59] rappdirs_0.3.3                DelayedArray_0.26.6          </span></span>
<span><span class="co">##  [61] bluster_1.10.0                rjson_0.2.21                 </span></span>
<span><span class="co">##  [63] tools_4.3.0                   vipor_0.4.5                  </span></span>
<span><span class="co">##  [65] beeswarm_0.4.0                interactiveDisplayBase_1.38.0</span></span>
<span><span class="co">##  [67] httpuv_1.6.11                 R.oo_1.25.0                  </span></span>
<span><span class="co">##  [69] glue_1.6.2                    rhdf5filters_1.12.1          </span></span>
<span><span class="co">##  [71] promises_1.2.0.1              grid_4.3.0                   </span></span>
<span><span class="co">##  [73] Rtsne_0.16                    cluster_2.1.4                </span></span>
<span><span class="co">##  [75] generics_0.1.3                gtable_0.3.3                 </span></span>
<span><span class="co">##  [77] R.methodsS3_1.8.2             metapod_1.8.0                </span></span>
<span><span class="co">##  [79] BiocSingular_1.16.0           ScaledMatrix_1.8.1           </span></span>
<span><span class="co">##  [81] utf8_1.2.3                    XVector_0.40.0               </span></span>
<span><span class="co">##  [83] ggrepel_0.9.3                 BiocVersion_3.17.1           </span></span>
<span><span class="co">##  [85] pillar_1.9.0                  stringr_1.5.0                </span></span>
<span><span class="co">##  [87] BumpyMatrix_1.8.0             later_1.3.1                  </span></span>
<span><span class="co">##  [89] splines_4.3.0                 dplyr_1.1.2                  </span></span>
<span><span class="co">##  [91] BiocFileCache_2.8.0           lattice_0.21-8               </span></span>
<span><span class="co">##  [93] bit_4.0.5                     tidyselect_1.2.0             </span></span>
<span><span class="co">##  [95] locfit_1.5-9.8                Biostrings_2.68.1            </span></span>
<span><span class="co">##  [97] knitr_1.43                    gridExtra_2.3                </span></span>
<span><span class="co">##  [99] xfun_0.39                     statmod_1.5.0                </span></span>
<span><span class="co">## [101] DropletUtils_1.20.0           stringi_1.7.12               </span></span>
<span><span class="co">## [103] yaml_2.3.7                    evaluate_0.21                </span></span>
<span><span class="co">## [105] codetools_0.2-19              tibble_3.2.1                 </span></span>
<span><span class="co">## [107] BiocManager_1.30.21           cli_3.6.1                    </span></span>
<span><span class="co">## [109] xtable_1.8-4                  systemfonts_1.0.4            </span></span>
<span><span class="co">## [111] munsell_0.5.0                 jquerylib_0.1.4              </span></span>
<span><span class="co">## [113] Rcpp_1.0.10                   dbplyr_2.3.2                 </span></span>
<span><span class="co">## [115] png_0.1-8                     parallel_4.3.0               </span></span>
<span><span class="co">## [117] ellipsis_0.3.2                pkgdown_2.0.7                </span></span>
<span><span class="co">## [119] blob_1.2.4                    sparseMatrixStats_1.12.2     </span></span>
<span><span class="co">## [121] bitops_1.0-7                  viridisLite_0.4.2            </span></span>
<span><span class="co">## [123] scales_1.2.1                  purrr_1.0.1                  </span></span>
<span><span class="co">## [125] crayon_1.5.2                  rlang_1.1.1                  </span></span>
<span><span class="co">## [127] formatR_1.14                  cowplot_1.1.1                </span></span>
<span><span class="co">## [129] KEGGREST_1.40.0</span></span></code></pre>
</div>
<div class="section level2">
<h2 id="further-reading">Further Reading<a class="anchor" aria-label="anchor" href="#further-reading"></a>
</h2>
<ul>
<li>OSCA book, Multi-sample analysis, <a href="https://bioconductor.org/books/release/OSCA.multisample" class="external-link">Chapters
1, 4, and 6</a>
</li>
</ul>
</div>
<div class="section level2">
<h2 id="exercises">Exercises<a class="anchor" aria-label="anchor" href="#exercises"></a>
</h2>
<ul>
<li><p>Test differential expressed genes with just 2 replicates per
condition and look into the changes in the results and possible emerging
issues.</p></li>
<li><p>Test differential expressed genes computed with
<code>muscat</code> package and check for differences in the
results.</p></li>
</ul>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">

        <nav id="toc" data-toggle="toc"><h2 data-toc-skip>Contents</h2>
    </nav>
</div>

</div>



      <footer><div class="copyright">
  <p></p>
<p>Developed by Davide Risso, Ludwig Geistlinger.</p>
</div>

<div class="pkgdown">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.7.</p>
</div>

      </footer>
</div>

  


  

  </body>
</html>
