<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>Multi-sample analyses • ISMB.OSCA</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link href="https://cdnjs.cloudflare.com/ajax/libs/bootswatch/3.4.0/flatly/bootstrap.min.css" rel="stylesheet" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="Multi-sample analyses">
<meta property="og:description" content="ISMB.OSCA">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">
    

    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">ISMB.OSCA</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">0.99.0</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Vignettes
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../articles/intro_bioc_sce.html">Introduction to Bioconductor</a>
    </li>
    <li>
      <a href="../articles/eda_qc.html">Exploratory data analysis</a>
    </li>
    <li>
      <a href="../articles/cell_type_annotation.html">Cell type annotation</a>
    </li>
    <li>
      <a href="../articles/multi_sample.html">Multi-sample analyses</a>
    </li>
    <li>
      <a href="../articles/large_data.html">Working with large data</a>
    </li>
    <li>
      <a href="../articles/hca.html">Accessing data from the Human Cell Atlas</a>
    </li>
  </ul>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://github.com/Bioconductor/ISMB.OSCA" class="external-link">
    <span class="fa fa-github"></span>
     
  </a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1 data-toc-skip>Multi-sample analyses</h1>
            
      
      <small class="dont-index">Source: <a href="https://github.com/Bioconductor/ISMB.OSCA/blob/HEAD/vignettes/multi_sample.Rmd" class="external-link"><code>vignettes/multi_sample.Rmd</code></a></small>
      <div class="hidden name"><code>multi_sample.Rmd</code></div>

    </div>

    
    
<div class="section level2">
<h2 id="setup-and-data-exploration">Setup and data exploration<a class="anchor" aria-label="anchor" href="#setup-and-data-exploration"></a>
</h2>
<p>As said, we will use the the data from the Tal1 chimera
experiment:</p>
<ul>
<li>Sample 5: E8.5 injected cells (tomato positive), pool 3</li>
<li>Sample 6: E8.5 host cells (tomato negative), pool 3</li>
<li>Sample 7: E8.5 injected cells (tomato positive), pool 4</li>
<li>Sample 8: E8.5 host cells (tomato negative), pool 4</li>
<li>Sample 9: E8.5 injected cells (tomato positive), pool 5</li>
<li>Sample 10: E8.5 host cells (tomato negative), pool 5</li>
</ul>
<p>Note that this is a paired design in which for each biological
replicate (pool 3, 4, and 5), we have both host and injected cells.</p>
<p>We start by loading the data and doing a quick exploratory analysis,
essentially applying the normalization and visualization techniques that
we have seen in the previous lectures to all samples.</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/MarioniLab/MouseGastrulationData" class="external-link">MouseGastrulationData</a></span><span class="op">)</span></span>
<span><span class="va">sce</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/MouseGastrulationData/man/WTChimeraData.html" class="external-link">WTChimeraData</a></span><span class="op">(</span>samples<span class="op">=</span><span class="fl">5</span><span class="op">:</span><span class="fl">10</span>, type <span class="op">=</span> <span class="st">"processed"</span><span class="op">)</span></span>
<span><span class="va">sce</span></span></code></pre></div>
<pre><code><span><span class="co">## class: SingleCellExperiment </span></span>
<span><span class="co">## dim: 29453 20935 </span></span>
<span><span class="co">## metadata(0):</span></span>
<span><span class="co">## assays(1): counts</span></span>
<span><span class="co">## rownames(29453): ENSMUSG00000051951 ENSMUSG00000089699 ...</span></span>
<span><span class="co">##   ENSMUSG00000095742 tomato-td</span></span>
<span><span class="co">## rowData names(2): ENSEMBL SYMBOL</span></span>
<span><span class="co">## colnames(20935): cell_9769 cell_9770 ... cell_30702 cell_30703</span></span>
<span><span class="co">## colData names(11): cell barcode ... doub.density sizeFactor</span></span>
<span><span class="co">## reducedDimNames(2): pca.corrected.E7.5 pca.corrected.E8.5</span></span>
<span><span class="co">## mainExpName: NULL</span></span>
<span><span class="co">## altExpNames(0):</span></span></code></pre>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">colData</a></span><span class="op">(</span><span class="va">sce</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## DataFrame with 20935 rows and 11 columns</span></span>
<span><span class="co">##                   cell          barcode    sample       stage    tomato</span></span>
<span><span class="co">##            &lt;character&gt;      &lt;character&gt; &lt;integer&gt; &lt;character&gt; &lt;logical&gt;</span></span>
<span><span class="co">## cell_9769    cell_9769 AAACCTGAGACTGTAA         5        E8.5      TRUE</span></span>
<span><span class="co">## cell_9770    cell_9770 AAACCTGAGATGCCTT         5        E8.5      TRUE</span></span>
<span><span class="co">## cell_9771    cell_9771 AAACCTGAGCAGCCTC         5        E8.5      TRUE</span></span>
<span><span class="co">## cell_9772    cell_9772 AAACCTGCATACTCTT         5        E8.5      TRUE</span></span>
<span><span class="co">## cell_9773    cell_9773 AAACGGGTCAACACCA         5        E8.5      TRUE</span></span>
<span><span class="co">## ...                ...              ...       ...         ...       ...</span></span>
<span><span class="co">## cell_30699  cell_30699 TTTGTCACAGCTCGCA        10        E8.5     FALSE</span></span>
<span><span class="co">## cell_30700  cell_30700 TTTGTCAGTCTAGTCA        10        E8.5     FALSE</span></span>
<span><span class="co">## cell_30701  cell_30701 TTTGTCATCATCGGAT        10        E8.5     FALSE</span></span>
<span><span class="co">## cell_30702  cell_30702 TTTGTCATCATTATCC        10        E8.5     FALSE</span></span>
<span><span class="co">## cell_30703  cell_30703 TTTGTCATCCCATTTA        10        E8.5     FALSE</span></span>
<span><span class="co">##                 pool stage.mapped        celltype.mapped closest.cell</span></span>
<span><span class="co">##            &lt;integer&gt;  &lt;character&gt;            &lt;character&gt;  &lt;character&gt;</span></span>
<span><span class="co">## cell_9769          3        E8.25             Mesenchyme   cell_24159</span></span>
<span><span class="co">## cell_9770          3         E8.5            Endothelium   cell_96660</span></span>
<span><span class="co">## cell_9771          3         E8.5              Allantois  cell_134982</span></span>
<span><span class="co">## cell_9772          3         E8.5             Erythroid3  cell_133892</span></span>
<span><span class="co">## cell_9773          3        E8.25             Erythroid1   cell_76296</span></span>
<span><span class="co">## ...              ...          ...                    ...          ...</span></span>
<span><span class="co">## cell_30699         5         E8.5             Erythroid3   cell_38810</span></span>
<span><span class="co">## cell_30700         5         E8.5       Surface ectoderm   cell_38588</span></span>
<span><span class="co">## cell_30701         5        E8.25 Forebrain/Midbrain/H..   cell_66082</span></span>
<span><span class="co">## cell_30702         5         E8.5             Erythroid3  cell_138114</span></span>
<span><span class="co">## cell_30703         5         E8.0                Doublet   cell_92644</span></span>
<span><span class="co">##            doub.density sizeFactor</span></span>
<span><span class="co">##               &lt;numeric&gt;  &lt;numeric&gt;</span></span>
<span><span class="co">## cell_9769    0.02985045    1.41243</span></span>
<span><span class="co">## cell_9770    0.00172753    1.22757</span></span>
<span><span class="co">## cell_9771    0.01338013    1.15439</span></span>
<span><span class="co">## cell_9772    0.00218402    1.28676</span></span>
<span><span class="co">## cell_9773    0.00211723    1.78719</span></span>
<span><span class="co">## ...                 ...        ...</span></span>
<span><span class="co">## cell_30699   0.00146287   0.389311</span></span>
<span><span class="co">## cell_30700   0.00374155   0.588784</span></span>
<span><span class="co">## cell_30701   0.05651258   0.624455</span></span>
<span><span class="co">## cell_30702   0.00108837   0.550807</span></span>
<span><span class="co">## cell_30703   0.82369305   1.184919</span></span></code></pre>
<p>We now normalize the data and visualize them in a tSNE plot.</p>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="http://bioconductor.org/packages/scater/" class="external-link">scater</a></span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## Loading required package: scuttle</span></span></code></pre>
<pre><code><span><span class="co">## Loading required package: ggplot2</span></span></code></pre>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://ggplot2.tidyverse.org" class="external-link">ggplot2</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">scran</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># remove doublets</span></span>
<span><span class="va">drop</span> <span class="op">&lt;-</span> <span class="va">sce</span><span class="op">$</span><span class="va">celltype.mapped</span> <span class="op"><a href="https://rdrr.io/r/base/match.html" class="external-link">%in%</a></span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"stripped"</span>, <span class="st">"Doublet"</span><span class="op">)</span></span>
<span><span class="va">sce</span> <span class="op">&lt;-</span> <span class="va">sce</span><span class="op">[</span>,<span class="op">!</span><span class="va">drop</span><span class="op">]</span></span>
<span></span>
<span><span class="co"># normalization</span></span>
<span><span class="va">sce</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/scuttle/man/logNormCounts.html" class="external-link">logNormCounts</a></span><span class="op">(</span><span class="va">sce</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># identify highly variable genes</span></span>
<span><span class="va">dec</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/scran/man/modelGeneVar.html" class="external-link">modelGeneVar</a></span><span class="op">(</span><span class="va">sce</span>, block<span class="op">=</span><span class="va">sce</span><span class="op">$</span><span class="va">sample</span><span class="op">)</span></span>
<span><span class="va">chosen.hvgs</span> <span class="op">&lt;-</span> <span class="va">dec</span><span class="op">$</span><span class="va">bio</span> <span class="op">&gt;</span> <span class="fl">0</span></span>
<span></span>
<span><span class="co"># dimensionality reduction</span></span>
<span><span class="va">sce</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/BiocSingular/man/runPCA.html" class="external-link">runPCA</a></span><span class="op">(</span><span class="va">sce</span>, subset_row <span class="op">=</span> <span class="va">chosen.hvgs</span>, ntop <span class="op">=</span> <span class="fl">1000</span><span class="op">)</span></span>
<span><span class="va">sce</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/scater/man/runTSNE.html" class="external-link">runTSNE</a></span><span class="op">(</span><span class="va">sce</span>, dimred <span class="op">=</span> <span class="st">"PCA"</span><span class="op">)</span></span>
<span></span>
<span><span class="va">sce</span><span class="op">$</span><span class="va">sample</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/factor.html" class="external-link">as.factor</a></span><span class="op">(</span><span class="va">sce</span><span class="op">$</span><span class="va">sample</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/scater/man/plot_reddim.html" class="external-link">plotTSNE</a></span><span class="op">(</span><span class="va">sce</span>, colour_by <span class="op">=</span> <span class="st">"sample"</span><span class="op">)</span></span></code></pre></div>
<p><img src="multi_sample_files/figure-html/unnamed-chunk-1-1.png" width="700"></p>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/scater/man/plot_reddim.html" class="external-link">plotTSNE</a></span><span class="op">(</span><span class="va">sce</span>, colour_by <span class="op">=</span> <span class="st">"celltype.mapped"</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_colour_discrete.html" class="external-link">scale_color_discrete</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/theme.html" class="external-link">theme</a></span><span class="op">(</span>legend.position <span class="op">=</span> <span class="st">"bottom"</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## Scale for <span style="color: #00BB00;">colour</span> is already present.</span></span>
<span><span class="co">## Adding another scale for <span style="color: #00BB00;">colour</span>, which will replace the existing scale.</span></span></code></pre>
<p><img src="multi_sample_files/figure-html/unnamed-chunk-1-2.png" width="700"></p>
<p>There are evident sample effects. Depending on the analysis that you
want to do you may want to remove or retain the sample effect. For
instance, if the goal is to identify cell types with a clustering
method, one may want to remove the sample effects with “batch effect”
correction methods.</p>
<p>For now, let’s assume that we want to remove this effect.</p>
</div>
<div class="section level2">
<h2 id="correcting-batch-effects">Correcting batch effects<a class="anchor" aria-label="anchor" href="#correcting-batch-effects"></a>
</h2>
<div class="sourceCode" id="cb11"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">## This needs work</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">batchelor</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/Random.html" class="external-link">set.seed</a></span><span class="op">(</span><span class="fl">10102</span><span class="op">)</span></span>
<span><span class="va">merged</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/batchelor/man/correctExperiments.html" class="external-link">correctExperiments</a></span><span class="op">(</span><span class="va">sce</span>, </span>
<span>    batch<span class="op">=</span><span class="va">sce</span><span class="op">$</span><span class="va">sample</span>, </span>
<span>    subset.row<span class="op">=</span><span class="va">chosen.hvgs</span>,</span>
<span>    PARAM<span class="op">=</span><span class="fu"><a href="https://rdrr.io/pkg/batchelor/man/BatchelorParam.html" class="external-link">FastMnnParam</a></span><span class="op">(</span></span>
<span>        merge.order<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span></span>
<span>            <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span><span class="fl">1</span>,<span class="fl">3</span>,<span class="fl">5</span><span class="op">)</span>, <span class="co"># WT (3 replicates)</span></span>
<span>            <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span><span class="fl">2</span>,<span class="fl">4</span>,<span class="fl">6</span><span class="op">)</span>  <span class="co"># td-Tomato (3 replicates)</span></span>
<span>        <span class="op">)</span></span>
<span>    <span class="op">)</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="va">merged</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/scater/man/runTSNE.html" class="external-link">runTSNE</a></span><span class="op">(</span><span class="va">merged</span>, dimred<span class="op">=</span><span class="st">"corrected"</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/scater/man/plot_reddim.html" class="external-link">plotTSNE</a></span><span class="op">(</span><span class="va">merged</span>, colour_by<span class="op">=</span><span class="st">"batch"</span><span class="op">)</span></span></code></pre></div>
<p><img src="multi_sample_files/figure-html/unnamed-chunk-2-1.png" width="700"></p>
<div class="sourceCode" id="cb12"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># do we want to include the diagnostics about the batch correction?</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="differential-expression">Differential Expression<a class="anchor" aria-label="anchor" href="#differential-expression"></a>
</h2>
<p>Section 4 of OSCA multisample.</p>
<p>In order to perform a Differential Expression Analysis, we need to
identify clusters of cells across samples/conditions (depending on the
experimental design and the final aim of the experiment) to proceed with
it</p>
<div class="section level3">
<h3 id="clustering-identification">clustering identification<a class="anchor" aria-label="anchor" href="#clustering-identification"></a>
</h3>
<div class="sourceCode" id="cb13"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">g</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/scran/man/buildSNNGraph.html" class="external-link">buildSNNGraph</a></span><span class="op">(</span><span class="va">merged</span>, use.dimred<span class="op">=</span><span class="st">"corrected"</span><span class="op">)</span></span>
<span><span class="va">clusters</span> <span class="op">&lt;-</span> <span class="fu">igraph</span><span class="fu">::</span><span class="fu"><a href="https://r.igraph.org/reference/cluster_louvain.html" class="external-link">cluster_louvain</a></span><span class="op">(</span><span class="va">g</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/SingleCellExperiment/man/colLabels.html" class="external-link">colLabels</a></span><span class="op">(</span><span class="va">merged</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/factor.html" class="external-link">factor</a></span><span class="op">(</span><span class="va">clusters</span><span class="op">$</span><span class="va">membership</span><span class="op">)</span></span>
<span><span class="va">merged</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/scater/man/runTSNE.html" class="external-link">runTSNE</a></span><span class="op">(</span><span class="va">merged</span>, dimred<span class="op">=</span><span class="st">"corrected"</span>, external_neighbors<span class="op">=</span><span class="cn">TRUE</span><span class="op">)</span></span>
<span><span class="va">merged</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/scater/man/runUMAP.html" class="external-link">runUMAP</a></span><span class="op">(</span><span class="va">merged</span>, dimred<span class="op">=</span><span class="st">"corrected"</span>, external_neighbors<span class="op">=</span><span class="cn">TRUE</span><span class="op">)</span></span></code></pre></div>
<div class="sourceCode" id="cb14"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># table(colLabels(merged), merged$tomato)</span></span>
<span><span class="fu">gridExtra</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/gridExtra/man/arrangeGrob.html" class="external-link">grid.arrange</a></span><span class="op">(</span></span>
<span>    <span class="fu"><a href="https://rdrr.io/pkg/scater/man/plot_reddim.html" class="external-link">plotTSNE</a></span><span class="op">(</span><span class="va">merged</span>, colour_by<span class="op">=</span><span class="st">"tomato"</span>, text_by<span class="op">=</span><span class="st">"label"</span><span class="op">)</span>,</span>
<span>    <span class="fu"><a href="https://rdrr.io/pkg/scater/man/plot_reddim.html" class="external-link">plotTSNE</a></span><span class="op">(</span><span class="va">merged</span>, colour_by<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/data.frame.html" class="external-link">data.frame</a></span><span class="op">(</span>pool<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/factor.html" class="external-link">factor</a></span><span class="op">(</span><span class="va">merged</span><span class="op">$</span><span class="va">pool</span><span class="op">)</span><span class="op">)</span><span class="op">)</span>,</span>
<span>    ncol<span class="op">=</span><span class="fl">2</span></span>
<span><span class="op">)</span></span></code></pre></div>
<p><img src="multi_sample_files/figure-html/unnamed-chunk-4-1.png" width="700"></p>
<p>As previously mentioned, to assign biological meaning to the found
clusters, we need to detect and assign gene-markers and provide cell
labels. For lack of time, we will skip this time and knowledge demanding
process and we use the annotation included with the dataset. This
annotation has been previously performed using a cell type atlas by
Pijuan-Sala et al. (2019), using it as a reference for our dataset. From
the next figure we can see that there are quite a few similarities
between the clusters and the assigned labels, but in some cases we can
notice that a cluster can contain multiple cell types, highlighting the
difficulty to identify cell types by using cell markers and
clusters.</p>
<div class="sourceCode" id="cb15"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">bluster</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/bluster/man/pairwiseRand.html" class="external-link">pairwiseRand</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/SingleCellExperiment/man/colLabels.html" class="external-link">colLabels</a></span><span class="op">(</span><span class="va">merged</span><span class="op">)</span>, <span class="va">merged</span><span class="op">$</span><span class="va">celltype.mapped</span>, <span class="st">"index"</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## [1] 0.5346738</span></span></code></pre>
<div class="sourceCode" id="cb17"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">by.label</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/table.html" class="external-link">table</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/SingleCellExperiment/man/colLabels.html" class="external-link">colLabels</a></span><span class="op">(</span><span class="va">merged</span><span class="op">)</span>, <span class="va">merged</span><span class="op">$</span><span class="va">celltype.mapped</span><span class="op">)</span></span>
<span><span class="fu">pheatmap</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/pheatmap/man/pheatmap.html" class="external-link">pheatmap</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/Log.html" class="external-link">log2</a></span><span class="op">(</span><span class="va">by.label</span><span class="op">+</span><span class="fl">1</span><span class="op">)</span>, color<span class="op">=</span><span class="fu">viridis</span><span class="fu">::</span><span class="fu"><a href="https://sjmgarnier.github.io/viridisLite/reference/viridis.html" class="external-link">viridis</a></span><span class="op">(</span><span class="fl">101</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<p><img src="multi_sample_files/figure-html/unnamed-chunk-5-1.png" width="700"></p>
</div>
<div class="section level3">
<h3 id="pseudo-bulk-samples">Pseudo-bulk samples<a class="anchor" aria-label="anchor" href="#pseudo-bulk-samples"></a>
</h3>
<div class="sourceCode" id="cb18"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Using 'label' and 'sample' as our two factors; each column of the output</span></span>
<span><span class="co"># corresponds to one unique combination of these two factors.</span></span>
<span><span class="va">summed</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/scuttle/man/aggregateAcrossCells.html" class="external-link">aggregateAcrossCells</a></span><span class="op">(</span><span class="va">merged</span>, </span>
<span>    id<span class="op">=</span><span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">colData</a></span><span class="op">(</span><span class="va">merged</span><span class="op">)</span><span class="op">[</span>,<span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"celltype.mapped"</span>, <span class="st">"sample"</span><span class="op">)</span><span class="op">]</span><span class="op">)</span></span>
<span><span class="va">summed</span></span></code></pre></div>
<pre><code><span><span class="co">## class: SingleCellExperiment </span></span>
<span><span class="co">## dim: 14699 186 </span></span>
<span><span class="co">## metadata(2): merge.info pca.info</span></span>
<span><span class="co">## assays(1): counts</span></span>
<span><span class="co">## rownames(14699): ENSMUSG00000051951 ENSMUSG00000025900 ...</span></span>
<span><span class="co">##   ENSMUSG00000096730 ENSMUSG00000095742</span></span>
<span><span class="co">## rowData names(3): rotation ENSEMBL SYMBOL</span></span>
<span><span class="co">## colnames: NULL</span></span>
<span><span class="co">## colData names(16): batch cell ... sample ncells</span></span>
<span><span class="co">## reducedDimNames(6): corrected pca.corrected.E7.5 ... TSNE UMAP</span></span>
<span><span class="co">## mainExpName: NULL</span></span>
<span><span class="co">## altExpNames(0):</span></span></code></pre>
</div>
<div class="section level3">
<h3 id="differential-expression-analysis">Differential Expression Analysis<a class="anchor" aria-label="anchor" href="#differential-expression-analysis"></a>
</h3>
<div class="sourceCode" id="cb20"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">label</span> <span class="op">&lt;-</span> <span class="st">"Mesenchyme"</span></span>
<span><span class="va">current</span> <span class="op">&lt;-</span> <span class="va">summed</span><span class="op">[</span>,<span class="va">label</span><span class="op">==</span><span class="va">summed</span><span class="op">$</span><span class="va">celltype.mapped</span><span class="op">]</span></span>
<span></span>
<span><span class="co"># Creating up a DGEList object for use in edgeR:</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://bioinf.wehi.edu.au/edgeR/" class="external-link">edgeR</a></span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## Loading required package: limma</span></span></code></pre>
<pre><code><span><span class="co">## </span></span>
<span><span class="co">## Attaching package: 'limma'</span></span></code></pre>
<pre><code><span><span class="co">## The following object is masked from 'package:scater':</span></span>
<span><span class="co">## </span></span>
<span><span class="co">##     plotMDS</span></span></code></pre>
<pre><code><span><span class="co">## The following object is masked from 'package:BiocGenerics':</span></span>
<span><span class="co">## </span></span>
<span><span class="co">##     plotMA</span></span></code></pre>
<pre><code><span><span class="co">## </span></span>
<span><span class="co">## Attaching package: 'edgeR'</span></span></code></pre>
<pre><code><span><span class="co">## The following object is masked from 'package:SingleCellExperiment':</span></span>
<span><span class="co">## </span></span>
<span><span class="co">##     cpm</span></span></code></pre>
<div class="sourceCode" id="cb27"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">y</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/edgeR/man/DGEList.html" class="external-link">DGEList</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/dge.html" class="external-link">counts</a></span><span class="op">(</span><span class="va">current</span><span class="op">)</span>, samples<span class="op">=</span><span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">colData</a></span><span class="op">(</span><span class="va">current</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">y</span></span></code></pre></div>
<pre><code><span><span class="co">## An object of class "DGEList"</span></span>
<span><span class="co">## $counts</span></span>
<span><span class="co">##                    Sample1 Sample2 Sample3 Sample4 Sample5 Sample6</span></span>
<span><span class="co">## ENSMUSG00000051951       2       0       0       0       3       0</span></span>
<span><span class="co">## ENSMUSG00000025900       0       0       1       0       0       0</span></span>
<span><span class="co">## ENSMUSG00000025902       7       0       3       0      14       9</span></span>
<span><span class="co">## ENSMUSG00000033845    1420     271    1009     379    1578     749</span></span>
<span><span class="co">## ENSMUSG00000002459       3       0       1       1       0       0</span></span>
<span><span class="co">## 14694 more rows ...</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## $samples</span></span>
<span><span class="co">##         group lib.size norm.factors batch cell barcode sample stage tomato pool</span></span>
<span><span class="co">## Sample1     1  4607053            1     5 &lt;NA&gt;    &lt;NA&gt;      5  E8.5   TRUE    3</span></span>
<span><span class="co">## Sample2     1  1064970            1     6 &lt;NA&gt;    &lt;NA&gt;      6  E8.5  FALSE    3</span></span>
<span><span class="co">## Sample3     1  2494010            1     7 &lt;NA&gt;    &lt;NA&gt;      7  E8.5   TRUE    4</span></span>
<span><span class="co">## Sample4     1  1028668            1     8 &lt;NA&gt;    &lt;NA&gt;      8  E8.5  FALSE    4</span></span>
<span><span class="co">## Sample5     1  4290221            1     9 &lt;NA&gt;    &lt;NA&gt;      9  E8.5   TRUE    5</span></span>
<span><span class="co">## Sample6     1  1950840            1    10 &lt;NA&gt;    &lt;NA&gt;     10  E8.5  FALSE    5</span></span>
<span><span class="co">##         stage.mapped celltype.mapped closest.cell doub.density sizeFactor label</span></span>
<span><span class="co">## Sample1         &lt;NA&gt;      Mesenchyme         &lt;NA&gt;           NA         NA  &lt;NA&gt;</span></span>
<span><span class="co">## Sample2         &lt;NA&gt;      Mesenchyme         &lt;NA&gt;           NA         NA  &lt;NA&gt;</span></span>
<span><span class="co">## Sample3         &lt;NA&gt;      Mesenchyme         &lt;NA&gt;           NA         NA  &lt;NA&gt;</span></span>
<span><span class="co">## Sample4         &lt;NA&gt;      Mesenchyme         &lt;NA&gt;           NA         NA  &lt;NA&gt;</span></span>
<span><span class="co">## Sample5         &lt;NA&gt;      Mesenchyme         &lt;NA&gt;           NA         NA  &lt;NA&gt;</span></span>
<span><span class="co">## Sample6         &lt;NA&gt;      Mesenchyme         &lt;NA&gt;           NA         NA  &lt;NA&gt;</span></span>
<span><span class="co">##         celltype.mapped.1 sample.1 ncells</span></span>
<span><span class="co">## Sample1        Mesenchyme        5    286</span></span>
<span><span class="co">## Sample2        Mesenchyme        6     55</span></span>
<span><span class="co">## Sample3        Mesenchyme        7    243</span></span>
<span><span class="co">## Sample4        Mesenchyme        8    134</span></span>
<span><span class="co">## Sample5        Mesenchyme        9    478</span></span>
<span><span class="co">## Sample6        Mesenchyme       10    299</span></span></code></pre>
<p>Discard lowly “sequenced” samples</p>
<div class="sourceCode" id="cb29"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">discarded</span> <span class="op">&lt;-</span> <span class="va">current</span><span class="op">$</span><span class="va">ncells</span> <span class="op">&lt;</span> <span class="fl">10</span></span>
<span><span class="va">y</span> <span class="op">&lt;-</span> <span class="va">y</span><span class="op">[</span>,<span class="op">!</span><span class="va">discarded</span><span class="op">]</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/summary.html" class="external-link">summary</a></span><span class="op">(</span><span class="va">discarded</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">##    Mode   FALSE </span></span>
<span><span class="co">## logical       6</span></span></code></pre>
<p>Discard lowly expressed genes</p>
<div class="sourceCode" id="cb31"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">keep</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/edgeR/man/filterByExpr.html" class="external-link">filterByExpr</a></span><span class="op">(</span><span class="va">y</span>, group<span class="op">=</span><span class="va">current</span><span class="op">$</span><span class="va">tomato</span><span class="op">)</span></span>
<span><span class="va">y</span> <span class="op">&lt;-</span> <span class="va">y</span><span class="op">[</span><span class="va">keep</span>,<span class="op">]</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/summary.html" class="external-link">summary</a></span><span class="op">(</span><span class="va">keep</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">##    Mode   FALSE    TRUE </span></span>
<span><span class="co">## logical    9011    5688</span></span></code></pre>
<p>Normalization factors by TMM</p>
<div class="sourceCode" id="cb33"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">y</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/edgeR/man/calcNormFactors.html" class="external-link">calcNormFactors</a></span><span class="op">(</span><span class="va">y</span><span class="op">)</span></span>
<span><span class="va">y</span><span class="op">$</span><span class="va">samples</span></span></code></pre></div>
<pre><code><span><span class="co">##         group lib.size norm.factors batch cell barcode sample stage tomato pool</span></span>
<span><span class="co">## Sample1     1  4607053    1.0683392     5 &lt;NA&gt;    &lt;NA&gt;      5  E8.5   TRUE    3</span></span>
<span><span class="co">## Sample2     1  1064970    1.0487418     6 &lt;NA&gt;    &lt;NA&gt;      6  E8.5  FALSE    3</span></span>
<span><span class="co">## Sample3     1  2494010    0.9582296     7 &lt;NA&gt;    &lt;NA&gt;      7  E8.5   TRUE    4</span></span>
<span><span class="co">## Sample4     1  1028668    0.9774156     8 &lt;NA&gt;    &lt;NA&gt;      8  E8.5  FALSE    4</span></span>
<span><span class="co">## Sample5     1  4290221    0.9707300     9 &lt;NA&gt;    &lt;NA&gt;      9  E8.5   TRUE    5</span></span>
<span><span class="co">## Sample6     1  1950840    0.9816914    10 &lt;NA&gt;    &lt;NA&gt;     10  E8.5  FALSE    5</span></span>
<span><span class="co">##         stage.mapped celltype.mapped closest.cell doub.density sizeFactor label</span></span>
<span><span class="co">## Sample1         &lt;NA&gt;      Mesenchyme         &lt;NA&gt;           NA         NA  &lt;NA&gt;</span></span>
<span><span class="co">## Sample2         &lt;NA&gt;      Mesenchyme         &lt;NA&gt;           NA         NA  &lt;NA&gt;</span></span>
<span><span class="co">## Sample3         &lt;NA&gt;      Mesenchyme         &lt;NA&gt;           NA         NA  &lt;NA&gt;</span></span>
<span><span class="co">## Sample4         &lt;NA&gt;      Mesenchyme         &lt;NA&gt;           NA         NA  &lt;NA&gt;</span></span>
<span><span class="co">## Sample5         &lt;NA&gt;      Mesenchyme         &lt;NA&gt;           NA         NA  &lt;NA&gt;</span></span>
<span><span class="co">## Sample6         &lt;NA&gt;      Mesenchyme         &lt;NA&gt;           NA         NA  &lt;NA&gt;</span></span>
<span><span class="co">##         celltype.mapped.1 sample.1 ncells</span></span>
<span><span class="co">## Sample1        Mesenchyme        5    286</span></span>
<span><span class="co">## Sample2        Mesenchyme        6     55</span></span>
<span><span class="co">## Sample3        Mesenchyme        7    243</span></span>
<span><span class="co">## Sample4        Mesenchyme        8    134</span></span>
<span><span class="co">## Sample5        Mesenchyme        9    478</span></span>
<span><span class="co">## Sample6        Mesenchyme       10    299</span></span></code></pre>
<p>MD plot</p>
<div class="sourceCode" id="cb35"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/graphics/par.html" class="external-link">par</a></span><span class="op">(</span>mfrow<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">2</span>,<span class="fl">3</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="kw">for</span> <span class="op">(</span><span class="va">i</span> <span class="kw">in</span> <span class="fu"><a href="https://rdrr.io/r/base/seq.html" class="external-link">seq_len</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/nrow.html" class="external-link">ncol</a></span><span class="op">(</span><span class="va">y</span><span class="op">)</span><span class="op">)</span><span class="op">)</span> <span class="op">{</span></span>
<span>    <span class="fu"><a href="https://rdrr.io/pkg/limma/man/plotMD.html" class="external-link">plotMD</a></span><span class="op">(</span><span class="va">y</span>, column<span class="op">=</span><span class="va">i</span><span class="op">)</span></span>
<span><span class="op">}</span></span></code></pre></div>
<p><img src="multi_sample_files/figure-html/unnamed-chunk-11-1.png" width="700"></p>
<p>MDS plot</p>
<div class="sourceCode" id="cb36"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/limma/man/plotMDS.html" class="external-link">plotMDS</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/edgeR/man/cpm.html" class="external-link">cpm</a></span><span class="op">(</span><span class="va">y</span>, log<span class="op">=</span><span class="cn">TRUE</span><span class="op">)</span>, </span>
<span>    col<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/ifelse.html" class="external-link">ifelse</a></span><span class="op">(</span><span class="va">y</span><span class="op">$</span><span class="va">samples</span><span class="op">$</span><span class="va">tomato</span>, <span class="st">"red"</span>, <span class="st">"blue"</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<p><img src="multi_sample_files/figure-html/unnamed-chunk-12-1.png" width="700"></p>
<div class="sourceCode" id="cb37"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">design</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/stats/model.matrix.html" class="external-link">model.matrix</a></span><span class="op">(</span><span class="op">~</span><span class="fu"><a href="https://rdrr.io/r/base/factor.html" class="external-link">factor</a></span><span class="op">(</span><span class="va">pool</span><span class="op">)</span> <span class="op">+</span> <span class="fu"><a href="https://rdrr.io/r/base/factor.html" class="external-link">factor</a></span><span class="op">(</span><span class="va">tomato</span><span class="op">)</span>, <span class="va">y</span><span class="op">$</span><span class="va">samples</span><span class="op">)</span></span>
<span><span class="va">design</span></span></code></pre></div>
<pre><code><span><span class="co">##         (Intercept) factor(pool)4 factor(pool)5 factor(tomato)TRUE</span></span>
<span><span class="co">## Sample1           1             0             0                  1</span></span>
<span><span class="co">## Sample2           1             0             0                  0</span></span>
<span><span class="co">## Sample3           1             1             0                  1</span></span>
<span><span class="co">## Sample4           1             1             0                  0</span></span>
<span><span class="co">## Sample5           1             0             1                  1</span></span>
<span><span class="co">## Sample6           1             0             1                  0</span></span>
<span><span class="co">## attr(,"assign")</span></span>
<span><span class="co">## [1] 0 1 1 2</span></span>
<span><span class="co">## attr(,"contrasts")</span></span>
<span><span class="co">## attr(,"contrasts")$`factor(pool)`</span></span>
<span><span class="co">## [1] "contr.treatment"</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## attr(,"contrasts")$`factor(tomato)`</span></span>
<span><span class="co">## [1] "contr.treatment"</span></span></code></pre>
<div class="sourceCode" id="cb39"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">y</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/edgeR/man/estimateDisp.html" class="external-link">estimateDisp</a></span><span class="op">(</span><span class="va">y</span>, <span class="va">design</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/summary.html" class="external-link">summary</a></span><span class="op">(</span><span class="va">y</span><span class="op">$</span><span class="va">trended.dispersion</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">##    Min. 1st Qu.  Median    Mean 3rd Qu.    Max. </span></span>
<span><span class="co">## 0.01026 0.01670 0.02128 0.02020 0.02349 0.02665</span></span></code></pre>
<div class="sourceCode" id="cb41"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/edgeR/man/plotBCV.html" class="external-link">plotBCV</a></span><span class="op">(</span><span class="va">y</span><span class="op">)</span></span></code></pre></div>
<p><img src="multi_sample_files/figure-html/unnamed-chunk-14-1.png" width="700"></p>
<div class="sourceCode" id="cb42"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">fit</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/edgeR/man/glmQLFTest.html" class="external-link">glmQLFit</a></span><span class="op">(</span><span class="va">y</span>, <span class="va">design</span>, robust<span class="op">=</span><span class="cn">TRUE</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/summary.html" class="external-link">summary</a></span><span class="op">(</span><span class="va">fit</span><span class="op">$</span><span class="va">var.prior</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">##    Min. 1st Qu.  Median    Mean 3rd Qu.    Max. </span></span>
<span><span class="co">##  0.3179  0.7143  0.8538  0.8045  0.9128  1.0665</span></span></code></pre>
<div class="sourceCode" id="cb44"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/summary.html" class="external-link">summary</a></span><span class="op">(</span><span class="va">fit</span><span class="op">$</span><span class="va">df.prior</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">##    Min. 1st Qu.  Median    Mean 3rd Qu.    Max. </span></span>
<span><span class="co">##  0.2267 12.6749 12.6749 12.3390 12.6749 12.6749</span></span></code></pre>
<div class="sourceCode" id="cb46"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/edgeR/man/plotQLDisp.html" class="external-link">plotQLDisp</a></span><span class="op">(</span><span class="va">fit</span><span class="op">)</span></span></code></pre></div>
<p><img src="multi_sample_files/figure-html/unnamed-chunk-15-1.png" width="700"></p>
<div class="sourceCode" id="cb47"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">res</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/edgeR/man/glmQLFTest.html" class="external-link">glmQLFTest</a></span><span class="op">(</span><span class="va">fit</span>, coef<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/nrow.html" class="external-link">ncol</a></span><span class="op">(</span><span class="va">design</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/summary.html" class="external-link">summary</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/limma/man/decideTests.html" class="external-link">decideTests</a></span><span class="op">(</span><span class="va">res</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">##        factor(tomato)TRUE</span></span>
<span><span class="co">## Down                    8</span></span>
<span><span class="co">## NotSig               5672</span></span>
<span><span class="co">## Up                      8</span></span></code></pre>
<div class="sourceCode" id="cb49"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/edgeR/man/topTags.html" class="external-link">topTags</a></span><span class="op">(</span><span class="va">res</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## Coefficient:  factor(tomato)TRUE </span></span>
<span><span class="co">##                         logFC   logCPM          F       PValue          FDR</span></span>
<span><span class="co">## ENSMUSG00000010760 -4.3873546 9.934130 1638.59469 1.812293e-16 1.030832e-12</span></span>
<span><span class="co">## ENSMUSG00000096768  2.0690698 8.832662  356.36590 1.060836e-11 3.017019e-08</span></span>
<span><span class="co">## ENSMUSG00000035299  1.5190728 6.931325  120.14656 1.844351e-08 3.496889e-05</span></span>
<span><span class="co">## ENSMUSG00000019188 -1.0596020 7.540121   80.79795 2.373001e-07 2.526790e-04</span></span>
<span><span class="co">## ENSMUSG00000101609  1.3762700 7.241651   83.30701 2.392052e-07 2.526790e-04</span></span>
<span><span class="co">## ENSMUSG00000045410 -1.7205826 5.127802   79.31386 2.665390e-07 2.526790e-04</span></span>
<span><span class="co">## ENSMUSG00000027520  1.8008336 6.797367   83.66324 6.808994e-07 5.532794e-04</span></span>
<span><span class="co">## ENSMUSG00000042607 -0.9234911 7.340648   53.44578 2.918297e-06 2.074909e-03</span></span>
<span><span class="co">## ENSMUSG00000024423  0.8516300 7.353208   50.31429 4.145416e-06 2.619903e-03</span></span>
<span><span class="co">## ENSMUSG00000036446 -0.6031413 9.274529   41.67104 1.204523e-05 6.851324e-03</span></span></code></pre>
</div>
</div>
<div class="section level2">
<h2 id="differential-abundance">Differential Abundance<a class="anchor" aria-label="anchor" href="#differential-abundance"></a>
</h2>
<p>Section 6 of OSCA multisample.</p>
</div>
<div class="section level2">
<h2 id="session-info">Session Info<a class="anchor" aria-label="anchor" href="#session-info"></a>
</h2>
<div class="sourceCode" id="cb51"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/sessionInfo.html" class="external-link">sessionInfo</a></span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## R version 4.3.0 (2023-04-21)</span></span>
<span><span class="co">## Platform: x86_64-pc-linux-gnu (64-bit)</span></span>
<span><span class="co">## Running under: Ubuntu 22.04.2 LTS</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## Matrix products: default</span></span>
<span><span class="co">## BLAS:   /usr/lib/x86_64-linux-gnu/openblas-pthread/libblas.so.3 </span></span>
<span><span class="co">## LAPACK: /usr/lib/x86_64-linux-gnu/openblas-pthread/libopenblasp-r0.3.20.so;  LAPACK version 3.10.0</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## locale:</span></span>
<span><span class="co">##  [1] LC_CTYPE=en_US.UTF-8       LC_NUMERIC=C              </span></span>
<span><span class="co">##  [3] LC_TIME=en_US.UTF-8        LC_COLLATE=en_US.UTF-8    </span></span>
<span><span class="co">##  [5] LC_MONETARY=en_US.UTF-8    LC_MESSAGES=en_US.UTF-8   </span></span>
<span><span class="co">##  [7] LC_PAPER=en_US.UTF-8       LC_NAME=C                 </span></span>
<span><span class="co">##  [9] LC_ADDRESS=C               LC_TELEPHONE=C            </span></span>
<span><span class="co">## [11] LC_MEASUREMENT=en_US.UTF-8 LC_IDENTIFICATION=C       </span></span>
<span><span class="co">## </span></span>
<span><span class="co">## time zone: Etc/UTC</span></span>
<span><span class="co">## tzcode source: system (glibc)</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## attached base packages:</span></span>
<span><span class="co">## [1] stats4    stats     graphics  grDevices utils     datasets  methods  </span></span>
<span><span class="co">## [8] base     </span></span>
<span><span class="co">## </span></span>
<span><span class="co">## other attached packages:</span></span>
<span><span class="co">##  [1] edgeR_3.42.4                 limma_3.56.2                </span></span>
<span><span class="co">##  [3] bluster_1.10.0               batchelor_1.16.0            </span></span>
<span><span class="co">##  [5] scran_1.28.1                 scater_1.28.0               </span></span>
<span><span class="co">##  [7] ggplot2_3.4.2                scuttle_1.10.1              </span></span>
<span><span class="co">##  [9] MouseGastrulationData_1.14.0 SpatialExperiment_1.10.0    </span></span>
<span><span class="co">## [11] SingleCellExperiment_1.22.0  SummarizedExperiment_1.30.2 </span></span>
<span><span class="co">## [13] Biobase_2.60.0               GenomicRanges_1.52.0        </span></span>
<span><span class="co">## [15] GenomeInfoDb_1.36.0          IRanges_2.34.0              </span></span>
<span><span class="co">## [17] S4Vectors_0.38.1             BiocGenerics_0.46.0         </span></span>
<span><span class="co">## [19] MatrixGenerics_1.12.2        matrixStats_1.0.0           </span></span>
<span><span class="co">## </span></span>
<span><span class="co">## loaded via a namespace (and not attached):</span></span>
<span><span class="co">##   [1] RColorBrewer_1.1-3            jsonlite_1.8.5               </span></span>
<span><span class="co">##   [3] magrittr_2.0.3                ggbeeswarm_0.7.2             </span></span>
<span><span class="co">##   [5] magick_2.7.4                  farver_2.1.1                 </span></span>
<span><span class="co">##   [7] rmarkdown_2.22                fs_1.6.2                     </span></span>
<span><span class="co">##   [9] zlibbioc_1.46.0               ragg_1.2.5                   </span></span>
<span><span class="co">##  [11] vctrs_0.6.3                   memoise_2.0.1                </span></span>
<span><span class="co">##  [13] DelayedMatrixStats_1.22.1     RCurl_1.98-1.12              </span></span>
<span><span class="co">##  [15] htmltools_0.5.5               S4Arrays_1.0.4               </span></span>
<span><span class="co">##  [17] AnnotationHub_3.8.0           curl_5.0.1                   </span></span>
<span><span class="co">##  [19] BiocNeighbors_1.18.0          Rhdf5lib_1.22.0              </span></span>
<span><span class="co">##  [21] rhdf5_2.44.0                  sass_0.4.6                   </span></span>
<span><span class="co">##  [23] bslib_0.5.0                   desc_1.4.2                   </span></span>
<span><span class="co">##  [25] cachem_1.0.8                  ResidualMatrix_1.10.0        </span></span>
<span><span class="co">##  [27] igraph_1.4.3                  mime_0.12                    </span></span>
<span><span class="co">##  [29] lifecycle_1.0.3               pkgconfig_2.0.3              </span></span>
<span><span class="co">##  [31] rsvd_1.0.5                    Matrix_1.5-4.1               </span></span>
<span><span class="co">##  [33] R6_2.5.1                      fastmap_1.1.1                </span></span>
<span><span class="co">##  [35] GenomeInfoDbData_1.2.10       shiny_1.7.4                  </span></span>
<span><span class="co">##  [37] digest_0.6.31                 colorspace_2.1-0             </span></span>
<span><span class="co">##  [39] AnnotationDbi_1.62.1          rprojroot_2.0.3              </span></span>
<span><span class="co">##  [41] dqrng_0.3.0                   irlba_2.3.5.1                </span></span>
<span><span class="co">##  [43] ExperimentHub_2.8.0           textshaping_0.3.6            </span></span>
<span><span class="co">##  [45] RSQLite_2.3.1                 beachmat_2.16.0              </span></span>
<span><span class="co">##  [47] labeling_0.4.2                filelock_1.0.2               </span></span>
<span><span class="co">##  [49] fansi_1.0.4                   httr_1.4.6                   </span></span>
<span><span class="co">##  [51] compiler_4.3.0                bit64_4.0.5                  </span></span>
<span><span class="co">##  [53] withr_2.5.0                   BiocParallel_1.34.2          </span></span>
<span><span class="co">##  [55] viridis_0.6.3                 DBI_1.1.3                    </span></span>
<span><span class="co">##  [57] highr_0.10                    HDF5Array_1.28.1             </span></span>
<span><span class="co">##  [59] R.utils_2.12.2                rappdirs_0.3.3               </span></span>
<span><span class="co">##  [61] DelayedArray_0.26.3           rjson_0.2.21                 </span></span>
<span><span class="co">##  [63] tools_4.3.0                   vipor_0.4.5                  </span></span>
<span><span class="co">##  [65] beeswarm_0.4.0                interactiveDisplayBase_1.38.0</span></span>
<span><span class="co">##  [67] httpuv_1.6.11                 R.oo_1.25.0                  </span></span>
<span><span class="co">##  [69] glue_1.6.2                    rhdf5filters_1.12.1          </span></span>
<span><span class="co">##  [71] promises_1.2.0.1              grid_4.3.0                   </span></span>
<span><span class="co">##  [73] Rtsne_0.16                    cluster_2.1.4                </span></span>
<span><span class="co">##  [75] generics_0.1.3                gtable_0.3.3                 </span></span>
<span><span class="co">##  [77] R.methodsS3_1.8.2             metapod_1.8.0                </span></span>
<span><span class="co">##  [79] BiocSingular_1.16.0           ScaledMatrix_1.8.1           </span></span>
<span><span class="co">##  [81] utf8_1.2.3                    XVector_0.40.0               </span></span>
<span><span class="co">##  [83] ggrepel_0.9.3                 BiocVersion_3.17.1           </span></span>
<span><span class="co">##  [85] pillar_1.9.0                  stringr_1.5.0                </span></span>
<span><span class="co">##  [87] BumpyMatrix_1.8.0             later_1.3.1                  </span></span>
<span><span class="co">##  [89] splines_4.3.0                 dplyr_1.1.2                  </span></span>
<span><span class="co">##  [91] BiocFileCache_2.8.0           lattice_0.21-8               </span></span>
<span><span class="co">##  [93] bit_4.0.5                     tidyselect_1.2.0             </span></span>
<span><span class="co">##  [95] locfit_1.5-9.8                Biostrings_2.68.1            </span></span>
<span><span class="co">##  [97] knitr_1.43                    gridExtra_2.3                </span></span>
<span><span class="co">##  [99] xfun_0.39                     statmod_1.5.0                </span></span>
<span><span class="co">## [101] DropletUtils_1.20.0           pheatmap_1.0.12              </span></span>
<span><span class="co">## [103] stringi_1.7.12                yaml_2.3.7                   </span></span>
<span><span class="co">## [105] evaluate_0.21                 codetools_0.2-19             </span></span>
<span><span class="co">## [107] tibble_3.2.1                  BiocManager_1.30.21          </span></span>
<span><span class="co">## [109] cli_3.6.1                     uwot_0.1.14                  </span></span>
<span><span class="co">## [111] xtable_1.8-4                  systemfonts_1.0.4            </span></span>
<span><span class="co">## [113] munsell_0.5.0                 jquerylib_0.1.4              </span></span>
<span><span class="co">## [115] Rcpp_1.0.10                   dbplyr_2.3.2                 </span></span>
<span><span class="co">## [117] png_0.1-8                     parallel_4.3.0               </span></span>
<span><span class="co">## [119] ellipsis_0.3.2                pkgdown_2.0.7                </span></span>
<span><span class="co">## [121] blob_1.2.4                    sparseMatrixStats_1.12.0     </span></span>
<span><span class="co">## [123] bitops_1.0-7                  viridisLite_0.4.2            </span></span>
<span><span class="co">## [125] scales_1.2.1                  purrr_1.0.1                  </span></span>
<span><span class="co">## [127] crayon_1.5.2                  rlang_1.1.1                  </span></span>
<span><span class="co">## [129] formatR_1.14                  cowplot_1.1.1                </span></span>
<span><span class="co">## [131] KEGGREST_1.40.0</span></span></code></pre>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">

        <nav id="toc" data-toggle="toc"><h2 data-toc-skip>Contents</h2>
    </nav>
</div>

</div>



      <footer><div class="copyright">
  <p></p>
<p>Developed by Davide Risso, Ludwig Geistlinger.</p>
</div>

<div class="pkgdown">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.7.</p>
</div>

      </footer>
</div>

  


  

  </body>
</html>
